{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b299f0c5-23a6-4609-9f91-c65d0994bed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bc4ad765-e905-4d20-b63c-202f99990d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference:\n",
    "ref_annot = pd.read_csv('../data/raw/ref_annot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "00b3cbf7-9d10-427d-b534-c343a37a267c",
   "metadata": {},
   "outputs": [],
   "source": [
    "REGION_SIZE = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5bc1a977-0763-4a16-b3c3-d3a4ff789e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "WINDOW_SIZE = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebe27959-21cb-4b6e-becd-7e8bffa4c30c",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = pd.read_csv('../data/processed/g4conf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1d5bbd36-6e65-4eb7-828f-f5fdce1156f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G4\n",
       "G4     43\n",
       "UG4    31\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf['G4'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7670f0dd-f84d-45e7-88cc-0c21fa83c9f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0758224e-1102-4938-a0ea-039dffaf7c53",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2a873713-15a1-4f9e-a418-b4cc5d493e62",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation \n",
    "## Step 1\n",
    "- auto-annotate G4 by G4Hunter (https://bioinformatics.cruk.cam.ac.uk/G4Hunter/)  \n",
    "- *verificate auto-annotate G4 in human RefSeq using data from africa paper and exel*  \n",
    "\n",
    "Dataset from paper [DOI:10.1093/nar/gkw006](DOI:10.1093/nar/gkw006) (G4Hunter and *in vitro* checking G4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "db421e06-a623-44fa-b428-05fdbe10799c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:79: UserWarning: Conditional Formatting extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n",
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n"
     ]
    }
   ],
   "source": [
    "datasetG4 = pd.read_excel(\"../data/raw/SupplementaryTableS2.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53e0a8a8-7591-4154-8312-7642e4063652",
   "metadata": {},
   "outputs": [],
   "source": [
    "africaG4 = datasetG4[datasetG4['Type'] != 'notG4'][['seqnames','start','end', 'width', 'strand', 'seq', 'G4Hscore', 'Type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53468540-2d03-46a3-affa-21fa082ef1ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# africaG4.to_excel('data/processed/africaG4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8d9ec877-2751-48b6-b770-e697bbfa0dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "refG4 = pd.read_table('../data/raw/NC_012920.1 Homo sapiens mitochondrion, complete genome_DNA_hl=1_k=25_G4Hseeked_2023-09-22.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "235ce597-bb87-4274-b451-8954b8fc6ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# refG4.to_excel('data/processed/refG4.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1348a1f9-3117-4868-9e94-fa00fffa65bd",
   "metadata": {},
   "source": [
    "## Manually fill in the data  \n",
    "Compare two datasets africaG4, refG4  \n",
    "If G4 present in both datasets put **1** in column *confirm*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e30ab88-66fe-42d4-9782-203264d269f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step 2\n",
    "### Interval size analysis  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f30b178d-e771-4e53-8fa9-c3e8ee52f1c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "refG4 = pd.read_excel('../data/processed/refG4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "571255a8-f1b3-42d2-acff-528e9d0f97b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = refG4[refG4['confirmed'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "be8770a4-3c3b-4834-8d00-cb2f4b34f2a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_untill_next_G4 = g4conf['start'].drop(1).values - g4conf['end'].drop(93).values\n",
    "bp_untill_next_G4 = bp_untill_next_G4.tolist()\n",
    "bp_untill_next_G4.append(294)  #294 bp between last G4 and first G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "67a246ea-61b1-40d7-9704-a4f6257f473d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_untill_previous_G4 = g4conf['start'].drop(1).values - g4conf['end'].drop(93).values\n",
    "bp_untill_previous_G4 = bp_untill_previous_G4.tolist()\n",
    "bp_untill_previous_G4.insert(0, 294)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "182926c9-5d1e-4ab0-97af-fb4c1bb0a1f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_11760\\2987571035.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  g4conf['bp_untill_next_G4'] = bp_untill_next_G4\n",
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_11760\\2987571035.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  g4conf['bp_untill_previous_G4'] = bp_untill_previous_G4\n"
     ]
    }
   ],
   "source": [
    "g4conf['bp_untill_next_G4'] = bp_untill_next_G4\n",
    "g4conf['bp_untill_previous_G4'] = bp_untill_previous_G4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7069d949-437e-49c7-93b9-21f42b9cb1e0",
   "metadata": {},
   "source": [
    "Adding G4_id for each confirmed G4:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "32a06658-7101-495b-80fd-bf01390b77d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_11760\\169449425.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  g4conf['G4_id'] = [*range(1,g4conf.shape[0]+1)]\n"
     ]
    }
   ],
   "source": [
    "g4conf['G4_id'] = [*range(1,g4conf.shape[0]+1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dedd0502-3fe2-496c-834f-ffa3291810a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_11760\\2099197523.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  g4conf['up_win'] = pd.concat([g4conf[g4conf['bp_untill_previous_G4'] <= 100][['bp_untill_previous_G4', 'start']].diff(axis=1)['start'],\\\n"
     ]
    }
   ],
   "source": [
    "g4conf['up_win'] = pd.concat([g4conf[g4conf['bp_untill_previous_G4'] <= 100][['bp_untill_previous_G4', 'start']].diff(axis=1)['start'],\\\n",
    "                                 g4conf[g4conf['bp_untill_previous_G4'] > 100]['start']-100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2524fb81-f61b-4cc7-bd31-790d2ea3b6d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_11760\\461284617.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  g4conf['down_win'] = pd.concat([g4conf[g4conf['bp_untill_next_G4'] <= 100][['end', 'bp_untill_next_G4']].sum(axis=1), \\\n"
     ]
    }
   ],
   "source": [
    "g4conf['down_win'] = pd.concat([g4conf[g4conf['bp_untill_next_G4'] <= 100][['end', 'bp_untill_next_G4']].sum(axis=1), \\\n",
    "                               g4conf[g4conf['bp_untill_next_G4'] > 100]['end']+100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dbe53171-6128-4cfb-b51b-dc2249d1c18d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4conf.to_csv('../data/processed/g4conf.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "154b760d-666f-4ff7-be60-c93ff4d7513d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step 2  \n",
    "### New interval size:\n",
    "- Take all confirmed G4   \n",
    "- **Dont sort G4 by size between G4 >= 50 bp**  \n",
    "- Obtain different interval size (0-100bp) between G4  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e084d609-6f9e-4e99-9514-354e2f751a31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# G4_id = 38 hasnt downstream windows\n",
    "# G4_id = 39 hasnt upsteram windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d3c32786-4097-4917-bc67-17455f8e7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>confirmed</th>\n",
       "      <th>G4</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>9243</td>\n",
       "      <td>9262</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.65</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAGCCCATGACCCCTAAC</td>\n",
       "      <td>1</td>\n",
       "      <td>UG4</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>9143</td>\n",
       "      <td>9362</td>\n",
       "      <td>6</td>\n",
       "      <td>285</td>\n",
       "      <td>38</td>\n",
       "      <td>9143</td>\n",
       "      <td>9268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>9268</td>\n",
       "      <td>9293</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.15</td>\n",
       "      <td>-1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTCTCAGCCCTCCTAATGACCTCC</td>\n",
       "      <td>1</td>\n",
       "      <td>UG4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9168</td>\n",
       "      <td>9393</td>\n",
       "      <td>212</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>9262</td>\n",
       "      <td>9393</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start   end  width  \\\n",
       "52  NC_012920.1 Homo sapiens mitochondrion, comple...   9243  9262     20   \n",
       "53  NC_012920.1 Homo sapiens mitochondrion, comple...   9268  9293     26   \n",
       "\n",
       "   strand  score  max_score  threshold  window                    sequence  \\\n",
       "52      -  -1.65      -1.32          1      25        CCCAGCCCATGACCCCTAAC   \n",
       "53      -  -1.15      -1.12          1      25  CCCTCTCAGCCCTCCTAATGACCTCC   \n",
       "\n",
       "    confirmed   G4              note  start-100bp  end+100bp  \\\n",
       "52          1  UG4  less than africa         9143       9362   \n",
       "53          1  UG4               NaN         9168       9393   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  G4_id  up_win  down_win  \n",
       "52                  6                    285     38    9143      9268  \n",
       "53                212                      6     39    9262      9393  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf[(g4conf['G4_id'] == 38) | (g4conf['G4_id'] == 39)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b0c66085-6236-45a3-be2c-8cda68f4f618",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_up = []\n",
    "\n",
    "for up_win, start, G4_id in zip(g4conf['up_win'], g4conf['start'], g4conf['G4_id']):\n",
    "    \n",
    "    count_window = start - up_win - WINDOW_SIZE + 1  # count window in upstream interval (before G4)\n",
    "    number_of_win = count_window + 1                 # reverse counting of windows; the closer to G4, the smaller the number\n",
    "    \n",
    "    for i in range(up_win, start):\n",
    "        row = []\n",
    "        \n",
    "        if i+WINDOW_SIZE <= start:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]\n",
    "        else:\n",
    "            break\n",
    "#         print(i, window)    \n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            if pos != 3107:\n",
    "                window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "            else:\n",
    "                window_seq += ''\n",
    "#         print(pos, window_seq)\n",
    "\n",
    "#       ref_annot[ref_annot['Pos'].isin(window)].Nuc.value_counts()\n",
    "        \n",
    "        number_of_win -= 1\n",
    "        \n",
    "        row.append(G4_id)\n",
    "        row.append('up')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_up.append(row)\n",
    "# print(data_up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1f4775d4-f58a-4e6b-b4c2-02525afcb545",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_down = []\n",
    "\n",
    "for end, down_win, G4_id in zip(g4conf['end'], g4conf['down_win'], g4conf['G4_id']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i in range(end, down_win):\n",
    "        row = []\n",
    "        \n",
    "        if i >= g4conf['end'].iloc[-1] | i <= ref_annot['Pos'].iloc[-1]:\n",
    "            window = [*range(i, ref_annot['Pos'].iloc[-1]+1)]\n",
    "            remain = WINDOW_SIZE - (ref_annot['Pos'].iloc[-1] - i)\n",
    "            window += [*range(ref_annot['Pos'].iloc[0], remain)]\n",
    "        \n",
    "        elif i > ref_annot['Pos'].iloc[-1] and i <= down_win - WINDOW_SIZE :\n",
    "            i = i - ref_annot['Pos'].iloc[-1]\n",
    "            window = [*range(i, i + WINDOW_SIZE)]\n",
    "        \n",
    "        elif i+WINDOW_SIZE < down_win:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]\n",
    "               \n",
    "        else:\n",
    "            break\n",
    "            \n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            window_seq += ref_annot[ref_annot['Pos'] == pos].Nuc.values[0]\n",
    "#         print(window[0], window_seq)\n",
    "\n",
    "#     ref_annot[ref_annot['Pos'].isin(window)].Nuc.value_counts()\n",
    "\n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('down')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_down.append(row)\n",
    "# print(data_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4790d0b7-9a65-4703-b99b-b7f5544182bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data_up + data_down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "56d20465-de48-480a-8fb5-b7e3ed3fed00",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win = pd.DataFrame(data=data, columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7920c25f-e851-4809-bc3a-89a89c2a892e",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_win = df_win['sequence'].apply(lambda x: Counter(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1b8de43a-731b-4dd8-9301-3f73e0dfa435",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_df_win = pd.DataFrame.from_records(nuc_count_win).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "679287e2-7c73-4f2e-9c20-e6c2f6f2b110",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win = df_win.join(nuc_count_df_win).fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59f23ba2-3fa3-4d61-81f3-64c7d7e0ed3b",
   "metadata": {},
   "source": [
    "### Finished dataset with nucleotide frequencies in ref_seq windows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b5b63c49-b04d-42f0-803f-16598e995f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win.to_csv('../data/processed/df_win.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78f5f2a-7b79-4043-aa17-e44b5ac0accf",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16545ac6-9c5f-48cd-8604-0e599d201848",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24f5a999-ac7e-48df-a16c-17e5552db03b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e2d190-41c1-47a1-bc37-036d401b6e6f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb990c7-119d-4e5b-a18b-f2e9cc2c0738",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b2d21ca-388c-4433-80b5-13bbc6bb1f9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "afef4e00-7932-4f3a-a496-c3b6e0cc77b9",
   "metadata": {},
   "source": [
    "# Data preparation around G4 for Cancer \n",
    "### Observed mutations Cancer dataset:\n",
    "\n",
    "df_win_cancer - observed mutations from Cancer dataset with nucleotide frequencies from df_win\n",
    "\n",
    "Add mutations from Cancer dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "935484e8-29c6-4e38-a802-84153feca618",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "951d376a-10a3-4648-869b-f694afef0e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win = []\n",
    "for window_start, window_end in zip(df_win['window_start'], df_win['window_end']):\n",
    "    row = [*range(window_start, window_end)]\n",
    "    Mut_win.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f566e6fc-a507-4cf8-8e1a-a30579b1d49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df = pd.DataFrame(Mut_win)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "09046d33-5408-49a1-bc76-75002900df83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>['A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>['A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>['A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>['A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>['A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;G]C' 'A[A&gt;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0\n",
       "0  ['A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>...\n",
       "1  ['A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>...\n",
       "2  ['A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>...\n",
       "3  ['A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>...\n",
       "4  ['A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>G]C' 'A[A>..."
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Mut_win_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6a91687c-34aa-4d13-b718-8dc6f696a3fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df[0] = Mut_win_df[0].str.findall(\"[A-T]>[A-T]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f86300eb-96ad-49e6-970b-f67ed2d9d0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutwin = pd.DataFrame.from_records(Mut_win_df[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "14f6dfe2-3095-4f3c-8822-29cf6cbcb907",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer = df_win.join(obsMutwin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dd7b7c40-d462-43cd-a31b-df818afeae53",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer['number_of_Mut'] = df_win_cancer[['A>C','A>G','A>T',\n",
    "                                                'C>A','C>G','C>T',\n",
    "                                                'G>A','G>C','G>T',\n",
    "                                                'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2ba2f37-3401-4376-beaf-edf92dcee049",
   "metadata": {},
   "source": [
    "### Finished dataset with observed mutations from cancer_df:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "21fe33c6-e94b-4b43-878e-8823fbb60def",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer.to_csv('../data/processed/df_win_cancer.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e72b704d-e1ba-444f-97fd-4155cc97d104",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation around G4 for Human Global Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5857365f-77a5-444f-ae4d-c4d77dbc5d0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt = pd.read_table('../data/raw/fulltreeCodons.csv', sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c7883c2f-3160-4909-aed1-05286d08bc5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/raw/hgt_alignment.blast') as fin:\n",
    "    while True:\n",
    "        line = fin.readline()\n",
    "        if 'Strand=Plus/Plus' in line:\n",
    "            fin.readline()\n",
    "            break\n",
    "    qseq = ''\n",
    "    rods = ''\n",
    "    sseq = ''\n",
    "    query, sbjct = 'Query', 'Sbjct'\n",
    "    i = 0\n",
    "    size = 60\n",
    "    start = 14\n",
    "    while query.startswith('Query') and sbjct.startswith('Sbjct'):\n",
    "        i += 1\n",
    "        query = fin.readline().strip()\n",
    "        cur_rods = fin.readline()\n",
    "        sbjct = fin.readline().strip()\n",
    "\n",
    "        if len(query[start:start+size]) == size:\n",
    "            qseq += query[start:start+size]\n",
    "            sseq += sbjct[start:start+size]\n",
    "            rods += cur_rods[start:start+size]\n",
    "        else:\n",
    "            pos = re.search('Query\\s{2}(\\d+)\\s+[\\w-]+\\s+(\\d+)', query).groups()\n",
    "            size = int(pos[1]) - int(pos[0]) + 1\n",
    "            \n",
    "            qseq += query[start:start+size]\n",
    "            sseq += sbjct[start:start+size]\n",
    "            rods += cur_rods[start:start+size]\n",
    "            break\n",
    "            \n",
    "        fin.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "95af0cf5-e3f2-4ac7-85bd-be659f34fd79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sseq = 16568 qseq = 16568\n"
     ]
    }
   ],
   "source": [
    "print('sseq =', len(sseq), 'qseq =', len(qseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "06f4b065-c8ea-48e2-b49e-2312f36be982",
   "metadata": {},
   "outputs": [],
   "source": [
    "sseq = 'G' + sseq\n",
    "qseq = 'G' + qseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3b7d7077-ca3c-4c2c-b583-276373dd4d9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sseq = 16569 qseq = 16569\n"
     ]
    }
   ],
   "source": [
    "print('sseq =', len(sseq), 'qseq =', len(qseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a8b7ee0c-f238-4d38-8e77-58d16cf2958f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_hgt = {i: qseq[i] for i in range(len(qseq))}\n",
    "dict_ref = {i: sseq[i] for i in range(len(sseq))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "05e513a6-65b2-49df-8906-51fd00ce3dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = pd.DataFrame.from_dict(dict_ref, orient='index', columns=['nuc_ref']) \\\n",
    "    .reset_index() \\\n",
    "    .rename(columns={'index':'Pos'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4113e24e-739f-4efb-a05b-58d1b6af7953",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref['nuc_hgt'] = df_ali_hgt_to_ref['Pos'].map(dict_hgt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3ecc70dc-76b2-410f-9340-e1c726c71ab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref['Pos'] = range(1, len(df_ali_hgt_to_ref['Pos'])+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "fd09aabc-b2fa-464e-8d39-2c9d2fabcbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = df_ali_hgt_to_ref.replace({'c':'C', 'n':'N', 't':'T', 'g':'G', 'a':'A'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133a84d6-868b-422b-bb73-92a65f903216",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique = df_hgt[['ref_pos', 'nuc_ref_in_ali']].sort_values(by='ref_pos').drop_duplicates(subset=['ref_pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5b2678c0-a075-4664-bb83-ce3542c096ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique = df_unique.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f1dae751-7d37-4ee4-9c1d-07ed2f003e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique_sort = df_unique[~((df_unique['nuc_ref_in_ali'] == '-') | (df_unique['nuc_ref_in_ali'] == 'N'))][['ref_pos', 'nuc_ref_in_ali']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "884975a1-7706-4217-8bab-de6f5baf659a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique_sort = df_unique_sort.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "592b006a-8361-4657-997c-0d91b5fa2ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Detele all posirions with N, -, and first position\n",
    "a = df_ali_hgt_to_ref[~((df_ali_hgt_to_ref['nuc_hgt'] == 'N') | \n",
    "                        (df_ali_hgt_to_ref['nuc_hgt'] == '-') | \n",
    "                        (df_ali_hgt_to_ref['Pos'] == 1))].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "be87253e-954f-4475-a221-07be82cb4a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos_ref_to_hgt = pd.concat([a, df_unique_sort], axis=1).rename(columns={'ref_pos':'hgt_ref_pos'}).set_index('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "972e8755-03f2-44b4-adb2-12309cd25b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos_ref_to_hgt.index.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7e9e6eac-96f0-4624-902e-e108e2c3a163",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = df_ali_hgt_to_ref.join(df_pos_ref_to_hgt, how='left', rsuffix='2') \\\n",
    "    .drop(columns=['Pos2', 'nuc_ref2', 'nuc_hgt2', 'nuc_hgt']) \\\n",
    "    .astype({'hgt_ref_pos':'Int64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0b77155c-c82c-4d30-8b00-89795bc6100f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add N in table\n",
    "df_ali_hgt_to_ref.iloc[3106, df_ali_hgt_to_ref.columns.get_loc('hgt_ref_pos')] = 3106\n",
    "df_ali_hgt_to_ref.iloc[3106, df_ali_hgt_to_ref.columns.get_loc('nuc_ref_in_ali')] = 'N'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "71fff04b-9d3a-48e0-8421-bb94aca06688",
   "metadata": {},
   "outputs": [],
   "source": [
    "for start, end, G4_id in zip(g4conf['start'], g4conf['end'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(start, end+1)]),'G4'] = 1 #is position G4? 1-yes, 0-no\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(start, end+1)]),'G4_id'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c9df48fd-023f-4672-a35c-4a9e9f9e9342",
   "metadata": {},
   "outputs": [],
   "source": [
    "for up_win, start, G4_id in zip(g4conf['up_win'], g4conf['start'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(up_win, start+1)]),'up_win'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f6abbe4e-ed75-4093-8fa2-c2ac748d418e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for end, down_win, G4_id in zip(g4conf['end'], g4conf['down_win'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(end, down_win+1)]),'down_win'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0ae3a737-8c80-4d9a-a351-3ac0816c3108",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref[['G4', 'G4_id', 'up_win', 'down_win']] = df_ali_hgt_to_ref[['G4',  'G4_id', \n",
    "                                                                              'up_win', 'down_win']].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "30cff4ed-0b9b-4c8d-ac76-79c87b33ed96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>Pos</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>G4</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos  Pos  G4_id   G4  up_win  down_win\n",
       "0     <NA>    1    0.0  0.0     0.0       0.0\n",
       "1        1    2    0.0  0.0     0.0       0.0\n",
       "2        2    3    0.0  0.0     0.0       0.0\n",
       "3        3    4    0.0  0.0     0.0       0.0\n",
       "4        4    5    0.0  0.0     0.0       0.0"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ali_hgt_to_ref[['hgt_ref_pos', 'Pos', 'G4_id', 'G4', 'up_win', 'down_win']] \\\n",
    "    .rename(columns={'hgt_ref_pos':'ref_pos'}).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "19e55ecf-5372-4392-82a3-2889866402c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt_g4 = df_hgt \\\n",
    "    .merge(df_ali_hgt_to_ref[['hgt_ref_pos', 'Pos', 'G4_id', 'G4', 'up_win', 'down_win']] \\\n",
    "           .rename(columns={'hgt_ref_pos':'ref_pos'}), how='left') \\\n",
    "    .sort_values(by='ref_pos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b38c2f91-81f6-4339-9ec5-2f58bc3d65c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt_g4['Mut'] = df_hgt_g4.ancestor.str[2] + '>' + df_hgt_g4.descendant.str[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b56cd6d7-8493-4db9-8175-8a99fbc53970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_hgt_g4.to_csv('data/processed/df_hgt_g4.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7078b016-cbd2-4057-b122-5db717a75411",
   "metadata": {},
   "source": [
    "### Collect observed mutations from HGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d1e14f14-0750-4d91-92c8-0b80449d7504",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_13856\\3731285902.py:1: DtypeWarning: Columns (10,11,12,14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')\n"
     ]
    }
   ],
   "source": [
    "df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e61e6ff7-e62d-4f33-b59c-2cb67a86a7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract mutations without gaps (-) and other (R, M, N, ect) bases in mut position:\n",
    "df_hgt_g4 = df_hgt_g4[\n",
    "    (df_hgt_g4.ancestor.str.lower().str.fullmatch(\"[acgt]{5}\")) &\n",
    "    (df_hgt_g4.descendant.str.lower().str.fullmatch(\"[acgt]{5}\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "cf42ee54-8933-4a17-9547-c6846dfff594",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_up = []\n",
    "for up_win, Pos, Mut in zip(df_hgt_g4['up_win'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if up_win > 0:\n",
    "        row.append(up_win)\n",
    "        row.append('up')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_up.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "63a09fcb-4f8c-422c-9dff-12bfeffd72c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_down = []\n",
    "for down_win, Pos, Mut in zip(df_hgt_g4['down_win'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if down_win > 0:\n",
    "        row.append(down_win)\n",
    "        row.append('down')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_down.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8e4fe658-50cc-4fc9-b12b-f1b568b975db",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt = obs_mut_hgt_up + obs_mut_hgt_down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "40aba95a-b118-4632-ba7b-fe5de90f037e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_obs_mut_hgt = pd.DataFrame(data=obs_mut_hgt, columns=['G4_id', 'side', 'Pos', 'Mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "7c2726ca-0560-45ed-8bd3-3bd17c29e9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win = pd.read_csv('data/processed/df_win.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4a788a6f-9920-4132-811e-8cbb0957962a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_hgt_win = []\n",
    "for window_start, window_end in zip(df_win['window_start'], df_win['window_end']):\n",
    "    row = []\n",
    "    \n",
    "    mut = df_obs_mut_hgt[(df_obs_mut_hgt['Pos'] >= window_start) & (df_obs_mut_hgt['Pos'] <= window_end)].Mut.tolist()\n",
    "    \n",
    "    row.append(mut)\n",
    "    \n",
    "    Mut_hgt_win.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "85104fc9-f0fb-473a-bffb-b7afa2cbd888",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Mut_hgt_win = pd.DataFrame(data=Mut_hgt_win)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "84f9c9e7-fde8-4607-8e07-458dcf2d94f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutG4_hgt = pd.DataFrame.from_records(df_Mut_hgt_win[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "f52fd3a1-ac53-46ea-9248-17057f847d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt = df_win.join(obsMutG4_hgt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3d5fcbb3-b07e-43a3-988b-2ef6b6db3dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt['number_of_Mut'] = df_win_hgt[['A>C','A>G','A>T','C>A','C>G','C>T','G>A','G>C','G>T','T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "2dbcc008-a0f0-47b4-b3f4-a482361d577e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_hgt.to_csv('data/processed/df_win_hgt.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "857d48de-74a0-4f74-9ae7-ee2b8f52fe45",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation within G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3aefd62-b76d-4267-ade4-d5703360467b",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = pd.read_csv('../data/processed/g4conf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a470c577-d88d-4969-9f19-4143ba4aeaec",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_within = []\n",
    "\n",
    "for start, end, G4_id, sequence in zip(g4conf['start'], g4conf['end'], g4conf['G4_id'], g4conf['sequence']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i, j in zip(range(start, end), range(len(sequence))):\n",
    "        row = []\n",
    "        \n",
    "        if i+WINDOW_SIZE < end:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            break\n",
    "            \n",
    "        window_seq = ''\n",
    "        window_seq += sequence[j: j + WINDOW_SIZE]\n",
    "\n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('within')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_within.append(row)\n",
    "# print(data_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "231dc2e7-3b3a-4af8-8391-5125cb8bdb78",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within = pd.DataFrame(data=data_within, \n",
    "                             columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "029c1751-4fb5-41f8-98a3-86881c2e01c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_win_within = df_win_within['sequence'].apply(lambda x: Counter(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "856e6540-7210-4d8c-9314-c530af7f297f",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_df_win_within = pd.DataFrame.from_records(nuc_count_win_within).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "8f628500-a3a8-40c3-a2e8-25aeb81493de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within = df_win_within.join(nuc_count_df_win_within).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d07639e6-e5e9-448b-bf8a-7af07026b60d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_within.to_csv('../data/processed/df_win_within.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13ad1802-5217-4408-839d-c858576e0737",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation Cancer within G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "63e4f3ca-4b41-46fa-b86d-009e1122a1ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('data/raw/mutations.csv')\n",
    "df_win_within = pd.read_csv('data/processed/df_win_within.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "d1677ac1-d462-43b4-810b-d0fa746f473f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_within = []\n",
    "for window_start, window_end in zip(df_win_within['window_start'], df_win_within['window_end']):\n",
    "    row = [*range(window_start, window_end)]\n",
    "    Mut_win_within.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "f272ebe7-4ce9-4070-9d6e-29f6bb4ea390",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_within = pd.DataFrame(Mut_win_within)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d3111899-fb7e-4132-b7ac-29d8ce680029",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_within[0] = Mut_win_df_within[0].str.findall(\"[A-T]>[A-T]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e0ddb35b-5df1-42d4-a63e-a0249161b9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutwin_within = pd.DataFrame.from_records(Mut_win_df_within[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "d1de0407-1297-42e9-a6d5-026aedb9ff37",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within = df_win_within.join(obsMutwin_within)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c7e030f8-161c-4a08-8464-4c573aa31d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within['number_of_Mut'] = df_win_cancer_within[['A>C','A>G','A>T',\n",
    "                                                              'C>A','C>G','C>T',\n",
    "                                                              'G>A','G>C','G>T',\n",
    "                                                              'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "afd9f9b3-e899-41d3-b0f2-5def69ae2211",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id    side  number_of_win  window_start  window_end    sequence    A  \\\n",
       "0      1  within              1           285         294  CAAAAAATTT  6.0   \n",
       "1      1  within              2           286         295  AAAAAATTTC  6.0   \n",
       "2      1  within              3           287         296  AAAAATTTCC  5.0   \n",
       "3      1  within              4           288         297  AAAATTTCCA  5.0   \n",
       "4      1  within              5           289         298  AAATTTCCAC  4.0   \n",
       "\n",
       "     C    G    T  ...  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G  \\\n",
       "0  1.0  0.0  3.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  2.0  0.0   \n",
       "1  1.0  0.0  3.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "2  2.0  0.0  3.0  ...  0.0  0.0  1.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "3  2.0  0.0  3.0  ...  0.0  0.0  4.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "4  3.0  0.0  3.0  ...  0.0  0.0  4.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0            2.0  \n",
       "1            4.0  \n",
       "2            5.0  \n",
       "3            8.0  \n",
       "4           10.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_cancer_within.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "57442f52-e2d3-411f-b052-a27e2042153b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_within.to_csv('data/processed/df_win_cancer_within.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8e7ebe-2140-43f2-98fb-edd20da2a679",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation HGT within G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "566b3bca-cc92-4185-a00c-9ea2cf475cf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_13856\\486489025.py:1: DtypeWarning: Columns (10,11,12,14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')\n"
     ]
    }
   ],
   "source": [
    "df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')\n",
    "df_win_within = pd.read_csv('data/processed/df_win_within.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "a121394c-bcfa-4d38-81b1-91499ca4e6d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract mutations without gaps (-) and other (R, M, N, ect) bases in mut position:\n",
    "df_hgt_g4 = df_hgt_g4[\n",
    "    (df_hgt_g4.ancestor.str.lower().str.fullmatch(\"[acgt]{5}\")) &\n",
    "    (df_hgt_g4.descendant.str.lower().str.fullmatch(\"[acgt]{5}\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "8993af2e-c946-4091-b0d4-e9f2e9069efa",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_within = []\n",
    "for G4, G4_id, Pos, Mut in zip(df_hgt_g4['G4'], df_hgt_g4['G4_id'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if G4 > 0:\n",
    "        row.append(G4_id)\n",
    "        row.append('within')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_within.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "ebad643b-3a2d-4e35-9cf8-490e40c47045",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_obs_mut_hgt_within = pd.DataFrame(data=obs_mut_hgt_within, columns=['G4_id', 'side', 'Pos', 'Mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "41c4894b-8a56-4a65-b724-11dd68033c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_hgt_win_within = []\n",
    "for window_start, window_end in zip(df_win_within['window_start'], df_win_within['window_end']):\n",
    "    row = []\n",
    "    \n",
    "    mut = df_obs_mut_hgt_within[(df_obs_mut_hgt_within['Pos'] >= window_start) & \n",
    "        (df_obs_mut_hgt_within['Pos'] <= window_end)]['Mut'].tolist()\n",
    "\n",
    "    row.append(mut)\n",
    "    \n",
    "    Mut_hgt_win_within.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "fe2c38b7-a5c0-4e61-b46d-6c0fc400fae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Mut_hgt_win_within = pd.DataFrame(data=Mut_hgt_win_within)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a5d3123f-c7c2-4486-9352-a69db93e328a",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutG4_hgt_within = pd.DataFrame.from_records(df_Mut_hgt_win_within[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "c3297f99-62d0-46c3-b726-1857efd4fc08",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within = df_win_within.join(obsMutG4_hgt_within)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "a9290ce1-2c53-46c2-a645-3cd32ae3ace9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within['number_of_Mut'] = df_win_hgt_within[['A>C','A>G','A>T',\n",
    "                                                        'C>A','C>G','C>T',\n",
    "                                                        'G>A','G>C','G>T',\n",
    "                                                        'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "919ed189-6d30-4cc3-afd9-5459016a6f0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>539.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id    side  number_of_win  window_start  window_end    sequence    A  \\\n",
       "0      1  within              1           285         294  CAAAAAATTT  6.0   \n",
       "1      1  within              2           286         295  AAAAAATTTC  6.0   \n",
       "2      1  within              3           287         296  AAAAATTTCC  5.0   \n",
       "3      1  within              4           288         297  AAAATTTCCA  5.0   \n",
       "4      1  within              5           289         298  AAATTTCCAC  4.0   \n",
       "\n",
       "     C    G    T  ...   C>A  C>G    C>T   G>A  G>C  G>T  T>A    T>C  T>G  \\\n",
       "0  1.0  0.0  3.0  ...   0.0  0.0   76.0   5.0  0.0  0.0  8.0   88.0  0.0   \n",
       "1  1.0  0.0  3.0  ...  26.0  0.0  128.0   5.0  0.0  0.0  9.0  140.0  0.0   \n",
       "2  2.0  0.0  3.0  ...  28.0  0.0  152.0   5.0  0.0  0.0  9.0  163.0  0.0   \n",
       "3  2.0  0.0  3.0  ...  34.0  0.0  152.0  72.0  0.0  0.0  9.0  163.0  0.0   \n",
       "4  3.0  0.0  3.0  ...  35.0  0.0  152.0  67.0  0.0  0.0  9.0  163.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0          189.0  \n",
       "1          346.0  \n",
       "2          397.0  \n",
       "3          548.0  \n",
       "4          539.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_hgt_within.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "09e7eedb-e37d-42d5-8d43-593e49dce9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_hgt_within.to_csv('data/processed/df_win_hgt_within.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64bf7068-990f-4cbb-a59e-04bf1414797f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "409edb6e-a55c-4d47-aebd-61fad4ab0158",
   "metadata": {},
   "source": [
    "# Data preparation within G4 with sides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60592576-eb4b-4bca-8cd0-be06b7c261d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = pd.read_csv('../data/processed/g4conf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f8c11d68-4420-4685-91ac-cfc64a21cb1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_within_start = []\n",
    "\n",
    "for start, end, G4_id, sequence in zip(g4conf['start'], g4conf['end'], g4conf['G4_id'], g4conf['sequence']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i, j in zip(range(start, end), range(len(sequence))):\n",
    "        row = []\n",
    "        \n",
    "        if i+WINDOW_SIZE <= end:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            break\n",
    "        \n",
    "        # print(i, window, WINDOW_SIZE, end)    \n",
    "        window_seq = ''\n",
    "        window_seq += sequence[j: j + WINDOW_SIZE]\n",
    "\n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('within_start')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_within_start.append(row)\n",
    "# print(data_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a446ca00-5c5b-421f-aa28-1ad665348cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_within_end = []\n",
    "\n",
    "for start, end, G4_id, sequence in zip(g4conf['start'], g4conf['end'], g4conf['G4_id'], g4conf['sequence']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i, j in zip(range(end, start, -1), range(len(sequence))):\n",
    "        row = []\n",
    "\n",
    "        \n",
    "        if i+WINDOW_SIZE > end:\n",
    "            continue  \n",
    "        elif (i+WINDOW_SIZE <= end):\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            continue\n",
    "        # print(window)\n",
    "        \n",
    "        window_seq = ''\n",
    "        window_seq += sequence[::-1][j-10:j+WINDOW_SIZE-10][::-1]   #window_seq from the end. works with '-10' \n",
    "\n",
    "        # print(i, j)\n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('within_end')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_within_end.append(row)\n",
    "# print(data_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b278d206-8cd3-4be7-827a-413e4e9642c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_within_sides = data_within_start + data_within_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b67d52dc-479b-4a29-8a73-852827989f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within_sides = pd.DataFrame(data=data_within_sides, \n",
    "                             columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce86e47d-fd7a-4da9-bb7e-5af5151cc99f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2816</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>1</td>\n",
       "      <td>322</td>\n",
       "      <td>331</td>\n",
       "      <td>GCCACAGCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2817</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>2</td>\n",
       "      <td>321</td>\n",
       "      <td>330</td>\n",
       "      <td>GGCCACAGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2818</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>3</td>\n",
       "      <td>320</td>\n",
       "      <td>329</td>\n",
       "      <td>TGGCCACAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2819</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>4</td>\n",
       "      <td>319</td>\n",
       "      <td>328</td>\n",
       "      <td>CTGGCCACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2820</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>5</td>\n",
       "      <td>318</td>\n",
       "      <td>327</td>\n",
       "      <td>TCTGGCCACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2821</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>6</td>\n",
       "      <td>317</td>\n",
       "      <td>326</td>\n",
       "      <td>TTCTGGCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2822</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>7</td>\n",
       "      <td>316</td>\n",
       "      <td>325</td>\n",
       "      <td>CTTCTGGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>8</td>\n",
       "      <td>315</td>\n",
       "      <td>324</td>\n",
       "      <td>GCTTCTGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2824</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>9</td>\n",
       "      <td>314</td>\n",
       "      <td>323</td>\n",
       "      <td>CGCTTCTGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2825</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>10</td>\n",
       "      <td>313</td>\n",
       "      <td>322</td>\n",
       "      <td>CCGCTTCTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2826</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>11</td>\n",
       "      <td>312</td>\n",
       "      <td>321</td>\n",
       "      <td>CCCGCTTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2827</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>12</td>\n",
       "      <td>311</td>\n",
       "      <td>320</td>\n",
       "      <td>CCCCGCTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2828</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>13</td>\n",
       "      <td>310</td>\n",
       "      <td>319</td>\n",
       "      <td>CCCCCGCTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2829</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>14</td>\n",
       "      <td>309</td>\n",
       "      <td>318</td>\n",
       "      <td>TCCCCCGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2830</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>15</td>\n",
       "      <td>308</td>\n",
       "      <td>317</td>\n",
       "      <td>CTCCCCCGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2831</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>16</td>\n",
       "      <td>307</td>\n",
       "      <td>316</td>\n",
       "      <td>CCTCCCCCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2832</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>17</td>\n",
       "      <td>306</td>\n",
       "      <td>315</td>\n",
       "      <td>CCCTCCCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2833</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>18</td>\n",
       "      <td>305</td>\n",
       "      <td>314</td>\n",
       "      <td>CCCCTCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2834</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>19</td>\n",
       "      <td>304</td>\n",
       "      <td>313</td>\n",
       "      <td>CCCCCTCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>20</td>\n",
       "      <td>303</td>\n",
       "      <td>312</td>\n",
       "      <td>CCCCCCTCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2836</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>21</td>\n",
       "      <td>302</td>\n",
       "      <td>311</td>\n",
       "      <td>CCCCCCCTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2837</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>22</td>\n",
       "      <td>301</td>\n",
       "      <td>310</td>\n",
       "      <td>ACCCCCCCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2838</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>23</td>\n",
       "      <td>300</td>\n",
       "      <td>309</td>\n",
       "      <td>AACCCCCCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2839</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>24</td>\n",
       "      <td>299</td>\n",
       "      <td>308</td>\n",
       "      <td>AAACCCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2840</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>25</td>\n",
       "      <td>298</td>\n",
       "      <td>307</td>\n",
       "      <td>CAAACCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2841</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>26</td>\n",
       "      <td>297</td>\n",
       "      <td>306</td>\n",
       "      <td>CCAAACCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2842</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>27</td>\n",
       "      <td>296</td>\n",
       "      <td>305</td>\n",
       "      <td>ACCAAACCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2843</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>28</td>\n",
       "      <td>295</td>\n",
       "      <td>304</td>\n",
       "      <td>CACCAAACCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2844</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>29</td>\n",
       "      <td>294</td>\n",
       "      <td>303</td>\n",
       "      <td>CCACCAAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2845</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>30</td>\n",
       "      <td>293</td>\n",
       "      <td>302</td>\n",
       "      <td>TCCACCAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2846</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>31</td>\n",
       "      <td>292</td>\n",
       "      <td>301</td>\n",
       "      <td>TTCCACCAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2847</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>32</td>\n",
       "      <td>291</td>\n",
       "      <td>300</td>\n",
       "      <td>TTTCCACCAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2848</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>33</td>\n",
       "      <td>290</td>\n",
       "      <td>299</td>\n",
       "      <td>ATTTCCACCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2849</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>34</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AATTTCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2850</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>35</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2851</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>36</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2852</th>\n",
       "      <td>1</td>\n",
       "      <td>within_end</td>\n",
       "      <td>37</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id        side  number_of_win  window_start  window_end    sequence\n",
       "2816      1  within_end              1           322         331  GCCACAGCAC\n",
       "2817      1  within_end              2           321         330  GGCCACAGCA\n",
       "2818      1  within_end              3           320         329  TGGCCACAGC\n",
       "2819      1  within_end              4           319         328  CTGGCCACAG\n",
       "2820      1  within_end              5           318         327  TCTGGCCACA\n",
       "2821      1  within_end              6           317         326  TTCTGGCCAC\n",
       "2822      1  within_end              7           316         325  CTTCTGGCCA\n",
       "2823      1  within_end              8           315         324  GCTTCTGGCC\n",
       "2824      1  within_end              9           314         323  CGCTTCTGGC\n",
       "2825      1  within_end             10           313         322  CCGCTTCTGG\n",
       "2826      1  within_end             11           312         321  CCCGCTTCTG\n",
       "2827      1  within_end             12           311         320  CCCCGCTTCT\n",
       "2828      1  within_end             13           310         319  CCCCCGCTTC\n",
       "2829      1  within_end             14           309         318  TCCCCCGCTT\n",
       "2830      1  within_end             15           308         317  CTCCCCCGCT\n",
       "2831      1  within_end             16           307         316  CCTCCCCCGC\n",
       "2832      1  within_end             17           306         315  CCCTCCCCCG\n",
       "2833      1  within_end             18           305         314  CCCCTCCCCC\n",
       "2834      1  within_end             19           304         313  CCCCCTCCCC\n",
       "2835      1  within_end             20           303         312  CCCCCCTCCC\n",
       "2836      1  within_end             21           302         311  CCCCCCCTCC\n",
       "2837      1  within_end             22           301         310  ACCCCCCCTC\n",
       "2838      1  within_end             23           300         309  AACCCCCCCT\n",
       "2839      1  within_end             24           299         308  AAACCCCCCC\n",
       "2840      1  within_end             25           298         307  CAAACCCCCC\n",
       "2841      1  within_end             26           297         306  CCAAACCCCC\n",
       "2842      1  within_end             27           296         305  ACCAAACCCC\n",
       "2843      1  within_end             28           295         304  CACCAAACCC\n",
       "2844      1  within_end             29           294         303  CCACCAAACC\n",
       "2845      1  within_end             30           293         302  TCCACCAAAC\n",
       "2846      1  within_end             31           292         301  TTCCACCAAA\n",
       "2847      1  within_end             32           291         300  TTTCCACCAA\n",
       "2848      1  within_end             33           290         299  ATTTCCACCA\n",
       "2849      1  within_end             34           289         298  AATTTCCACC\n",
       "2850      1  within_end             35           288         297  AAATTTCCAC\n",
       "2851      1  within_end             36           287         296  AAAATTTCCA\n",
       "2852      1  within_end             37           286         295  AAAAATTTCC"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_within_sides[(df_win_within_sides['side'] == 'within_end') & (df_win_within_sides['G4_id'] == 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6add867f-2342-4282-b411-460fa634443e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>6</td>\n",
       "      <td>290</td>\n",
       "      <td>299</td>\n",
       "      <td>AATTTCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>7</td>\n",
       "      <td>291</td>\n",
       "      <td>300</td>\n",
       "      <td>ATTTCCACCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>8</td>\n",
       "      <td>292</td>\n",
       "      <td>301</td>\n",
       "      <td>TTTCCACCAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>9</td>\n",
       "      <td>293</td>\n",
       "      <td>302</td>\n",
       "      <td>TTCCACCAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>10</td>\n",
       "      <td>294</td>\n",
       "      <td>303</td>\n",
       "      <td>TCCACCAAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>11</td>\n",
       "      <td>295</td>\n",
       "      <td>304</td>\n",
       "      <td>CCACCAAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>12</td>\n",
       "      <td>296</td>\n",
       "      <td>305</td>\n",
       "      <td>CACCAAACCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>13</td>\n",
       "      <td>297</td>\n",
       "      <td>306</td>\n",
       "      <td>ACCAAACCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>14</td>\n",
       "      <td>298</td>\n",
       "      <td>307</td>\n",
       "      <td>CCAAACCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>15</td>\n",
       "      <td>299</td>\n",
       "      <td>308</td>\n",
       "      <td>CAAACCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>16</td>\n",
       "      <td>300</td>\n",
       "      <td>309</td>\n",
       "      <td>AAACCCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>17</td>\n",
       "      <td>301</td>\n",
       "      <td>310</td>\n",
       "      <td>AACCCCCCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>18</td>\n",
       "      <td>302</td>\n",
       "      <td>311</td>\n",
       "      <td>ACCCCCCCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>19</td>\n",
       "      <td>303</td>\n",
       "      <td>312</td>\n",
       "      <td>CCCCCCCTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>20</td>\n",
       "      <td>304</td>\n",
       "      <td>313</td>\n",
       "      <td>CCCCCCTCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>21</td>\n",
       "      <td>305</td>\n",
       "      <td>314</td>\n",
       "      <td>CCCCCTCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>22</td>\n",
       "      <td>306</td>\n",
       "      <td>315</td>\n",
       "      <td>CCCCTCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>23</td>\n",
       "      <td>307</td>\n",
       "      <td>316</td>\n",
       "      <td>CCCTCCCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>24</td>\n",
       "      <td>308</td>\n",
       "      <td>317</td>\n",
       "      <td>CCTCCCCCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>25</td>\n",
       "      <td>309</td>\n",
       "      <td>318</td>\n",
       "      <td>CTCCCCCGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>26</td>\n",
       "      <td>310</td>\n",
       "      <td>319</td>\n",
       "      <td>TCCCCCGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>27</td>\n",
       "      <td>311</td>\n",
       "      <td>320</td>\n",
       "      <td>CCCCCGCTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>28</td>\n",
       "      <td>312</td>\n",
       "      <td>321</td>\n",
       "      <td>CCCCGCTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>29</td>\n",
       "      <td>313</td>\n",
       "      <td>322</td>\n",
       "      <td>CCCGCTTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>30</td>\n",
       "      <td>314</td>\n",
       "      <td>323</td>\n",
       "      <td>CCGCTTCTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>31</td>\n",
       "      <td>315</td>\n",
       "      <td>324</td>\n",
       "      <td>CGCTTCTGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>32</td>\n",
       "      <td>316</td>\n",
       "      <td>325</td>\n",
       "      <td>GCTTCTGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>33</td>\n",
       "      <td>317</td>\n",
       "      <td>326</td>\n",
       "      <td>CTTCTGGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>34</td>\n",
       "      <td>318</td>\n",
       "      <td>327</td>\n",
       "      <td>TTCTGGCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>35</td>\n",
       "      <td>319</td>\n",
       "      <td>328</td>\n",
       "      <td>TCTGGCCACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>36</td>\n",
       "      <td>320</td>\n",
       "      <td>329</td>\n",
       "      <td>CTGGCCACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>37</td>\n",
       "      <td>321</td>\n",
       "      <td>330</td>\n",
       "      <td>TGGCCACAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>38</td>\n",
       "      <td>322</td>\n",
       "      <td>331</td>\n",
       "      <td>GGCCACAGCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id          side  number_of_win  window_start  window_end    sequence\n",
       "0       1  within_start              1           285         294  CAAAAAATTT\n",
       "1       1  within_start              2           286         295  AAAAAATTTC\n",
       "2       1  within_start              3           287         296  AAAAATTTCC\n",
       "3       1  within_start              4           288         297  AAAATTTCCA\n",
       "4       1  within_start              5           289         298  AAATTTCCAC\n",
       "5       1  within_start              6           290         299  AATTTCCACC\n",
       "6       1  within_start              7           291         300  ATTTCCACCA\n",
       "7       1  within_start              8           292         301  TTTCCACCAA\n",
       "8       1  within_start              9           293         302  TTCCACCAAA\n",
       "9       1  within_start             10           294         303  TCCACCAAAC\n",
       "10      1  within_start             11           295         304  CCACCAAACC\n",
       "11      1  within_start             12           296         305  CACCAAACCC\n",
       "12      1  within_start             13           297         306  ACCAAACCCC\n",
       "13      1  within_start             14           298         307  CCAAACCCCC\n",
       "14      1  within_start             15           299         308  CAAACCCCCC\n",
       "15      1  within_start             16           300         309  AAACCCCCCC\n",
       "16      1  within_start             17           301         310  AACCCCCCCT\n",
       "17      1  within_start             18           302         311  ACCCCCCCTC\n",
       "18      1  within_start             19           303         312  CCCCCCCTCC\n",
       "19      1  within_start             20           304         313  CCCCCCTCCC\n",
       "20      1  within_start             21           305         314  CCCCCTCCCC\n",
       "21      1  within_start             22           306         315  CCCCTCCCCC\n",
       "22      1  within_start             23           307         316  CCCTCCCCCG\n",
       "23      1  within_start             24           308         317  CCTCCCCCGC\n",
       "24      1  within_start             25           309         318  CTCCCCCGCT\n",
       "25      1  within_start             26           310         319  TCCCCCGCTT\n",
       "26      1  within_start             27           311         320  CCCCCGCTTC\n",
       "27      1  within_start             28           312         321  CCCCGCTTCT\n",
       "28      1  within_start             29           313         322  CCCGCTTCTG\n",
       "29      1  within_start             30           314         323  CCGCTTCTGG\n",
       "30      1  within_start             31           315         324  CGCTTCTGGC\n",
       "31      1  within_start             32           316         325  GCTTCTGGCC\n",
       "32      1  within_start             33           317         326  CTTCTGGCCA\n",
       "33      1  within_start             34           318         327  TTCTGGCCAC\n",
       "34      1  within_start             35           319         328  TCTGGCCACA\n",
       "35      1  within_start             36           320         329  CTGGCCACAG\n",
       "36      1  within_start             37           321         330  TGGCCACAGC\n",
       "37      1  within_start             38           322         331  GGCCACAGCA"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_within_sides[(df_win_within_sides['side'] == 'within_start') & (df_win_within_sides['G4_id'] == 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "972b1ad8-728f-4875-8361-cbcf9393c20b",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_win_within_sides = df_win_within_sides['sequence'].apply(lambda x: Counter(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "23ed8ec0-5e15-4d57-a34d-0c5963ade617",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_df_win_within_sides = pd.DataFrame.from_records(nuc_count_win_within_sides).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d0306a1d-1c08-4974-870a-2fc99ece9284",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within_sides = df_win_within_sides.join(nuc_count_df_win_within_sides).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f3d6801d-f02a-4f40-a5bc-e69f51154b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_within_sides.to_csv('../data/processed/df_win_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8201f772-c292-4b89-af76-2695b79b3dee",
   "metadata": {},
   "source": [
    "# Within sides Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1e19fc2d-fb5a-4997-b490-fe4e0e305d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win_within_sides = pd.read_csv('../data/processed/df_win_within_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a35fa438-536c-4f82-92e9-96e2dc9dbf8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_within_sides = []\n",
    "for window_start, window_end in zip(df_win_within_sides['window_start'], df_win_within_sides['window_end']):\n",
    "    row = [*range(window_start, window_end)]\n",
    "    Mut_win_within_sides.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "46ee04bd-393f-49a6-af4e-45ee435b202b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_within_sides = pd.DataFrame(Mut_win_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5fbd5e66-4a48-4615-8b02-9a4529725eda",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_within_sides[0] = Mut_win_df_within_sides[0].str.findall(\"[A-T]>[A-T]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "78176695-5ecd-426e-87d3-9a782e822a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutwin_within_sides = pd.DataFrame.from_records(Mut_win_df_within_sides[0].apply(lambda x: Counter(x))) \\\n",
    "    .fillna(0) \\\n",
    "    .sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "1a041c92-2ae8-4295-92e5-6c99a0f51031",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within_sides = df_win_within_sides.join(obsMutwin_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b375cdfb-5dba-413f-9c12-5d643f7c4831",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within_sides['number_of_Mut'] = df_win_cancer_within_sides[['A>C','A>G','A>T',\n",
    "                                                              'C>A','C>G','C>T',\n",
    "                                                              'G>A','G>C','G>T',\n",
    "                                                              'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "7f56f107-4ed0-410f-8c4c-d9ad4db431a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id          side  number_of_win  window_start  window_end    sequence  \\\n",
       "0      1  within_start              1           285         294  CAAAAAATTT   \n",
       "1      1  within_start              2           286         295  AAAAAATTTC   \n",
       "2      1  within_start              3           287         296  AAAAATTTCC   \n",
       "3      1  within_start              4           288         297  AAAATTTCCA   \n",
       "4      1  within_start              5           289         298  AAATTTCCAC   \n",
       "\n",
       "     A    C    G    T  ...  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G  \\\n",
       "0  6.0  1.0  0.0  3.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  2.0  0.0   \n",
       "1  6.0  1.0  0.0  3.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "2  5.0  2.0  0.0  3.0  ...  0.0  0.0  1.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "3  5.0  2.0  0.0  3.0  ...  0.0  0.0  4.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "4  4.0  3.0  0.0  3.0  ...  0.0  0.0  4.0  0.0  0.0  0.0  0.0  4.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0            2.0  \n",
       "1            4.0  \n",
       "2            5.0  \n",
       "3            8.0  \n",
       "4           10.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_cancer_within_sides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "9eadd5ed-cea7-469d-851e-2ad4e081f2ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_within_sides.to_csv('../data/processed/df_win_cancer_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e455a9ac-bbfc-4374-b3e3-2352c7bcb33d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Within sides HGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "24c1a84c-1131-481c-b0e3-20f560764b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_13260\\1406226394.py:1: DtypeWarning: Columns (10,11,12,14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_hgt_g4 = pd.read_csv('../data/processed/df_hgt_g4.csv')\n"
     ]
    }
   ],
   "source": [
    "df_hgt_g4 = pd.read_csv('../data/processed/df_hgt_g4.csv')\n",
    "df_win_within_sides = pd.read_csv('../data/processed/df_win_within_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "bb4f24bf-54ee-4c91-8969-0fc301989c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract mutations without gaps (-) and other (R, M, N, ect) bases in mut position:\n",
    "df_hgt_g4 = df_hgt_g4[\n",
    "    (df_hgt_g4.ancestor.str.lower().str.fullmatch(\"[acgt]{5}\")) &\n",
    "    (df_hgt_g4.descendant.str.lower().str.fullmatch(\"[acgt]{5}\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f13f219f-8e59-4811-9ba8-c65fa84ee6fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_within_sides = []\n",
    "for G4, G4_id, Pos, Mut in zip(df_hgt_g4['G4'], df_hgt_g4['G4_id'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if G4 > 0:\n",
    "        row.append(G4_id)\n",
    "        row.append('within')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_within_sides.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "96ac0880-d4ef-4c5c-8972-deca9ab19fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_obs_mut_hgt_within_sides = pd.DataFrame(data=obs_mut_hgt_within_sides, columns=['G4_id', 'side', 'Pos', 'Mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a238997a-ddf4-4d5a-802c-5a9b2014a3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_hgt_win_within_sides = []\n",
    "for window_start, window_end in zip(df_win_within_sides['window_start'], df_win_within_sides['window_end']):\n",
    "    row = []\n",
    "    \n",
    "    mut = df_obs_mut_hgt_within_sides[(df_obs_mut_hgt_within_sides['Pos'] >= window_start) & \n",
    "        (df_obs_mut_hgt_within_sides['Pos'] <= window_end)]['Mut'].tolist()\n",
    "\n",
    "    row.append(mut)\n",
    "    \n",
    "    Mut_hgt_win_within_sides.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "e7d2b5b1-2536-4a36-a9cd-0b03fe9c624b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Mut_hgt_win_within_sides = pd.DataFrame(data=Mut_hgt_win_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ae247255-4877-4aaa-b560-30330023a8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutG4_hgt_within_sides = pd.DataFrame.from_records(df_Mut_hgt_win_within_sides[0].apply(lambda x: Counter(x))) \\\n",
    "    .fillna(0) \\\n",
    "    .sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "82bf9307-75fc-4f0c-a8a4-e36fc38917c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within_sides = df_win_within_sides.join(obsMutG4_hgt_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9a17414b-3b4f-491c-bb59-1eb5be0e6a4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within_sides['number_of_Mut'] = df_win_hgt_within_sides[['A>C','A>G','A>T',\n",
    "                                                        'C>A','C>G','C>T',\n",
    "                                                        'G>A','G>C','G>T',\n",
    "                                                        'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "0cf8821a-4985-4111-a181-7c0281d9ac08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>539.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id          side  number_of_win  window_start  window_end    sequence  \\\n",
       "0      1  within_start              1           285         294  CAAAAAATTT   \n",
       "1      1  within_start              2           286         295  AAAAAATTTC   \n",
       "2      1  within_start              3           287         296  AAAAATTTCC   \n",
       "3      1  within_start              4           288         297  AAAATTTCCA   \n",
       "4      1  within_start              5           289         298  AAATTTCCAC   \n",
       "\n",
       "     A    C    G    T  ...   C>A  C>G    C>T   G>A  G>C  G>T  T>A    T>C  T>G  \\\n",
       "0  6.0  1.0  0.0  3.0  ...   0.0  0.0   76.0   5.0  0.0  0.0  8.0   88.0  0.0   \n",
       "1  6.0  1.0  0.0  3.0  ...  26.0  0.0  128.0   5.0  0.0  0.0  9.0  140.0  0.0   \n",
       "2  5.0  2.0  0.0  3.0  ...  28.0  0.0  152.0   5.0  0.0  0.0  9.0  163.0  0.0   \n",
       "3  5.0  2.0  0.0  3.0  ...  34.0  0.0  152.0  72.0  0.0  0.0  9.0  163.0  0.0   \n",
       "4  4.0  3.0  0.0  3.0  ...  35.0  0.0  152.0  67.0  0.0  0.0  9.0  163.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0          189.0  \n",
       "1          346.0  \n",
       "2          397.0  \n",
       "3          548.0  \n",
       "4          539.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_hgt_within_sides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6908ca3d-6cb5-4e63-9762-6028447d53af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_hgt_within_sides.to_csv('../data/processed/df_win_hgt_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f5ecbb-abea-496d-8954-4cd968d2a740",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Location of g4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9b25f5fd-29f8-465b-8078-ba0572244a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4_location = []\n",
    "for G4_id, start, end in zip(g4conf['G4_id'], g4conf['start'], g4conf['end']):\n",
    "    location = []\n",
    "    row = []\n",
    "    for i in range(start, end+1):\n",
    "        # row.append(G4_id)\n",
    "        # row.append(i)\n",
    "        location.append(ref_annot[ref_annot['Pos'] == i]['Type'].to_string(index=False))\n",
    "        \n",
    "    row.append(G4_id)\n",
    "    row.append('/'.join(list(set(location))))\n",
    "    g4_location.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5ac157a1-9c4c-4630-9830-76b013eb3f7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1, 'D-loop'],\n",
       " [2, 'D-loop'],\n",
       " [3, 'D-loop/tRNA'],\n",
       " [4, 'rRNA'],\n",
       " [5, 'rRNA'],\n",
       " [6, 'tRNA/rRNA'],\n",
       " [7, 'rRNA'],\n",
       " [8, 'rRNA'],\n",
       " [9, 'rRNA'],\n",
       " [10, 'rRNA'],\n",
       " [11, 'CDS'],\n",
       " [12, 'CDS'],\n",
       " [13, 'CDS'],\n",
       " [14, 'CDS'],\n",
       " [15, 'CDS'],\n",
       " [16, 'CDS'],\n",
       " [17, 'tRNA'],\n",
       " [18, 'tRNA/CDS'],\n",
       " [19, 'CDS'],\n",
       " [20, 'CDS'],\n",
       " [21, 'CDS'],\n",
       " [22, 'CDS'],\n",
       " [23, 'CDS'],\n",
       " [24, 'tRNA/NaN/CDS'],\n",
       " [25, 'CDS'],\n",
       " [26, 'CDS'],\n",
       " [27, 'CDS'],\n",
       " [28, 'CDS'],\n",
       " [29, 'CDS'],\n",
       " [30, 'CDS'],\n",
       " [31, 'CDS'],\n",
       " [32, 'NaN/CDS/tRNA'],\n",
       " [33, 'CDS'],\n",
       " [34, 'CDS'],\n",
       " [35, 'CDS'],\n",
       " [36, 'CDS'],\n",
       " [37, 'CDS'],\n",
       " [38, 'CDS'],\n",
       " [39, 'CDS'],\n",
       " [40, 'CDS'],\n",
       " [41, 'CDS'],\n",
       " [42, 'CDS'],\n",
       " [43, 'CDS'],\n",
       " [44, 'CDS'],\n",
       " [45, 'CDS'],\n",
       " [46, 'CDS'],\n",
       " [47, 'CDS'],\n",
       " [48, 'CDS'],\n",
       " [49, 'CDS'],\n",
       " [50, 'CDS'],\n",
       " [51, 'CDS'],\n",
       " [52, 'CDS'],\n",
       " [53, 'CDS'],\n",
       " [54, 'CDS'],\n",
       " [55, 'CDS'],\n",
       " [56, 'CDS'],\n",
       " [57, 'CDS'],\n",
       " [58, 'CDS'],\n",
       " [59, 'CDS'],\n",
       " [60, 'CDS'],\n",
       " [61, 'CDS'],\n",
       " [62, 'CDS'],\n",
       " [63, 'CDS'],\n",
       " [64, 'CDS'],\n",
       " [65, 'CDS'],\n",
       " [66, 'CDS'],\n",
       " [67, 'CDS'],\n",
       " [68, 'CDS'],\n",
       " [69, 'CDS'],\n",
       " [70, 'CDS'],\n",
       " [71, 'D-loop'],\n",
       " [72, 'D-loop'],\n",
       " [73, 'D-loop'],\n",
       " [74, 'D-loop']]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6996452e-ce34-4554-8198-91d265e23823",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_location = pd.DataFrame(data=g4_location, columns=['G4_id', 'location'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d8b008b9-8161-490c-8aa0-18949456d5fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>D-loop/tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>70</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>72</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>73</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>74</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id     location\n",
       "0       1       D-loop\n",
       "1       2       D-loop\n",
       "2       3  D-loop/tRNA\n",
       "3       4         rRNA\n",
       "4       5         rRNA\n",
       "..    ...          ...\n",
       "69     70          CDS\n",
       "70     71       D-loop\n",
       "71     72       D-loop\n",
       "72     73       D-loop\n",
       "73     74       D-loop\n",
       "\n",
       "[74 rows x 2 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3f80e031-6250-4ac5-bbbb-e99a7ad713a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_location = df_g4_location.set_index('G4_id')['location'].str.split('/', expand=True) \\\n",
    "    .stack() \\\n",
    "    .reset_index(name='location') \\\n",
    "    .drop('level_1', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "619c81a1-64cd-4aeb-8736-77b00a279e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>70</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>71</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>72</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>73</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>74</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id location\n",
       "0       1   D-loop\n",
       "1       2   D-loop\n",
       "2       3   D-loop\n",
       "3       3     tRNA\n",
       "4       4     rRNA\n",
       "..    ...      ...\n",
       "76     70      CDS\n",
       "77     71   D-loop\n",
       "78     72   D-loop\n",
       "79     73   D-loop\n",
       "80     74   D-loop\n",
       "\n",
       "[81 rows x 2 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "238b70b6-bbb5-4d35-b151-fb6dd91d1c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_location.to_csv('../data/processed/df_g4_location.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e217401f-778f-471c-8944-d60363780fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqnames', 'start', 'end', 'width', 'strand', 'score', 'max_score',\n",
       "       'threshold', 'window', 'sequence', 'confirmed', 'G4', 'note',\n",
       "       'start-100bp', 'end+100bp', 'start-50bp', 'end+50bp',\n",
       "       'bp_untill_next_G4', 'bp_untill_previous_G4', 'G4_id', 'up_win',\n",
       "       'down_win'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "207182b6-3926-440c-beaa-e5d8debac76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_type = g4conf[['G4_id', 'G4']].rename(columns={'G4':'Type_of_G4'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1a3802d-9cb0-45ef-ae1b-667476e4815e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_type.to_csv('../data/processed/df_g4_type.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea9edee-8842-4dd8-afd3-ad23e066bb24",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# G4 location *in vivo* (doi.org/10.1093/nar/gkad535)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "10eb5e60-ca71-4f9c-99cd-3165eb22bc2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "G4_peaks = [[1, 2425, 3986], [2, 4330, 4624], [3, 5570, 7168], [4, 7139, 7559], [5, 7800, 8276], [6, 8958, 9869], [7, 12762, 13771]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fa699b23-3911-42c9-a74a-ae2ac50bf4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_vivo_peaks = pd.DataFrame(G4_peaks, columns=['number', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8c79b7f6-7683-4f16-8f5c-68c50cc9a383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2425</td>\n",
       "      <td>3986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4330</td>\n",
       "      <td>4624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>5570</td>\n",
       "      <td>7168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>7139</td>\n",
       "      <td>7559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>7800</td>\n",
       "      <td>8276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>8958</td>\n",
       "      <td>9869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>12762</td>\n",
       "      <td>13771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   number  start    end\n",
       "0       1   2425   3986\n",
       "1       2   4330   4624\n",
       "2       3   5570   7168\n",
       "3       4   7139   7559\n",
       "4       5   7800   8276\n",
       "5       6   8958   9869\n",
       "6       7  12762  13771"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_vivo_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9864ebf4-92b7-4dff-b323-d16e64db9c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6ef9e448-5e23-47e0-a09f-7d651fe49230",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4_vivo = []\n",
    "for start, end, G4_id in zip(g4conf['start'], g4conf['end'], g4conf['G4_id']):\n",
    "    for p_start, p_end, number in zip(df_g4_vivo_peaks['start'], df_g4_vivo_peaks['end'], df_g4_vivo_peaks['number']):\n",
    "        row = []\n",
    "        if start in range(p_start, p_end) or end in range(p_start, p_end):\n",
    "            row.append(number)\n",
    "            row.append(G4_id)\n",
    "            row.append(1)\n",
    "            g4_vivo.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c339a343-cc39-421a-9ff3-af75aea04821",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_vivo = pd.DataFrame(g4_vivo, columns=['peak_number', 'G4_id', 'in_vivo'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "21443f3b-5492-4c6f-a9a9-883b243ffed8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_number</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>in_vivo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>6</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>6</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>6</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>7</td>\n",
       "      <td>52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>7</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>7</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>7</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>7</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    peak_number  G4_id  in_vivo\n",
       "0             1      9        1\n",
       "1             1     10        1\n",
       "2             1     11        1\n",
       "3             1     12        1\n",
       "4             1     13        1\n",
       "5             2     17        1\n",
       "6             2     18        1\n",
       "7             3     24        1\n",
       "8             3     25        1\n",
       "9             3     26        1\n",
       "10            3     27        1\n",
       "11            3     28        1\n",
       "12            4     29        1\n",
       "13            5     30        1\n",
       "14            5     31        1\n",
       "15            5     32        1\n",
       "16            6     37        1\n",
       "17            6     38        1\n",
       "18            6     39        1\n",
       "19            6     40        1\n",
       "20            7     52        1\n",
       "21            7     53        1\n",
       "22            7     54        1\n",
       "23            7     55        1\n",
       "24            7     56        1\n",
       "25            7     57        1"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_vivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "95362125-0985-4a77-9152-8dbee0d20b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_vivo.to_csv('../data/processed/df_g4_vivo.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "193291d6-4be2-4d02-ae83-78c1cc1bb043",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Check decrease of >T  \n",
    "Make new windows which cover all length G4 and +-25 bp \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f454cf2d-1bc9-4f35-8c67-3f29be88c1a2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "      <th>start-50bp</th>\n",
       "      <th>end+50bp</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.250</td>\n",
       "      <td>-2.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>185</td>\n",
       "      <td>432</td>\n",
       "      <td>235</td>\n",
       "      <td>382</td>\n",
       "      <td>16</td>\n",
       "      <td>294</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>...</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>248</td>\n",
       "      <td>478</td>\n",
       "      <td>298</td>\n",
       "      <td>428</td>\n",
       "      <td>40</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>332</td>\n",
       "      <td>418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "      <td>368</td>\n",
       "      <td>642</td>\n",
       "      <td>193</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>378</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>785</td>\n",
       "      <td>808</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCTTAGCCTAGCCACACCCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "      <td>735</td>\n",
       "      <td>858</td>\n",
       "      <td>146</td>\n",
       "      <td>193</td>\n",
       "      <td>4</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "      <td>904</td>\n",
       "      <td>1034</td>\n",
       "      <td>683</td>\n",
       "      <td>146</td>\n",
       "      <td>5</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>15536</td>\n",
       "      <td>15783</td>\n",
       "      <td>15586</td>\n",
       "      <td>15733</td>\n",
       "      <td>484</td>\n",
       "      <td>80</td>\n",
       "      <td>70</td>\n",
       "      <td>15556</td>\n",
       "      <td>15783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16067</td>\n",
       "      <td>16311</td>\n",
       "      <td>16117</td>\n",
       "      <td>16261</td>\n",
       "      <td>28</td>\n",
       "      <td>484</td>\n",
       "      <td>71</td>\n",
       "      <td>16067</td>\n",
       "      <td>16239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>63</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16139</td>\n",
       "      <td>16401</td>\n",
       "      <td>16189</td>\n",
       "      <td>16351</td>\n",
       "      <td>43</td>\n",
       "      <td>28</td>\n",
       "      <td>72</td>\n",
       "      <td>16211</td>\n",
       "      <td>16344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16244</td>\n",
       "      <td>16482</td>\n",
       "      <td>16294</td>\n",
       "      <td>16432</td>\n",
       "      <td>145</td>\n",
       "      <td>43</td>\n",
       "      <td>73</td>\n",
       "      <td>16301</td>\n",
       "      <td>16482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16527</td>\n",
       "      <td>16560</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.853</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16427</td>\n",
       "      <td>91</td>\n",
       "      <td>16477</td>\n",
       "      <td>16610</td>\n",
       "      <td>294</td>\n",
       "      <td>145</td>\n",
       "      <td>74</td>\n",
       "      <td>16427</td>\n",
       "      <td>16660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start    end  width  \\\n",
       "0   NC_012920.1 Homo sapiens mitochondrion, comple...    285    332     48   \n",
       "1   NC_012920.1 Homo sapiens mitochondrion, comple...    348    378     31   \n",
       "2   NC_012920.1 Homo sapiens mitochondrion, comple...    418    592    175   \n",
       "3   NC_012920.1 Homo sapiens mitochondrion, comple...    785    808     24   \n",
       "4   NC_012920.1 Homo sapiens mitochondrion, comple...    954    984     31   \n",
       "..                                                ...    ...    ...    ...   \n",
       "69  NC_012920.1 Homo sapiens mitochondrion, comple...  15636  15683     48   \n",
       "70  NC_012920.1 Homo sapiens mitochondrion, comple...  16167  16211     45   \n",
       "71  NC_012920.1 Homo sapiens mitochondrion, comple...  16239  16301     63   \n",
       "72  NC_012920.1 Homo sapiens mitochondrion, comple...  16344  16382     39   \n",
       "73  NC_012920.1 Homo sapiens mitochondrion, comple...  16527  16560     34   \n",
       "\n",
       "   strand  score  max_score  threshold  window  \\\n",
       "0       - -1.250      -2.24          1      25   \n",
       "1       - -1.100      -1.16          1      25   \n",
       "2       - -1.150      -1.88          1      25   \n",
       "3       - -1.210      -1.16          1      25   \n",
       "4       - -1.390      -1.48          1      25   \n",
       "..    ...    ...        ...        ...     ...   \n",
       "69      - -0.896      -1.16          1      25   \n",
       "70      - -1.160      -1.80          1      25   \n",
       "71      - -0.968      -1.40          1      25   \n",
       "72      - -1.210      -1.48          1      25   \n",
       "73      - -0.853      -1.16          1      25   \n",
       "\n",
       "                                             sequence  ...  \\\n",
       "0    CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC  ...   \n",
       "1                     CCAAACCCCAAAAACAAAGAACCCTAACACC  ...   \n",
       "2   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...  ...   \n",
       "3                            CGCTTAGCCTAGCCACACCCCCAC  ...   \n",
       "4                     CACCCCCTCCCCAATAAAGCTAAAACTCACC  ...   \n",
       "..                                                ...  ...   \n",
       "69   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC  ...   \n",
       "70      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC  ...   \n",
       "71  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...  ...   \n",
       "72            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC  ...   \n",
       "73                 CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC  ...   \n",
       "\n",
       "                          note start-100bp end+100bp  start-50bp  end+50bp  \\\n",
       "0             more than africa         185       432         235       382   \n",
       "1             less than africa         248       478         298       428   \n",
       "2   5 overlapping g4 in africa         318       692         368       642   \n",
       "3                          NaN         685       908         735       858   \n",
       "4                          NaN         854      1084         904      1034   \n",
       "..                         ...         ...       ...         ...       ...   \n",
       "69            more than africa       15536     15783       15586     15733   \n",
       "70            more than africa       16067     16311       16117     16261   \n",
       "71            more than africa       16139     16401       16189     16351   \n",
       "72            more than africa       16244     16482       16294     16432   \n",
       "73            more than africa       16427        91       16477     16610   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  G4_id  up_win  down_win  \n",
       "0                  16                    294      1     185       348  \n",
       "1                  40                     16      2     332       418  \n",
       "2                 193                     40      3     378       692  \n",
       "3                 146                    193      4     685       908  \n",
       "4                 683                    146      5     854      1084  \n",
       "..                ...                    ...    ...     ...       ...  \n",
       "69                484                     80     70   15556     15783  \n",
       "70                 28                    484     71   16067     16239  \n",
       "71                 43                     28     72   16211     16344  \n",
       "72                145                     43     73   16301     16482  \n",
       "73                294                    145     74   16427     16660  \n",
       "\n",
       "[74 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "682becd2-66c4-48b5-b8a7-2585d1818f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_25bp = g4conf[(g4conf['bp_untill_previous_G4'] >= 25) & \n",
    "            (g4conf['bp_untill_next_G4'] >= 25) & \n",
    "            (g4conf['end'] + 25 <= 16569)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf69aff9-8a67-4b42-8b60-9e7439654dfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "      <th>start-50bp</th>\n",
       "      <th>end+50bp</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "      <td>368</td>\n",
       "      <td>642</td>\n",
       "      <td>193</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>378</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>785</td>\n",
       "      <td>808</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCTTAGCCTAGCCACACCCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "      <td>735</td>\n",
       "      <td>858</td>\n",
       "      <td>146</td>\n",
       "      <td>193</td>\n",
       "      <td>4</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "      <td>904</td>\n",
       "      <td>1034</td>\n",
       "      <td>683</td>\n",
       "      <td>146</td>\n",
       "      <td>5</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>1667</td>\n",
       "      <td>1707</td>\n",
       "      <td>41</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTGAGCTAAACCTAGCCCCAAACCCACTCCACCTTACTACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than in africa</td>\n",
       "      <td>1567</td>\n",
       "      <td>1807</td>\n",
       "      <td>1617</td>\n",
       "      <td>1757</td>\n",
       "      <td>350</td>\n",
       "      <td>683</td>\n",
       "      <td>6</td>\n",
       "      <td>1567</td>\n",
       "      <td>1807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>2057</td>\n",
       "      <td>2079</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.520</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCACAGAACCCTCTAAATCCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1957</td>\n",
       "      <td>2179</td>\n",
       "      <td>2007</td>\n",
       "      <td>2129</td>\n",
       "      <td>133</td>\n",
       "      <td>350</td>\n",
       "      <td>7</td>\n",
       "      <td>1957</td>\n",
       "      <td>2179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>2212</td>\n",
       "      <td>2236</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACACCCACTACCTAAAAAATCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2112</td>\n",
       "      <td>2336</td>\n",
       "      <td>2162</td>\n",
       "      <td>2286</td>\n",
       "      <td>912</td>\n",
       "      <td>133</td>\n",
       "      <td>8</td>\n",
       "      <td>2112</td>\n",
       "      <td>2336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3471</td>\n",
       "      <td>3604</td>\n",
       "      <td>134</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.170</td>\n",
       "      <td>-2.76</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>3371</td>\n",
       "      <td>3704</td>\n",
       "      <td>3421</td>\n",
       "      <td>3654</td>\n",
       "      <td>270</td>\n",
       "      <td>249</td>\n",
       "      <td>11</td>\n",
       "      <td>3371</td>\n",
       "      <td>3704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3874</td>\n",
       "      <td>3909</td>\n",
       "      <td>36</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.917</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3774</td>\n",
       "      <td>4009</td>\n",
       "      <td>3824</td>\n",
       "      <td>3959</td>\n",
       "      <td>42</td>\n",
       "      <td>270</td>\n",
       "      <td>12</td>\n",
       "      <td>3774</td>\n",
       "      <td>3951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3951</td>\n",
       "      <td>3984</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.882</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>3851</td>\n",
       "      <td>4084</td>\n",
       "      <td>3901</td>\n",
       "      <td>4034</td>\n",
       "      <td>102</td>\n",
       "      <td>42</td>\n",
       "      <td>13</td>\n",
       "      <td>3909</td>\n",
       "      <td>4084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4228</td>\n",
       "      <td>4260</td>\n",
       "      <td>33</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCATTACAATCTCCAGCATTCCCCCTCAAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4128</td>\n",
       "      <td>4360</td>\n",
       "      <td>4178</td>\n",
       "      <td>4310</td>\n",
       "      <td>112</td>\n",
       "      <td>67</td>\n",
       "      <td>16</td>\n",
       "      <td>4161</td>\n",
       "      <td>4360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4372</td>\n",
       "      <td>4398</td>\n",
       "      <td>27</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.220</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCCGTGCCACCTATCACACCCCATCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4272</td>\n",
       "      <td>4498</td>\n",
       "      <td>4322</td>\n",
       "      <td>4448</td>\n",
       "      <td>58</td>\n",
       "      <td>112</td>\n",
       "      <td>17</td>\n",
       "      <td>4272</td>\n",
       "      <td>4456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4456</td>\n",
       "      <td>4505</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTTCCCGTACTAATTAATCCCCTGGCCCAACCCGTCATCTACTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in africa</td>\n",
       "      <td>4356</td>\n",
       "      <td>4605</td>\n",
       "      <td>4406</td>\n",
       "      <td>4555</td>\n",
       "      <td>370</td>\n",
       "      <td>58</td>\n",
       "      <td>18</td>\n",
       "      <td>4398</td>\n",
       "      <td>4605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4875</td>\n",
       "      <td>4899</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.080</td>\n",
       "      <td>-1.08</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGCCCCCATCTCAATCATATACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4775</td>\n",
       "      <td>4999</td>\n",
       "      <td>4825</td>\n",
       "      <td>4949</td>\n",
       "      <td>288</td>\n",
       "      <td>370</td>\n",
       "      <td>19</td>\n",
       "      <td>4775</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5187</td>\n",
       "      <td>5255</td>\n",
       "      <td>69</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.957</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGC...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>5087</td>\n",
       "      <td>5355</td>\n",
       "      <td>5137</td>\n",
       "      <td>5305</td>\n",
       "      <td>33</td>\n",
       "      <td>288</td>\n",
       "      <td>20</td>\n",
       "      <td>5087</td>\n",
       "      <td>5288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5424</td>\n",
       "      <td>5499</td>\n",
       "      <td>76</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.200</td>\n",
       "      <td>-1.92</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>5324</td>\n",
       "      <td>5599</td>\n",
       "      <td>5374</td>\n",
       "      <td>5549</td>\n",
       "      <td>376</td>\n",
       "      <td>32</td>\n",
       "      <td>23</td>\n",
       "      <td>5392</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5875</td>\n",
       "      <td>5916</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.881</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTCAGCCATTTTACCTCACCCCCACTGATGTTCGCCGACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>5775</td>\n",
       "      <td>6016</td>\n",
       "      <td>5825</td>\n",
       "      <td>5966</td>\n",
       "      <td>297</td>\n",
       "      <td>376</td>\n",
       "      <td>24</td>\n",
       "      <td>5775</td>\n",
       "      <td>6016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6213</td>\n",
       "      <td>6242</td>\n",
       "      <td>30</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.900</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCTTACCTCCCTCTCTCCTACTCCTGCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>6113</td>\n",
       "      <td>6342</td>\n",
       "      <td>6163</td>\n",
       "      <td>6292</td>\n",
       "      <td>168</td>\n",
       "      <td>297</td>\n",
       "      <td>25</td>\n",
       "      <td>6113</td>\n",
       "      <td>6342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6410</td>\n",
       "      <td>6464</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.050</td>\n",
       "      <td>-1.44</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>6310</td>\n",
       "      <td>6564</td>\n",
       "      <td>6360</td>\n",
       "      <td>6514</td>\n",
       "      <td>81</td>\n",
       "      <td>168</td>\n",
       "      <td>26</td>\n",
       "      <td>6310</td>\n",
       "      <td>6545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6545</td>\n",
       "      <td>6572</td>\n",
       "      <td>28</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTCAACACCACCTTCTTCGACCCCGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>6445</td>\n",
       "      <td>6672</td>\n",
       "      <td>6495</td>\n",
       "      <td>6622</td>\n",
       "      <td>254</td>\n",
       "      <td>81</td>\n",
       "      <td>27</td>\n",
       "      <td>6464</td>\n",
       "      <td>6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6826</td>\n",
       "      <td>6851</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.040</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTACCATAATCATCGCTATCCCCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6726</td>\n",
       "      <td>6951</td>\n",
       "      <td>6776</td>\n",
       "      <td>6901</td>\n",
       "      <td>535</td>\n",
       "      <td>254</td>\n",
       "      <td>28</td>\n",
       "      <td>6726</td>\n",
       "      <td>6951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7386</td>\n",
       "      <td>7427</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>7286</td>\n",
       "      <td>7527</td>\n",
       "      <td>7336</td>\n",
       "      <td>7477</td>\n",
       "      <td>371</td>\n",
       "      <td>535</td>\n",
       "      <td>29</td>\n",
       "      <td>7286</td>\n",
       "      <td>7527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7798</td>\n",
       "      <td>7840</td>\n",
       "      <td>43</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.977</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7698</td>\n",
       "      <td>7940</td>\n",
       "      <td>7748</td>\n",
       "      <td>7890</td>\n",
       "      <td>96</td>\n",
       "      <td>371</td>\n",
       "      <td>30</td>\n",
       "      <td>7698</td>\n",
       "      <td>7936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7936</td>\n",
       "      <td>7974</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.923</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7836</td>\n",
       "      <td>8074</td>\n",
       "      <td>7886</td>\n",
       "      <td>8024</td>\n",
       "      <td>278</td>\n",
       "      <td>96</td>\n",
       "      <td>31</td>\n",
       "      <td>7840</td>\n",
       "      <td>8074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8252</td>\n",
       "      <td>8305</td>\n",
       "      <td>54</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.220</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCGTATTTACCCTATAGCACCCCCTCTACCCCCTCTAGAGCCCAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>8152</td>\n",
       "      <td>8405</td>\n",
       "      <td>8202</td>\n",
       "      <td>8355</td>\n",
       "      <td>150</td>\n",
       "      <td>278</td>\n",
       "      <td>32</td>\n",
       "      <td>8152</td>\n",
       "      <td>8405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8455</td>\n",
       "      <td>8488</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.090</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACAAACTACCACCTACCTCCCTCACCAAAGCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8355</td>\n",
       "      <td>8588</td>\n",
       "      <td>8405</td>\n",
       "      <td>8538</td>\n",
       "      <td>56</td>\n",
       "      <td>150</td>\n",
       "      <td>33</td>\n",
       "      <td>8355</td>\n",
       "      <td>8544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8791</td>\n",
       "      <td>8815</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTCACTCATTTACACCAACCACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8691</td>\n",
       "      <td>8915</td>\n",
       "      <td>8741</td>\n",
       "      <td>8865</td>\n",
       "      <td>102</td>\n",
       "      <td>148</td>\n",
       "      <td>36</td>\n",
       "      <td>8691</td>\n",
       "      <td>8915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8917</td>\n",
       "      <td>8958</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.881</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>8817</td>\n",
       "      <td>9058</td>\n",
       "      <td>8867</td>\n",
       "      <td>9008</td>\n",
       "      <td>285</td>\n",
       "      <td>102</td>\n",
       "      <td>37</td>\n",
       "      <td>8817</td>\n",
       "      <td>9058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>9505</td>\n",
       "      <td>9607</td>\n",
       "      <td>103</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>9405</td>\n",
       "      <td>9707</td>\n",
       "      <td>9455</td>\n",
       "      <td>9657</td>\n",
       "      <td>568</td>\n",
       "      <td>212</td>\n",
       "      <td>40</td>\n",
       "      <td>9405</td>\n",
       "      <td>9707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10175</td>\n",
       "      <td>10214</td>\n",
       "      <td>40</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.200</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>10075</td>\n",
       "      <td>10314</td>\n",
       "      <td>10125</td>\n",
       "      <td>10264</td>\n",
       "      <td>43</td>\n",
       "      <td>568</td>\n",
       "      <td>41</td>\n",
       "      <td>10075</td>\n",
       "      <td>10257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10257</td>\n",
       "      <td>10303</td>\n",
       "      <td>47</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.894</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>10157</td>\n",
       "      <td>10403</td>\n",
       "      <td>10207</td>\n",
       "      <td>10353</td>\n",
       "      <td>300</td>\n",
       "      <td>43</td>\n",
       "      <td>42</td>\n",
       "      <td>10214</td>\n",
       "      <td>10403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10603</td>\n",
       "      <td>10637</td>\n",
       "      <td>35</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCTCATAACCCTCAACACCCACTCCCTCTTAGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10503</td>\n",
       "      <td>10737</td>\n",
       "      <td>10553</td>\n",
       "      <td>10687</td>\n",
       "      <td>275</td>\n",
       "      <td>300</td>\n",
       "      <td>43</td>\n",
       "      <td>10503</td>\n",
       "      <td>10737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10912</td>\n",
       "      <td>10990</td>\n",
       "      <td>79</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.330</td>\n",
       "      <td>-2.08</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>10812</td>\n",
       "      <td>11090</td>\n",
       "      <td>10862</td>\n",
       "      <td>11040</td>\n",
       "      <td>132</td>\n",
       "      <td>275</td>\n",
       "      <td>44</td>\n",
       "      <td>10812</td>\n",
       "      <td>11090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11122</td>\n",
       "      <td>11162</td>\n",
       "      <td>41</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.854</td>\n",
       "      <td>-1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11022</td>\n",
       "      <td>11262</td>\n",
       "      <td>11072</td>\n",
       "      <td>11212</td>\n",
       "      <td>50</td>\n",
       "      <td>132</td>\n",
       "      <td>45</td>\n",
       "      <td>11022</td>\n",
       "      <td>11212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11212</td>\n",
       "      <td>11249</td>\n",
       "      <td>38</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.947</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11112</td>\n",
       "      <td>11349</td>\n",
       "      <td>11162</td>\n",
       "      <td>11299</td>\n",
       "      <td>243</td>\n",
       "      <td>50</td>\n",
       "      <td>46</td>\n",
       "      <td>11162</td>\n",
       "      <td>11349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11492</td>\n",
       "      <td>11555</td>\n",
       "      <td>64</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACC...</td>\n",
       "      <td>...</td>\n",
       "      <td>4 overlapping g4 in africa</td>\n",
       "      <td>11392</td>\n",
       "      <td>11655</td>\n",
       "      <td>11442</td>\n",
       "      <td>11605</td>\n",
       "      <td>100</td>\n",
       "      <td>243</td>\n",
       "      <td>47</td>\n",
       "      <td>11392</td>\n",
       "      <td>11655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11655</td>\n",
       "      <td>11692</td>\n",
       "      <td>38</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.816</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11555</td>\n",
       "      <td>11792</td>\n",
       "      <td>11605</td>\n",
       "      <td>11742</td>\n",
       "      <td>371</td>\n",
       "      <td>100</td>\n",
       "      <td>48</td>\n",
       "      <td>11555</td>\n",
       "      <td>11792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12063</td>\n",
       "      <td>12124</td>\n",
       "      <td>62</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.60</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11963</td>\n",
       "      <td>12224</td>\n",
       "      <td>12013</td>\n",
       "      <td>12174</td>\n",
       "      <td>222</td>\n",
       "      <td>371</td>\n",
       "      <td>49</td>\n",
       "      <td>11963</td>\n",
       "      <td>12224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12346</td>\n",
       "      <td>12450</td>\n",
       "      <td>105</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.76</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>12246</td>\n",
       "      <td>12550</td>\n",
       "      <td>12296</td>\n",
       "      <td>12500</td>\n",
       "      <td>92</td>\n",
       "      <td>222</td>\n",
       "      <td>50</td>\n",
       "      <td>12246</td>\n",
       "      <td>12542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12542</td>\n",
       "      <td>12573</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCACAACCCAAACAACCCAGCTCTCCCTAAGC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>12442</td>\n",
       "      <td>12673</td>\n",
       "      <td>12492</td>\n",
       "      <td>12623</td>\n",
       "      <td>378</td>\n",
       "      <td>92</td>\n",
       "      <td>51</td>\n",
       "      <td>12450</td>\n",
       "      <td>12673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13107</td>\n",
       "      <td>13150</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.909</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>13007</td>\n",
       "      <td>13250</td>\n",
       "      <td>13057</td>\n",
       "      <td>13200</td>\n",
       "      <td>280</td>\n",
       "      <td>25</td>\n",
       "      <td>54</td>\n",
       "      <td>13082</td>\n",
       "      <td>13250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13430</td>\n",
       "      <td>13458</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.030</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATACCTCTCACTTCAACCTCCCTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>13330</td>\n",
       "      <td>13558</td>\n",
       "      <td>13380</td>\n",
       "      <td>13508</td>\n",
       "      <td>167</td>\n",
       "      <td>280</td>\n",
       "      <td>55</td>\n",
       "      <td>13330</td>\n",
       "      <td>13558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13625</td>\n",
       "      <td>13705</td>\n",
       "      <td>81</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.050</td>\n",
       "      <td>-1.68</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13525</td>\n",
       "      <td>13805</td>\n",
       "      <td>13575</td>\n",
       "      <td>13755</td>\n",
       "      <td>30</td>\n",
       "      <td>167</td>\n",
       "      <td>56</td>\n",
       "      <td>13525</td>\n",
       "      <td>13735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13735</td>\n",
       "      <td>13809</td>\n",
       "      <td>75</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.170</td>\n",
       "      <td>-1.84</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13635</td>\n",
       "      <td>13909</td>\n",
       "      <td>13685</td>\n",
       "      <td>13859</td>\n",
       "      <td>150</td>\n",
       "      <td>30</td>\n",
       "      <td>57</td>\n",
       "      <td>13705</td>\n",
       "      <td>13909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13959</td>\n",
       "      <td>14004</td>\n",
       "      <td>46</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.935</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13859</td>\n",
       "      <td>14104</td>\n",
       "      <td>13909</td>\n",
       "      <td>14054</td>\n",
       "      <td>45</td>\n",
       "      <td>150</td>\n",
       "      <td>58</td>\n",
       "      <td>13859</td>\n",
       "      <td>14049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14049</td>\n",
       "      <td>14074</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCCACCTCCATCATCACCTCAACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13949</td>\n",
       "      <td>14174</td>\n",
       "      <td>13999</td>\n",
       "      <td>14124</td>\n",
       "      <td>35</td>\n",
       "      <td>45</td>\n",
       "      <td>59</td>\n",
       "      <td>14004</td>\n",
       "      <td>14109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14109</td>\n",
       "      <td>14169</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.951</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAACCTATT...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>14009</td>\n",
       "      <td>14269</td>\n",
       "      <td>14059</td>\n",
       "      <td>14219</td>\n",
       "      <td>56</td>\n",
       "      <td>35</td>\n",
       "      <td>60</td>\n",
       "      <td>14074</td>\n",
       "      <td>14225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14225</td>\n",
       "      <td>14292</td>\n",
       "      <td>68</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-1.52</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCCATAATCATACAAAGCCCCCGCACCAATAGGATCCTCCCGAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in africa</td>\n",
       "      <td>14125</td>\n",
       "      <td>14392</td>\n",
       "      <td>14175</td>\n",
       "      <td>14342</td>\n",
       "      <td>36</td>\n",
       "      <td>56</td>\n",
       "      <td>61</td>\n",
       "      <td>14169</td>\n",
       "      <td>14328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14751</td>\n",
       "      <td>14832</td>\n",
       "      <td>82</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.110</td>\n",
       "      <td>-1.84</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in afica</td>\n",
       "      <td>14651</td>\n",
       "      <td>14932</td>\n",
       "      <td>14701</td>\n",
       "      <td>14882</td>\n",
       "      <td>528</td>\n",
       "      <td>98</td>\n",
       "      <td>67</td>\n",
       "      <td>14653</td>\n",
       "      <td>14932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15360</td>\n",
       "      <td>15391</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAACAACCCCCTAGGAATCACCTCCCATTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15260</td>\n",
       "      <td>15491</td>\n",
       "      <td>15310</td>\n",
       "      <td>15441</td>\n",
       "      <td>111</td>\n",
       "      <td>528</td>\n",
       "      <td>68</td>\n",
       "      <td>15260</td>\n",
       "      <td>15491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15502</td>\n",
       "      <td>15556</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.450</td>\n",
       "      <td>-2.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15402</td>\n",
       "      <td>15656</td>\n",
       "      <td>15452</td>\n",
       "      <td>15606</td>\n",
       "      <td>80</td>\n",
       "      <td>111</td>\n",
       "      <td>69</td>\n",
       "      <td>15402</td>\n",
       "      <td>15636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>15536</td>\n",
       "      <td>15783</td>\n",
       "      <td>15586</td>\n",
       "      <td>15733</td>\n",
       "      <td>484</td>\n",
       "      <td>80</td>\n",
       "      <td>70</td>\n",
       "      <td>15556</td>\n",
       "      <td>15783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16067</td>\n",
       "      <td>16311</td>\n",
       "      <td>16117</td>\n",
       "      <td>16261</td>\n",
       "      <td>28</td>\n",
       "      <td>484</td>\n",
       "      <td>71</td>\n",
       "      <td>16067</td>\n",
       "      <td>16239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>63</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16139</td>\n",
       "      <td>16401</td>\n",
       "      <td>16189</td>\n",
       "      <td>16351</td>\n",
       "      <td>43</td>\n",
       "      <td>28</td>\n",
       "      <td>72</td>\n",
       "      <td>16211</td>\n",
       "      <td>16344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16244</td>\n",
       "      <td>16482</td>\n",
       "      <td>16294</td>\n",
       "      <td>16432</td>\n",
       "      <td>145</td>\n",
       "      <td>43</td>\n",
       "      <td>73</td>\n",
       "      <td>16301</td>\n",
       "      <td>16482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start    end  width  \\\n",
       "2   NC_012920.1 Homo sapiens mitochondrion, comple...    418    592    175   \n",
       "3   NC_012920.1 Homo sapiens mitochondrion, comple...    785    808     24   \n",
       "4   NC_012920.1 Homo sapiens mitochondrion, comple...    954    984     31   \n",
       "5   NC_012920.1 Homo sapiens mitochondrion, comple...   1667   1707     41   \n",
       "6   NC_012920.1 Homo sapiens mitochondrion, comple...   2057   2079     23   \n",
       "7   NC_012920.1 Homo sapiens mitochondrion, comple...   2212   2236     25   \n",
       "10  NC_012920.1 Homo sapiens mitochondrion, comple...   3471   3604    134   \n",
       "11  NC_012920.1 Homo sapiens mitochondrion, comple...   3874   3909     36   \n",
       "12  NC_012920.1 Homo sapiens mitochondrion, comple...   3951   3984     34   \n",
       "15  NC_012920.1 Homo sapiens mitochondrion, comple...   4228   4260     33   \n",
       "16  NC_012920.1 Homo sapiens mitochondrion, comple...   4372   4398     27   \n",
       "17  NC_012920.1 Homo sapiens mitochondrion, comple...   4456   4505     50   \n",
       "18  NC_012920.1 Homo sapiens mitochondrion, comple...   4875   4899     25   \n",
       "19  NC_012920.1 Homo sapiens mitochondrion, comple...   5187   5255     69   \n",
       "22  NC_012920.1 Homo sapiens mitochondrion, comple...   5424   5499     76   \n",
       "23  NC_012920.1 Homo sapiens mitochondrion, comple...   5875   5916     42   \n",
       "24  NC_012920.1 Homo sapiens mitochondrion, comple...   6213   6242     30   \n",
       "25  NC_012920.1 Homo sapiens mitochondrion, comple...   6410   6464     55   \n",
       "26  NC_012920.1 Homo sapiens mitochondrion, comple...   6545   6572     28   \n",
       "27  NC_012920.1 Homo sapiens mitochondrion, comple...   6826   6851     26   \n",
       "28  NC_012920.1 Homo sapiens mitochondrion, comple...   7386   7427     42   \n",
       "29  NC_012920.1 Homo sapiens mitochondrion, comple...   7798   7840     43   \n",
       "30  NC_012920.1 Homo sapiens mitochondrion, comple...   7936   7974     39   \n",
       "31  NC_012920.1 Homo sapiens mitochondrion, comple...   8252   8305     54   \n",
       "32  NC_012920.1 Homo sapiens mitochondrion, comple...   8455   8488     34   \n",
       "35  NC_012920.1 Homo sapiens mitochondrion, comple...   8791   8815     25   \n",
       "36  NC_012920.1 Homo sapiens mitochondrion, comple...   8917   8958     42   \n",
       "39  NC_012920.1 Homo sapiens mitochondrion, comple...   9505   9607    103   \n",
       "40  NC_012920.1 Homo sapiens mitochondrion, comple...  10175  10214     40   \n",
       "41  NC_012920.1 Homo sapiens mitochondrion, comple...  10257  10303     47   \n",
       "42  NC_012920.1 Homo sapiens mitochondrion, comple...  10603  10637     35   \n",
       "43  NC_012920.1 Homo sapiens mitochondrion, comple...  10912  10990     79   \n",
       "44  NC_012920.1 Homo sapiens mitochondrion, comple...  11122  11162     41   \n",
       "45  NC_012920.1 Homo sapiens mitochondrion, comple...  11212  11249     38   \n",
       "46  NC_012920.1 Homo sapiens mitochondrion, comple...  11492  11555     64   \n",
       "47  NC_012920.1 Homo sapiens mitochondrion, comple...  11655  11692     38   \n",
       "48  NC_012920.1 Homo sapiens mitochondrion, comple...  12063  12124     62   \n",
       "49  NC_012920.1 Homo sapiens mitochondrion, comple...  12346  12450    105   \n",
       "50  NC_012920.1 Homo sapiens mitochondrion, comple...  12542  12573     32   \n",
       "53  NC_012920.1 Homo sapiens mitochondrion, comple...  13107  13150     44   \n",
       "54  NC_012920.1 Homo sapiens mitochondrion, comple...  13430  13458     29   \n",
       "55  NC_012920.1 Homo sapiens mitochondrion, comple...  13625  13705     81   \n",
       "56  NC_012920.1 Homo sapiens mitochondrion, comple...  13735  13809     75   \n",
       "57  NC_012920.1 Homo sapiens mitochondrion, comple...  13959  14004     46   \n",
       "58  NC_012920.1 Homo sapiens mitochondrion, comple...  14049  14074     26   \n",
       "59  NC_012920.1 Homo sapiens mitochondrion, comple...  14109  14169     61   \n",
       "60  NC_012920.1 Homo sapiens mitochondrion, comple...  14225  14292     68   \n",
       "66  NC_012920.1 Homo sapiens mitochondrion, comple...  14751  14832     82   \n",
       "67  NC_012920.1 Homo sapiens mitochondrion, comple...  15360  15391     32   \n",
       "68  NC_012920.1 Homo sapiens mitochondrion, comple...  15502  15556     55   \n",
       "69  NC_012920.1 Homo sapiens mitochondrion, comple...  15636  15683     48   \n",
       "70  NC_012920.1 Homo sapiens mitochondrion, comple...  16167  16211     45   \n",
       "71  NC_012920.1 Homo sapiens mitochondrion, comple...  16239  16301     63   \n",
       "72  NC_012920.1 Homo sapiens mitochondrion, comple...  16344  16382     39   \n",
       "\n",
       "   strand  score  max_score  threshold  window  \\\n",
       "2       - -1.150      -1.88          1      25   \n",
       "3       - -1.210      -1.16          1      25   \n",
       "4       - -1.390      -1.48          1      25   \n",
       "5       - -1.020      -1.48          1      25   \n",
       "6       - -1.520      -1.40          1      25   \n",
       "7       - -1.000      -1.00          1      25   \n",
       "10      - -1.170      -2.76          1      25   \n",
       "11      - -0.917      -1.28          1      25   \n",
       "12      - -0.882      -1.20          1      25   \n",
       "15      - -1.210      -1.28          1      25   \n",
       "16      - -1.220      -1.28          1      25   \n",
       "17      - -1.100      -1.36          1      25   \n",
       "18      - -1.080      -1.08          1      25   \n",
       "19      - -0.957      -1.24          1      25   \n",
       "22      - -1.200      -1.92          1      25   \n",
       "23      - -0.881      -1.24          1      25   \n",
       "24      - -0.900      -1.04          1      25   \n",
       "25      - -1.050      -1.44          1      25   \n",
       "26      - -1.210      -1.24          1      25   \n",
       "27      - -1.040      -1.04          1      25   \n",
       "28      - -1.020      -1.56          1      25   \n",
       "29      - -0.977      -1.32          1      25   \n",
       "30      - -0.923      -1.20          1      25   \n",
       "31      - -1.220      -2.00          1      25   \n",
       "32      - -1.090      -1.36          1      25   \n",
       "35      - -1.000      -1.00          1      25   \n",
       "36      - -0.881      -1.56          1      25   \n",
       "39      - -1.020      -1.96          1      25   \n",
       "40      - -1.200      -1.80          1      25   \n",
       "41      - -0.894      -1.36          1      25   \n",
       "42      - -1.060      -1.28          1      25   \n",
       "43      - -1.330      -2.08          1      25   \n",
       "44      - -0.854      -1.12          1      25   \n",
       "45      - -0.947      -1.24          1      25   \n",
       "46      - -1.000      -1.32          1      25   \n",
       "47      - -0.816      -1.20          1      25   \n",
       "48      - -1.210      -1.60          1      25   \n",
       "49      - -1.060      -1.76          1      25   \n",
       "50      - -1.060      -1.20          1      25   \n",
       "53      - -0.909      -1.24          1      25   \n",
       "54      - -1.030      -1.04          1      25   \n",
       "55      - -1.050      -1.68          1      25   \n",
       "56      - -1.170      -1.84          1      25   \n",
       "57      - -0.935      -1.28          1      25   \n",
       "58      - -1.120      -1.12          1      25   \n",
       "59      - -0.951      -1.20          1      25   \n",
       "60      - -1.070      -1.52          1      25   \n",
       "66      - -1.110      -1.84          1      25   \n",
       "67      - -1.120      -1.36          1      25   \n",
       "68      - -1.450      -2.12          1      25   \n",
       "69      - -0.896      -1.16          1      25   \n",
       "70      - -1.160      -1.80          1      25   \n",
       "71      - -0.968      -1.40          1      25   \n",
       "72      - -1.210      -1.48          1      25   \n",
       "\n",
       "                                             sequence  ...  \\\n",
       "2   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...  ...   \n",
       "3                            CGCTTAGCCTAGCCACACCCCCAC  ...   \n",
       "4                     CACCCCCTCCCCAATAAAGCTAAAACTCACC  ...   \n",
       "5           CTGAGCTAAACCTAGCCCCAAACCCACTCCACCTTACTACC  ...   \n",
       "6                             CCCACAGAACCCTCTAAATCCCC  ...   \n",
       "7                           CAACACCCACTACCTAAAAAATCCC  ...   \n",
       "10  CTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTC...  ...   \n",
       "11               CTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCC  ...   \n",
       "12                 CGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCC  ...   \n",
       "15                  CCCATTACAATCTCCAGCATTCCCCCTCAAACC  ...   \n",
       "16                        CTCCGTGCCACCTATCACACCCCATCC  ...   \n",
       "17  CCCTTCCCGTACTAATTAATCCCCTGGCCCAACCCGTCATCTACTC...  ...   \n",
       "18                          CTAGCCCCCATCTCAATCATATACC  ...   \n",
       "19  CTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGC...  ...   \n",
       "22  CATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCA...  ...   \n",
       "23         CACTCAGCCATTTTACCTCACCCCCACTGATGTTCGCCGACC  ...   \n",
       "24                     CTCTTACCTCCCTCTCTCCTACTCCTGCTC  ...   \n",
       "25  CAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTC...  ...   \n",
       "26                       CCTCAACACCACCTTCTTCGACCCCGCC  ...   \n",
       "27                         CTACCATAATCATCGCTATCCCCACC  ...   \n",
       "28         CTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCC  ...   \n",
       "29        CATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTAC  ...   \n",
       "30            CTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACC  ...   \n",
       "31  CCCGTATTTACCCTATAGCACCCCCTCTACCCCCTCTAGAGCCCAC...  ...   \n",
       "32                 CACAAACTACCACCTACCTCCCTCACCAAAGCCC  ...   \n",
       "35                          CCTCACTCATTTACACCAACCACCC  ...   \n",
       "36         CAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATC  ...   \n",
       "39  CCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCA...  ...   \n",
       "40           CGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCC  ...   \n",
       "41    CTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAAC  ...   \n",
       "42                CTCTCATAACCCTCAACACCCACTCCCTCTTAGCC  ...   \n",
       "43  CTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTA...  ...   \n",
       "44          CTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCC  ...   \n",
       "45             CTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCAC  ...   \n",
       "46  CGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACC...  ...   \n",
       "47             CAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGC  ...   \n",
       "48  CCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAA...  ...   \n",
       "49  CACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCA...  ...   \n",
       "50                   CCACAACCCAAACAACCCAGCTCTCCCTAAGC  ...   \n",
       "53       CTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCAC  ...   \n",
       "54                      CCATACCTCTCACTTCAACCTCCCTCACC  ...   \n",
       "55  CCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGA...  ...   \n",
       "56  CTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAA...  ...   \n",
       "57     CCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACC  ...   \n",
       "58                         CTCCACCTCCATCATCACCTCAACCC  ...   \n",
       "59  CTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAACCTATT...  ...   \n",
       "60  CGCCCATAATCATACAAAGCCCCCGCACCAATAGGATCCTCCCGAA...  ...   \n",
       "66  CCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTC...  ...   \n",
       "67                   CAAACAACCCCCTAGGAATCACCTCCCATTCC  ...   \n",
       "68  CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...  ...   \n",
       "69   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC  ...   \n",
       "70      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC  ...   \n",
       "71  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...  ...   \n",
       "72            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC  ...   \n",
       "\n",
       "                          note start-100bp end+100bp  start-50bp  end+50bp  \\\n",
       "2   5 overlapping g4 in africa         318       692         368       642   \n",
       "3                          NaN         685       908         735       858   \n",
       "4                          NaN         854      1084         904      1034   \n",
       "5          more than in africa        1567      1807        1617      1757   \n",
       "6                          NaN        1957      2179        2007      2129   \n",
       "7                          NaN        2112      2336        2162      2286   \n",
       "10  5 overlapping g4 in africa        3371      3704        3421      3654   \n",
       "11                         NaN        3774      4009        3824      3959   \n",
       "12            more than africa        3851      4084        3901      4034   \n",
       "15                         NaN        4128      4360        4178      4310   \n",
       "16                         NaN        4272      4498        4322      4448   \n",
       "17  3 overlapping g4 in africa        4356      4605        4406      4555   \n",
       "18                         NaN        4775      4999        4825      4949   \n",
       "19  2 overlapping g4 in africa        5087      5355        5137      5305   \n",
       "22              2 g4 in africa        5324      5599        5374      5549   \n",
       "23            more than africa        5775      6016        5825      5966   \n",
       "24  2 overlapping g4 in africa        6113      6342        6163      6292   \n",
       "25            more than africa        6310      6564        6360      6514   \n",
       "26            less than africa        6445      6672        6495      6622   \n",
       "27                         NaN        6726      6951        6776      6901   \n",
       "28            more than africa        7286      7527        7336      7477   \n",
       "29                         NaN        7698      7940        7748      7890   \n",
       "30                         NaN        7836      8074        7886      8024   \n",
       "31            more than africa        8152      8405        8202      8355   \n",
       "32                         NaN        8355      8588        8405      8538   \n",
       "35                         NaN        8691      8915        8741      8865   \n",
       "36            more than africa        8817      9058        8867      9008   \n",
       "39              2 g4 in africa        9405      9707        9455      9657   \n",
       "40            more than africa       10075     10314       10125     10264   \n",
       "41            more than africa       10157     10403       10207     10353   \n",
       "42                         NaN       10503     10737       10553     10687   \n",
       "43              2 g4 in africa       10812     11090       10862     11040   \n",
       "44            more than africa       11022     11262       11072     11212   \n",
       "45            more than africa       11112     11349       11162     11299   \n",
       "46  4 overlapping g4 in africa       11392     11655       11442     11605   \n",
       "47            more than africa       11555     11792       11605     11742   \n",
       "48            more than africa       11963     12224       12013     12174   \n",
       "49  2 overlapping g4 in africa       12246     12550       12296     12500   \n",
       "50            more than africa       12442     12673       12492     12623   \n",
       "53  2 overlapping g4 in africa       13007     13250       13057     13200   \n",
       "54  2 overlapping g4 in africa       13330     13558       13380     13508   \n",
       "55            more than africa       13525     13805       13575     13755   \n",
       "56            more than africa       13635     13909       13685     13859   \n",
       "57            more than africa       13859     14104       13909     14054   \n",
       "58                         NaN       13949     14174       13999     14124   \n",
       "59  2 overlapping g4 in africa       14009     14269       14059     14219   \n",
       "60  3 overlapping g4 in africa       14125     14392       14175     14342   \n",
       "66   3 overlapping g4 in afica       14651     14932       14701     14882   \n",
       "67                         NaN       15260     15491       15310     15441   \n",
       "68                         NaN       15402     15656       15452     15606   \n",
       "69            more than africa       15536     15783       15586     15733   \n",
       "70            more than africa       16067     16311       16117     16261   \n",
       "71            more than africa       16139     16401       16189     16351   \n",
       "72            more than africa       16244     16482       16294     16432   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  G4_id  up_win  down_win  \n",
       "2                 193                     40      3     378       692  \n",
       "3                 146                    193      4     685       908  \n",
       "4                 683                    146      5     854      1084  \n",
       "5                 350                    683      6    1567      1807  \n",
       "6                 133                    350      7    1957      2179  \n",
       "7                 912                    133      8    2112      2336  \n",
       "10                270                    249     11    3371      3704  \n",
       "11                 42                    270     12    3774      3951  \n",
       "12                102                     42     13    3909      4084  \n",
       "15                112                     67     16    4161      4360  \n",
       "16                 58                    112     17    4272      4456  \n",
       "17                370                     58     18    4398      4605  \n",
       "18                288                    370     19    4775      4999  \n",
       "19                 33                    288     20    5087      5288  \n",
       "22                376                     32     23    5392      5599  \n",
       "23                297                    376     24    5775      6016  \n",
       "24                168                    297     25    6113      6342  \n",
       "25                 81                    168     26    6310      6545  \n",
       "26                254                     81     27    6464      6672  \n",
       "27                535                    254     28    6726      6951  \n",
       "28                371                    535     29    7286      7527  \n",
       "29                 96                    371     30    7698      7936  \n",
       "30                278                     96     31    7840      8074  \n",
       "31                150                    278     32    8152      8405  \n",
       "32                 56                    150     33    8355      8544  \n",
       "35                102                    148     36    8691      8915  \n",
       "36                285                    102     37    8817      9058  \n",
       "39                568                    212     40    9405      9707  \n",
       "40                 43                    568     41   10075     10257  \n",
       "41                300                     43     42   10214     10403  \n",
       "42                275                    300     43   10503     10737  \n",
       "43                132                    275     44   10812     11090  \n",
       "44                 50                    132     45   11022     11212  \n",
       "45                243                     50     46   11162     11349  \n",
       "46                100                    243     47   11392     11655  \n",
       "47                371                    100     48   11555     11792  \n",
       "48                222                    371     49   11963     12224  \n",
       "49                 92                    222     50   12246     12542  \n",
       "50                378                     92     51   12450     12673  \n",
       "53                280                     25     54   13082     13250  \n",
       "54                167                    280     55   13330     13558  \n",
       "55                 30                    167     56   13525     13735  \n",
       "56                150                     30     57   13705     13909  \n",
       "57                 45                    150     58   13859     14049  \n",
       "58                 35                     45     59   14004     14109  \n",
       "59                 56                     35     60   14074     14225  \n",
       "60                 36                     56     61   14169     14328  \n",
       "66                528                     98     67   14653     14932  \n",
       "67                111                    528     68   15260     15491  \n",
       "68                 80                    111     69   15402     15636  \n",
       "69                484                     80     70   15556     15783  \n",
       "70                 28                    484     71   16067     16239  \n",
       "71                 43                     28     72   16211     16344  \n",
       "72                145                     43     73   16301     16482  \n",
       "\n",
       "[54 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_25bp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b7c62411-85ae-47ce-be2d-a8ff82ff1582",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_25bp['start'] = df_25bp['start'].apply(lambda x: x-25)\n",
    "# df_25bp['end'] = df_25bp['end'].apply(lambda x: x+25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3aee0d0d-a77a-467f-be17-1f1434aa79a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_start = []\n",
    "\n",
    "for start, end, G4_id in zip(df_25bp['start'], \n",
    "                             df_25bp['end'], \n",
    "                             df_25bp['G4_id']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i in range(start-25, end+25):\n",
    "        row = []\n",
    "        \n",
    "        if i+WINDOW_SIZE <= end+25:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            break\n",
    "\n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            if pos != 3107:\n",
    "                window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "            else:\n",
    "                window_seq += ''\n",
    "        \n",
    "        # print(i, window, WINDOW_SIZE, end)  \n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('check_start')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_check_start.append(row)\n",
    "# print(data_check_start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "05d43e76-5d2a-4d22-9cb8-d017a355dfc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_end = []\n",
    "\n",
    "for start, end, G4_id in zip(df_25bp['start'], \n",
    "                             df_25bp['end'], \n",
    "                             df_25bp['G4_id']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i in range(end+25, start-25, -1):\n",
    "        row = []\n",
    "\n",
    "        \n",
    "        if i+WINDOW_SIZE > end+25:\n",
    "            continue  \n",
    "        elif (i+WINDOW_SIZE <= end+25):\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            if pos != 3107:\n",
    "                window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "            else:\n",
    "                window_seq += ''\n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('check_end')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_check_end.append(row)\n",
    "# print(data_check_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "51dad145-e328-43d4-861b-5a0427c3b1fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_sides = data_check_start + data_check_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc4b7853-81ed-4d14-97e6-e2e9ab836451",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_check_sides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b22715ac-2844-46f0-a0be-8962237f98d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = pd.DataFrame(data=data_check_sides, \n",
    "                             columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "00dd3f17-098e-48af-9313-7bfefbb2a03a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4901</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>1</td>\n",
       "      <td>607</td>\n",
       "      <td>616</td>\n",
       "      <td>CACTGAAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4902</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>2</td>\n",
       "      <td>606</td>\n",
       "      <td>615</td>\n",
       "      <td>ACACTGAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4903</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>3</td>\n",
       "      <td>605</td>\n",
       "      <td>614</td>\n",
       "      <td>TACACTGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4904</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>4</td>\n",
       "      <td>604</td>\n",
       "      <td>613</td>\n",
       "      <td>ATACACTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4905</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>5</td>\n",
       "      <td>603</td>\n",
       "      <td>612</td>\n",
       "      <td>AATACACTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5110</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>210</td>\n",
       "      <td>398</td>\n",
       "      <td>407</td>\n",
       "      <td>TTTTATCTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5111</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>211</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5112</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>212</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5113</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>213</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5114</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>214</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id       side  number_of_win  window_start  window_end    sequence\n",
       "4901      3  check_end              1           607         616  CACTGAAAAT\n",
       "4902      3  check_end              2           606         615  ACACTGAAAA\n",
       "4903      3  check_end              3           605         614  TACACTGAAA\n",
       "4904      3  check_end              4           604         613  ATACACTGAA\n",
       "4905      3  check_end              5           603         612  AATACACTGA\n",
       "...     ...        ...            ...           ...         ...         ...\n",
       "5110      3  check_end            210           398         407  TTTTATCTTT\n",
       "5111      3  check_end            211           397         406  ATTTTATCTT\n",
       "5112      3  check_end            212           396         405  AATTTTATCT\n",
       "5113      3  check_end            213           395         404  AAATTTTATC\n",
       "5114      3  check_end            214           394         403  CAAATTTTAT\n",
       "\n",
       "[214 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_check_sides[(df_win_check_sides['side'] == 'check_end') & (df_win_check_sides['G4_id'] == 3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "358e7e47-6499-4e36-9995-c248bfa19092",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_win_check_sides = df_win_check_sides['sequence'].apply(lambda x: Counter(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "45f31c06-2989-4e49-9cc6-f99a3a499ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_df_win_check_sides = pd.DataFrame.from_records(nuc_count_win_check_sides).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "56b82dde-2e52-493f-8621-b2a785802bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = df_win_check_sides.join(nuc_count_df_win_check_sides).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "07ae5c08-8751-49d0-82e9-5988297cd0a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>1</td>\n",
       "      <td>393</td>\n",
       "      <td>402</td>\n",
       "      <td>TCAAATTTTA</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>2</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>3</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>4</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>5</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9743</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>74</td>\n",
       "      <td>16324</td>\n",
       "      <td>16333</td>\n",
       "      <td>TTACCGTACA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9744</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>75</td>\n",
       "      <td>16323</td>\n",
       "      <td>16332</td>\n",
       "      <td>TTTACCGTAC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9745</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>76</td>\n",
       "      <td>16322</td>\n",
       "      <td>16331</td>\n",
       "      <td>ATTTACCGTA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9746</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>77</td>\n",
       "      <td>16321</td>\n",
       "      <td>16330</td>\n",
       "      <td>CATTTACCGT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9747</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>78</td>\n",
       "      <td>16320</td>\n",
       "      <td>16329</td>\n",
       "      <td>CCATTTACCG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9748 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id         side  number_of_win  window_start  window_end    sequence  \\\n",
       "0         3  check_start              1           393         402  TCAAATTTTA   \n",
       "1         3  check_start              2           394         403  CAAATTTTAT   \n",
       "2         3  check_start              3           395         404  AAATTTTATC   \n",
       "3         3  check_start              4           396         405  AATTTTATCT   \n",
       "4         3  check_start              5           397         406  ATTTTATCTT   \n",
       "...     ...          ...            ...           ...         ...         ...   \n",
       "9743     73    check_end             74         16324       16333  TTACCGTACA   \n",
       "9744     73    check_end             75         16323       16332  TTTACCGTAC   \n",
       "9745     73    check_end             76         16322       16331  ATTTACCGTA   \n",
       "9746     73    check_end             77         16321       16330  CATTTACCGT   \n",
       "9747     73    check_end             78         16320       16329  CCATTTACCG   \n",
       "\n",
       "        A    C    G    T  \n",
       "0     4.0  1.0  0.0  5.0  \n",
       "1     4.0  1.0  0.0  5.0  \n",
       "2     4.0  1.0  0.0  5.0  \n",
       "3     3.0  1.0  0.0  6.0  \n",
       "4     2.0  1.0  0.0  7.0  \n",
       "...   ...  ...  ...  ...  \n",
       "9743  3.0  3.0  1.0  3.0  \n",
       "9744  2.0  3.0  1.0  4.0  \n",
       "9745  3.0  2.0  1.0  4.0  \n",
       "9746  2.0  3.0  1.0  4.0  \n",
       "9747  2.0  4.0  1.0  3.0  \n",
       "\n",
       "[9748 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_check_sides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "60162525-7f65-423f-b8d1-de2ab56c24e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_check_sides.to_csv('../data/processed/df_win_check_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65bfe663-615f-49cc-a352-7e56b9a2db5a",
   "metadata": {},
   "source": [
    "Collect mutations from cancer dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ec18b2df-8934-43bf-ac78-663d5f46ee5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0b3f4996-771e-4be3-84ec-11ba5e205960",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = pd.read_csv('../data/processed/df_win_check_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1f9bb812-c75a-48e2-860e-b173bb19fca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_check_sides = []\n",
    "for window_start, window_end in zip(df_win_check_sides['window_start'], df_win_check_sides['window_end']):\n",
    "    row = [*range(window_start, window_end)]\n",
    "    Mut_win_check_sides.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0f238d1e-4d9c-47a3-bfad-cad3a809ba60",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_check_sides = pd.DataFrame(Mut_win_check_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "22564adc-bbdd-4285-ab84-ac9094f5b063",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_check_sides[0] = Mut_win_df_check_sides[0].str.findall(\"[A-T]>[A-T]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "15b6e619-bfe3-47ab-abdc-e285c92fa083",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutwin_check_sides = pd.DataFrame.from_records(Mut_win_df_check_sides[0].apply(lambda x: Counter(x))) \\\n",
    "    .fillna(0) \\\n",
    "    .sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3220bf42-2b40-4f03-be42-9f42b350ad09",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_check_sides = df_win_check_sides.join(obsMutwin_check_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9f520752-d0df-489a-8a98-250e2a33eff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_check_sides['number_of_Mut'] = df_win_cancer_check_sides[['A>C','A>G','A>T',\n",
    "                                                              'C>A','C>G','C>T',\n",
    "                                                              'G>A','G>C','G>T',\n",
    "                                                              'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2e333902-32ab-4cb6-b040-a1460db43805",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>1</td>\n",
       "      <td>393</td>\n",
       "      <td>402</td>\n",
       "      <td>TCAAATTTTA</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>2</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>3</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>4</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>5</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id         side  number_of_win  window_start  window_end    sequence  \\\n",
       "0      3  check_start              1           393         402  TCAAATTTTA   \n",
       "1      3  check_start              2           394         403  CAAATTTTAT   \n",
       "2      3  check_start              3           395         404  AAATTTTATC   \n",
       "3      3  check_start              4           396         405  AATTTTATCT   \n",
       "4      3  check_start              5           397         406  ATTTTATCTT   \n",
       "\n",
       "     A    C    G    T  ...  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G  \\\n",
       "0  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "1  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "2  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3  3.0  1.0  0.0  6.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "4  2.0  1.0  0.0  7.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0            0.0  \n",
       "1            0.0  \n",
       "2            1.0  \n",
       "3            1.0  \n",
       "4            1.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_cancer_check_sides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7be9f257-ac80-4d4b-8bb3-38b245f1ab31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_check_sides.to_csv('../data/processed/df_win_cancer_check_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b97fbb8e-0bed-4e14-805f-bd06e7af308a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
