{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b299f0c5-23a6-4609-9f91-c65d0994bed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bc4ad765-e905-4d20-b63c-202f99990d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference:\n",
    "ref_annot = pd.read_csv('../data/raw/ref_annot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "00b3cbf7-9d10-427d-b534-c343a37a267c",
   "metadata": {},
   "outputs": [],
   "source": [
    "REGION_SIZE = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5bc1a977-0763-4a16-b3c3-d3a4ff789e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "WINDOW_SIZE = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebe27959-21cb-4b6e-becd-7e8bffa4c30c",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = pd.read_csv('../data/processed/g4conf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7670f0dd-f84d-45e7-88cc-0c21fa83c9f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqs = pd.read_csv('../data/raw/pqsfinder_mtDNA_human_th15.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d397b691-746a-4a6e-9dab-b800f223c8ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf = pd.read_csv('../data/processed/g4pqsconf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4cd36d-cefe-4e7b-ad0b-2ed81003d3fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab916c37-d366-455b-8984-ae8837c7f252",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_columns(df:pd.DataFrame): # data_preparation step 2\n",
    "    \n",
    "    bp_untill_next_G4 = df['start'].drop(0).values - df['end'].drop(df.shape[0]-1).values\n",
    "    bp_untill_next_G4 = bp_untill_next_G4.tolist()\n",
    "    bp_untill_next_G4.append(len(ref_annot['Pos']) - df['end'][len(df['start'])-1] + df['start'][0])  #bp between last G4 and first G4\n",
    "\n",
    "    bp_untill_previous_G4 = df['start'].drop(0).values - df['end'].drop(df.shape[0]-1).values\n",
    "    bp_untill_previous_G4 = bp_untill_previous_G4.tolist()\n",
    "    bp_untill_previous_G4.insert(0, len(ref_annot['Pos']) - df['end'][len(df['start'])-1] + df['start'][0]) #bp between fist and last\n",
    "\n",
    "    df['bp_untill_next_G4'] = bp_untill_next_G4\n",
    "    df['bp_untill_previous_G4'] = bp_untill_previous_G4\n",
    "\n",
    "    if 'G4_id' in df.columns:\n",
    "        None\n",
    "    else:\n",
    "        df['G4_id'] = [*range(1,df.shape[0]+1)]\n",
    "\n",
    "    df['up_win'] = pd.concat([df[df['bp_untill_previous_G4'] <= 100][['bp_untill_previous_G4', 'start']]\\\n",
    "                                  .diff(axis=1)['start'],\\\n",
    "                                 df[df['bp_untill_previous_G4'] > 100]['start']-100])\n",
    "    \n",
    "    df['down_win'] = pd.concat([df[df['bp_untill_next_G4'] <= 100][['end', 'bp_untill_next_G4']]\\\n",
    "                                    .sum(axis=1), \\\n",
    "                               df[df['bp_untill_next_G4'] > 100]['end']+100])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f0f1357-5ab0-4d24-be4c-6137205f0972",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "79d5f841-f56a-402f-84e0-cc431b028e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_win_up(df:pd.DataFrame()): #data preparation step 3, windows around G4\n",
    "    \n",
    "    data_up = []\n",
    "    \n",
    "    for up_win, start, G4_id in zip(df['up_win'], df['start'], df['G4_id']):\n",
    "        \n",
    "        count_window = start - up_win - WINDOW_SIZE + 1  # count window in upstream interval (before G4)\n",
    "        number_of_win = count_window + 1                 # reverse counting of windows; the closer to G4, the smaller the number\n",
    "        \n",
    "        for i in range(up_win, start):\n",
    "            row = []\n",
    "            \n",
    "            if i+WINDOW_SIZE <= start:\n",
    "                window = [*range(i, i + WINDOW_SIZE)]\n",
    "            else:\n",
    "                break\n",
    "    #         print(i, window)    \n",
    "            window_seq = ''\n",
    "            for pos in window:\n",
    "                if pos != 3107:\n",
    "                    window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "                else:\n",
    "                    window_seq += ''\n",
    "    #         print(pos, window_seq)\n",
    "    \n",
    "    #       ref_annot[ref_annot['Pos'].isin(window)].Nuc.value_counts()\n",
    "            \n",
    "            number_of_win -= 1\n",
    "            \n",
    "            row.append(G4_id)\n",
    "            row.append('up')\n",
    "            row.append(number_of_win)\n",
    "            row.append(window[0])\n",
    "            row.append(window[-1])\n",
    "            row.append(window_seq)\n",
    "            \n",
    "            data_up.append(row)\n",
    "    # print(data_up)\n",
    "    return data_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0aafe402-746c-4c85-864e-7142967905d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_win_down(df:pd.DataFrame()): #data preparation step 3, windows around G4\n",
    "    data_down = []\n",
    "    \n",
    "    for end, down_win, G4_id in zip(df['end'], df['down_win'], df['G4_id']):\n",
    "        \n",
    "        number_of_win = 0\n",
    "        \n",
    "        for i in range(end, down_win):\n",
    "            row = []\n",
    "            \n",
    "            if i >= g4conf['end'].iloc[-1] | i <= ref_annot['Pos'].iloc[-1]:\n",
    "                window = [*range(i, ref_annot['Pos'].iloc[-1]+1)]\n",
    "                remain = WINDOW_SIZE - (ref_annot['Pos'].iloc[-1] - i)\n",
    "                window += [*range(ref_annot['Pos'].iloc[0], remain)]\n",
    "            \n",
    "            elif i > ref_annot['Pos'].iloc[-1] and i <= down_win - WINDOW_SIZE :\n",
    "                i = i - ref_annot['Pos'].iloc[-1]\n",
    "                window = [*range(i, i + WINDOW_SIZE)]\n",
    "            \n",
    "            elif i+WINDOW_SIZE < down_win:\n",
    "                window = [*range(i, i + WINDOW_SIZE)]\n",
    "                   \n",
    "            else:\n",
    "                break\n",
    "                \n",
    "            window_seq = ''\n",
    "            for pos in window:\n",
    "                window_seq += ref_annot[ref_annot['Pos'] == pos].Nuc.values[0]\n",
    "    #         print(window[0], window_seq)\n",
    "    \n",
    "    #     ref_annot[ref_annot['Pos'].isin(window)].Nuc.value_counts()\n",
    "    \n",
    "            number_of_win += 1\n",
    "    \n",
    "            row.append(G4_id)\n",
    "            row.append('down')\n",
    "            row.append(number_of_win)\n",
    "            row.append(window[0])\n",
    "            row.append(window[-1])\n",
    "            row.append(window_seq)\n",
    "            \n",
    "            data_down.append(row)\n",
    "    # print(data_down)\n",
    "    return data_down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6e575fba-f7bb-488a-b115-4d2241535473",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_win_df(data_up, data_down): #data preparation step 3, df_win around G4 and within G4\n",
    "    data = data_up + data_down\n",
    "    df_win = pd.DataFrame(data=data, columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])\n",
    "    nuc_count_win = df_win['sequence'].apply(lambda x: Counter(x))\n",
    "    nuc_count_df_win = pd.DataFrame.from_records(nuc_count_win).sort_index(axis=1)\n",
    "    df_win = df_win.join(nuc_count_df_win).fillna(0)\n",
    "    return df_win"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8c11d68-4420-4685-91ac-cfc64a21cb1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_win_start_within(df:pd.DataFrame):\n",
    "    data_within_start = []\n",
    "    \n",
    "    for start, end, G4_id, sequence in zip(df['start'], df['end'], df['G4_id'], df['sequence']):\n",
    "        \n",
    "        number_of_win = 0\n",
    "        \n",
    "        for i, j in zip(range(start, end), range(len(sequence))):\n",
    "            row = []\n",
    "            \n",
    "            if i+WINDOW_SIZE <= end:\n",
    "                window = [*range(i, i + WINDOW_SIZE)]       \n",
    "            else:\n",
    "                break\n",
    "            \n",
    "            # print(i, window, WINDOW_SIZE, end)    \n",
    "            window_seq = ''\n",
    "            window_seq += sequence[j: j + WINDOW_SIZE]\n",
    "    \n",
    "            \n",
    "            number_of_win += 1\n",
    "    \n",
    "            row.append(G4_id)\n",
    "            row.append('within_start')\n",
    "            row.append(number_of_win)\n",
    "            row.append(window[0])\n",
    "            row.append(window[-1])\n",
    "            row.append(window_seq)\n",
    "            \n",
    "            data_within_start.append(row)\n",
    "    # print(data_down)\n",
    "    return data_within_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a446ca00-5c5b-421f-aa28-1ad665348cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_win_end_within(df:pd.DataFrame):\n",
    "    data_within_end = []\n",
    "    \n",
    "    for start, end, G4_id, sequence in zip(df['start'], df['end'], df['G4_id'], df['sequence']):\n",
    "        \n",
    "        number_of_win = 0\n",
    "        \n",
    "        for i, j in zip(range(end, start, -1), range(len(sequence))):\n",
    "            row = []\n",
    "    \n",
    "            \n",
    "            if i+WINDOW_SIZE > end:\n",
    "                continue  \n",
    "            elif (i+WINDOW_SIZE <= end):\n",
    "                window = [*range(i, i + WINDOW_SIZE)]       \n",
    "            else:\n",
    "                continue\n",
    "            # print(window)\n",
    "            \n",
    "            window_seq = ''\n",
    "            window_seq += sequence[::-1][j-10:j+WINDOW_SIZE-10][::-1]   #window_seq from the end. works with '-10' \n",
    "    \n",
    "            # print(i, j)\n",
    "            \n",
    "            number_of_win += 1\n",
    "    \n",
    "            row.append(G4_id)\n",
    "            row.append('within_end')\n",
    "            row.append(number_of_win)\n",
    "            row.append(window[0])\n",
    "            row.append(window[-1])\n",
    "            row.append(window_seq)\n",
    "            \n",
    "            data_within_end.append(row)\n",
    "    # print(data_down)\n",
    "    return data_within_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5dabd4-33f1-4262-aea3-e8b39782a8d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "386d4cc5-13e7-4243-9b73-a8ed31fed007",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collect_mut_cancer(df_cancers:pd.DataFrame, df_win:pd.DataFrame, regime:str):\n",
    "    if regime == '2col':\n",
    "        Mut_win = []\n",
    "        for window_start, window_end in zip(df_win['window_start'], df_win['window_end']):\n",
    "            row = [*range(window_start, window_end)]\n",
    "            Mut_win.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))\n",
    "    elif regime == '1col':\n",
    "        Mut_win = []\n",
    "        for row in df_win['positions']:\n",
    "            Mut_win.append(str(df_cancers[df_cancers['Pos'].isin( \\\n",
    "                list(map(int, row.strip('[]').replace(' ', '').split(','))) \\\n",
    "                                                                )].Mut.values))\n",
    "            \n",
    "    elif regime == 'pos':\n",
    "        Mut_win = []\n",
    "        for i in df_win['pos_in_ref']:\n",
    "            Mut_win.append(str(df_cancers[df_cancers['Pos'] == i].Mut.values))\n",
    "    \n",
    "    Mut_win_df = pd.DataFrame(Mut_win)\n",
    "    Mut_win_df[0] = Mut_win_df[0].str.findall(\"[A-T]>[A-T]\")\n",
    "    obsMutwin = pd.DataFrame.from_records(Mut_win_df[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)\n",
    "    df_win_cancer = df_win.join(obsMutwin)\n",
    "    df_win_cancer['number_of_Mut'] = df_win_cancer[['A>C','A>G','A>T',\n",
    "                                                    'C>A','C>G','C>T',\n",
    "                                                'G>A','G>C','G>T',\n",
    "                                                'T>A','T>C','T>G']].sum(axis=1)\n",
    "    return df_win_cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "32cd8fa2-4820-48af-ac48-35a665906fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_columns_15(df:pd.DataFrame): #for category\n",
    "    df['15_up'] = pd.concat([df[df['bp_untill_previous_G4'] < 15][['bp_untill_previous_G4', 'start']]\\\n",
    "                                      .diff(axis=1)['bp_untill_previous_G4'],\\\n",
    "                                     df[df['bp_untill_previous_G4'] >= 15]['start']-15])\n",
    "        \n",
    "    df['15_down'] = pd.concat([df[df['bp_untill_next_G4'] < 15][['end', 'bp_untill_next_G4']]\\\n",
    "                                        .diff(axis=1)['end'], \\\n",
    "                                   df[df['bp_untill_next_G4'] >= 15]['end']+15])\n",
    "    \n",
    "    df['30_up'] = pd.concat([df[df['bp_untill_previous_G4'] < 30][['bp_untill_previous_G4', 'start']]\\\n",
    "                                      .diff(axis=1)['bp_untill_previous_G4'],\\\n",
    "                                     df[df['bp_untill_previous_G4'] >= 30]['start']-30])\n",
    "        \n",
    "    df['30_down'] = pd.concat([df[df['bp_untill_next_G4'] < 30][['end', 'bp_untill_next_G4']]\\\n",
    "                                        .diff(axis=1)['end'], \\\n",
    "                                   df[df['bp_untill_next_G4'] >= 30]['end']+30])\n",
    "    \n",
    "    df['45_up'] = pd.concat([df[df['bp_untill_previous_G4'] < 45][['bp_untill_previous_G4', 'start']]\\\n",
    "                                      .diff(axis=1)['bp_untill_previous_G4'],\\\n",
    "                                     df[df['bp_untill_previous_G4'] >= 45]['start']-45])\n",
    "        \n",
    "    df['45_down'] = pd.concat([df[df['bp_untill_next_G4'] < 45][['end', 'bp_untill_next_G4']]\\\n",
    "                                        .diff(axis=1)['end'], \\\n",
    "                                   df[df['bp_untill_next_G4'] >= 45]['end']+45])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a873713-15a1-4f9e-a418-b4cc5d493e62",
   "metadata": {},
   "source": [
    "# Data preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "183b19f0-ce75-4d67-a0c0-4a871c18953a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step 1\n",
    "- auto-annotate G4 by G4Hunter (https://bioinformatics.cruk.cam.ac.uk/G4Hunter/)  \n",
    "- *verificate auto-annotate G4 in human RefSeq using data from africa paper and exel*\n",
    "- auto-annotate and auto-labeling G4 by pqsfinder (https://pqsfinder.fi.muni.cz/)\n",
    "- compare two datasets (from G4hunter and from pqsfinder)\n",
    "\n",
    "Dataset from paper [DOI:10.1093/nar/gkw006](DOI:10.1093/nar/gkw006) (G4Hunter and *in vitro* checking G4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "db421e06-a623-44fa-b428-05fdbe10799c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:79: UserWarning: Conditional Formatting extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n",
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\openpyxl\\worksheet\\_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n"
     ]
    }
   ],
   "source": [
    "datasetG4 = pd.read_excel(\"../data/raw/SupplementaryTableS2.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53e0a8a8-7591-4154-8312-7642e4063652",
   "metadata": {},
   "outputs": [],
   "source": [
    "africaG4 = datasetG4[datasetG4['Type'] != 'notG4'][['seqnames','start','end', 'width', 'strand', 'seq', 'G4Hscore', 'Type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53468540-2d03-46a3-affa-21fa082ef1ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# africaG4.to_excel('data/processed/africaG4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8d9ec877-2751-48b6-b770-e697bbfa0dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "refG4 = pd.read_table('../data/raw/NC_012920.1 Homo sapiens mitochondrion, complete genome_DNA_hl=1_k=25_G4Hseeked_2023-09-22.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "235ce597-bb87-4274-b451-8954b8fc6ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# refG4.to_excel('data/processed/refG4.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1348a1f9-3117-4868-9e94-fa00fffa65bd",
   "metadata": {},
   "source": [
    "## OLD Manually fill in the data  \n",
    "Compare two datasets africaG4, refG4  \n",
    "If G4 present in both datasets put **1** in column *confirm*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c697e0c-f104-4427-be9b-516d1b2ca539",
   "metadata": {},
   "source": [
    "## NEW pqsfinder labeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "53bafbea-0578-4971-b3b4-85a871eca5fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "refG4 = pd.read_excel('../data/processed/refG4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6a4d4b9c-6375-47b8-a219-cd83af4fda96",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>confirmed</th>\n",
       "      <th>G4</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>47</td>\n",
       "      <td>90</td>\n",
       "      <td>44</td>\n",
       "      <td>+</td>\n",
       "      <td>0.636</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>GCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTG</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16516</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.250</td>\n",
       "      <td>-2.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>185</td>\n",
       "      <td>432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>1</td>\n",
       "      <td>UG4</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>248</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>702</td>\n",
       "      <td>728</td>\n",
       "      <td>27</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.040</td>\n",
       "      <td>-1.08</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CATCCCCGTTCCAGTGAGTTCACCCTC</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>602</td>\n",
       "      <td>828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16033</td>\n",
       "      <td>16049</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>1.530</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>GGGGAAGCAGATTTGGG</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15933</td>\n",
       "      <td>16149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16067</td>\n",
       "      <td>16311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>63</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16139</td>\n",
       "      <td>16401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>1</td>\n",
       "      <td>UG4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16244</td>\n",
       "      <td>16482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16527</td>\n",
       "      <td>16560</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.853</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16427</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start    end  width  \\\n",
       "0   NC_012920.1 Homo sapiens mitochondrion, comple...     47     90     44   \n",
       "1   NC_012920.1 Homo sapiens mitochondrion, comple...    285    332     48   \n",
       "2   NC_012920.1 Homo sapiens mitochondrion, comple...    348    378     31   \n",
       "3   NC_012920.1 Homo sapiens mitochondrion, comple...    418    592    175   \n",
       "4   NC_012920.1 Homo sapiens mitochondrion, comple...    702    728     27   \n",
       "..                                                ...    ...    ...    ...   \n",
       "89  NC_012920.1 Homo sapiens mitochondrion, comple...  16033  16049     17   \n",
       "90  NC_012920.1 Homo sapiens mitochondrion, comple...  16167  16211     45   \n",
       "91  NC_012920.1 Homo sapiens mitochondrion, comple...  16239  16301     63   \n",
       "92  NC_012920.1 Homo sapiens mitochondrion, comple...  16344  16382     39   \n",
       "93  NC_012920.1 Homo sapiens mitochondrion, comple...  16527  16560     34   \n",
       "\n",
       "   strand  score  max_score  threshold  window  \\\n",
       "0       +  0.636       1.12          1      25   \n",
       "1       - -1.250      -2.24          1      25   \n",
       "2       - -1.100      -1.16          1      25   \n",
       "3       - -1.150      -1.88          1      25   \n",
       "4       - -1.040      -1.08          1      25   \n",
       "..    ...    ...        ...        ...     ...   \n",
       "89      +  1.530       1.00          1      25   \n",
       "90      - -1.160      -1.80          1      25   \n",
       "91      - -0.968      -1.40          1      25   \n",
       "92      - -1.210      -1.48          1      25   \n",
       "93      - -0.853      -1.16          1      25   \n",
       "\n",
       "                                             sequence  confirmed   G4  \\\n",
       "0        GCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTG          0    -   \n",
       "1    CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC          1   G4   \n",
       "2                     CCAAACCCCAAAAACAAAGAACCCTAACACC          1  UG4   \n",
       "3   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...          1   G4   \n",
       "4                         CATCCCCGTTCCAGTGAGTTCACCCTC          0    -   \n",
       "..                                                ...        ...  ...   \n",
       "89                                  GGGGAAGCAGATTTGGG          0    -   \n",
       "90      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC          1   G4   \n",
       "91  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...          1   G4   \n",
       "92            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC          1  UG4   \n",
       "93                 CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC          1   G4   \n",
       "\n",
       "                          note  start-100bp  end+100bp  \n",
       "0                          NaN        16516        190  \n",
       "1             more than africa          185        432  \n",
       "2             less than africa          248        478  \n",
       "3   5 overlapping g4 in africa          318        692  \n",
       "4                          NaN          602        828  \n",
       "..                         ...          ...        ...  \n",
       "89                         NaN        15933      16149  \n",
       "90            more than africa        16067      16311  \n",
       "91            more than africa        16139      16401  \n",
       "92            more than africa        16244      16482  \n",
       "93            more than africa        16427         91  \n",
       "\n",
       "[94 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refG4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a32e238d-e46f-483f-826e-f430e2220440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequenceName</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>pattern</th>\n",
       "      <th>nt</th>\n",
       "      <th>nb</th>\n",
       "      <th>nm</th>\n",
       "      <th>rl1</th>\n",
       "      <th>rl2</th>\n",
       "      <th>rl3</th>\n",
       "      <th>ll1</th>\n",
       "      <th>ll2</th>\n",
       "      <th>ll3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>34</td>\n",
       "      <td>71</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>[GGG]AGCTCTCCATGCATTT[GG(TATTTTC)G]TCT[GGG][GGG]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>-</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>51</td>\n",
       "      <td>-</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NC_012920.1_Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16393</td>\n",
       "      <td>16427</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>[CCC]TTGA[CC(ACCATC)C]T[CC(GTGAAAT)C]AATAT[CCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         sequenceName  start    end  score  \\\n",
       "0   NC_012920.1_Homo sapiens mitochondrion, comple...     34     71     30   \n",
       "1   NC_012920.1_Homo sapiens mitochondrion, comple...    295    315     64   \n",
       "2   NC_012920.1_Homo sapiens mitochondrion, comple...    353    387     33   \n",
       "3   NC_012920.1_Homo sapiens mitochondrion, comple...    458    506     55   \n",
       "4   NC_012920.1_Homo sapiens mitochondrion, comple...    534    571     88   \n",
       "..                                                ...    ...    ...    ...   \n",
       "86  NC_012920.1_Homo sapiens mitochondrion, comple...  15624  15668     29   \n",
       "87  NC_012920.1_Homo sapiens mitochondrion, comple...  16159  16193     51   \n",
       "88  NC_012920.1_Homo sapiens mitochondrion, comple...  16260  16296     44   \n",
       "89  NC_012920.1_Homo sapiens mitochondrion, comple...  16353  16380     55   \n",
       "90  NC_012920.1_Homo sapiens mitochondrion, comple...  16393  16427     21   \n",
       "\n",
       "   strand                                            pattern  nt  nb  nm  rl1  \\\n",
       "0       +   [GGG]AGCTCTCCATGCATTT[GG(TATTTTC)G]TCT[GGG][GGG]   3   1   0    3   \n",
       "1       -                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   4   1   1    5   \n",
       "2       -    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   4   2   1    4   \n",
       "3       -  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   5   2   1    6   \n",
       "4       -     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   4   0   0    4   \n",
       "..    ...                                                ...  ..  ..  ..  ...   \n",
       "86      -  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   4   3   0    9   \n",
       "87      -      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   4   1   1   11   \n",
       "88      -      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   3   0   0    3   \n",
       "89      -               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   3   0   0    3   \n",
       "90      -    [CCC]TTGA[CC(ACCATC)C]T[CC(GTGAAAT)C]AATAT[CCC]   3   2   0    3   \n",
       "\n",
       "    rl2  rl3  ll1  ll2  ll3  \n",
       "0    10    3   16    3    0  \n",
       "1     4    4    3    1    0  \n",
       "2    10    4    5    3    2  \n",
       "3     5    5    1   24    1  \n",
       "4     4    4    3   12    7  \n",
       "..  ...  ...  ...  ...  ...  \n",
       "86    8    4    7   10    2  \n",
       "87    4    4    3    8    1  \n",
       "88    3    3    3   21    1  \n",
       "89    3    3    8    8    0  \n",
       "90    9   10    4    1    5  \n",
       "\n",
       "[91 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d3fc3eb8-985e-4865-ac33-07e551ff9cdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqs['score'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "619f0593-eef8-4c16-9c59-6b2e8c95d57f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "31b6d0fe-92e4-4190-b05b-fbe9baf4b626",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_intervals = []\n",
    "for index, row, in g4pqs.iterrows():\n",
    "    for index_h, row_h in refG4.iterrows():\n",
    "        lst = []\n",
    "        for i in range(row['start'], row['end']+1):\n",
    "            if i < row_h['start']:\n",
    "                continue\n",
    "            elif i >= row_h['start']-20 and i <= row_h['end']:\n",
    "                lst = [i, row['start'], row['end'], row_h['start'], row_h['end'], \n",
    "                       row['score'], row_h['score'], row_h['G4'], row['pattern'], row_h['sequence'],\n",
    "                      row['ll1'], row['ll2'], row['ll3']] \n",
    "            # elif i >= row_h['end']+10:\n",
    "            #     lst = [i, row['start'], row['end'], row_h['start'], row_h['end'], row['score'], row_h['score'], row_h['G4']] \n",
    "            else:\n",
    "                continue\n",
    "                \n",
    "        data_intervals.append(lst)\n",
    "\n",
    "data_intervals = list(filter(None, data_intervals))                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e6b73aca-345a-4912-861e-71dd03c7af97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "6cb077e5-dd01-4335-acbb-00488f13f3cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_intervals = pd.DataFrame(data_intervals, columns=['i', 'start', 'end', 'start_h', 'end_h', 'score', 'score_h', 'G4', 'pattern', 'seq_h', 'll1', 'll2', 'll3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "5e3fafc3-29ff-48b8-9458-78c097aabe01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>score</th>\n",
       "      <th>score_h</th>\n",
       "      <th>G4</th>\n",
       "      <th>pattern</th>\n",
       "      <th>seq_h</th>\n",
       "      <th>ll1</th>\n",
       "      <th>ll2</th>\n",
       "      <th>ll3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>315</td>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>64</td>\n",
       "      <td>-1.250</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>378</td>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>33</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>506</td>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>55</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>571</td>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>88</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>984</td>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>22</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>15556</td>\n",
       "      <td>15526</td>\n",
       "      <td>15556</td>\n",
       "      <td>15502</td>\n",
       "      <td>15556</td>\n",
       "      <td>83</td>\n",
       "      <td>-1.450</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]</td>\n",
       "      <td>CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>15668</td>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>29</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>16193</td>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>51</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>G4</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>16296</td>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>44</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>16380</td>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>55</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        i  start    end  start_h  end_h  score  score_h   G4  \\\n",
       "1     315    295    315      285    332     64   -1.250   G4   \n",
       "2     378    353    387      348    378     33   -1.100  UG4   \n",
       "3     506    458    506      418    592     55   -1.150   G4   \n",
       "4     571    534    571      418    592     88   -1.150   G4   \n",
       "6     984    957   1002      954    984     22   -1.390   G4   \n",
       "..    ...    ...    ...      ...    ...    ...      ...  ...   \n",
       "80  15556  15526  15556    15502  15556     83   -1.450   G4   \n",
       "81  15668  15624  15668    15636  15683     29   -0.896  UG4   \n",
       "82  16193  16159  16193    16167  16211     51   -1.160   G4   \n",
       "83  16296  16260  16296    16239  16301     44   -0.968   G4   \n",
       "84  16380  16353  16380    16344  16382     55   -1.210  UG4   \n",
       "\n",
       "                                              pattern  \\\n",
       "1                     [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "2     [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "3   [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "4      [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "6   [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "..                                                ...   \n",
       "80          [CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]   \n",
       "81  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   \n",
       "82      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   \n",
       "83      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   \n",
       "84               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   \n",
       "\n",
       "                                                seq_h  ll1  ll2  ll3  \n",
       "1    CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC    3    1    0  \n",
       "2                     CCAAACCCCAAAAACAAAGAACCCTAACACC    5    3    2  \n",
       "3   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...    1   24    1  \n",
       "4   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...    3   12    7  \n",
       "6                     CACCCCCTCCCCAATAAAGCTAAAACTCACC    1   15   14  \n",
       "..                                                ...  ...  ...  ...  \n",
       "80  CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...    7    1    1  \n",
       "81   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC    7   10    2  \n",
       "82      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC    3    8    1  \n",
       "83  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...    3   21    1  \n",
       "84            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC    8    8    0  \n",
       "\n",
       "[70 rows x 13 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_intervals[df_intervals['G4'] != '-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "4129f36d-e3ed-4bc0-929b-160ad0682a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_intervals = []  #comparing results from G4Hunter + in_vitro and pqsfinder\n",
    "for index, row, in g4pqs.iterrows():\n",
    "    for index_h, row_h in refG4.iterrows():\n",
    "        lst = []\n",
    "        \n",
    "        s = row['start']\n",
    "        s1 = row_h['start']\n",
    "        e = row['end']\n",
    "        e1 = row_h['end']\n",
    "        \n",
    "        if e < s1:\n",
    "            continue\n",
    "        elif e1 < s:\n",
    "            continue\n",
    "        elif s < s1 and e >= s1 and e <= e1:\n",
    "            lst = [s, e, s1, e1, \n",
    "                   row['score'], row_h['score'], row_h['G4'], row['pattern'], row_h['sequence'],row['ll1'], row['ll2'], row['ll3']] \n",
    "        elif s1 < s and e1 >= s and e1 <= e:\n",
    "            lst = [s, e, s1, e1, \n",
    "                   row['score'], row_h['score'], row_h['G4'], row['pattern'], row_h['sequence'],row['ll1'], row['ll2'], row['ll3']] \n",
    "        elif s >= s1 and e <= e1:\n",
    "            lst = [s, e, s1, e1, \n",
    "                   row['score'], row_h['score'], row_h['G4'], row['pattern'], row_h['sequence'],row['ll1'], row['ll2'], row['ll3']] \n",
    "        elif s1 >= s and e1 <= e:\n",
    "            lst = [s, e, s1, e1, \n",
    "                   row['score'], row_h['score'], row_h['G4'], row['pattern'], row_h['sequence'],row['ll1'], row['ll2'], row['ll3']] \n",
    "        else:\n",
    "            continue\n",
    "                \n",
    "        data_intervals.append(lst)\n",
    "\n",
    "data_intervals = list(filter(None, data_intervals))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "17199c1c-d83f-43f5-bb8e-6db7b7708474",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interv = pd.DataFrame(data_intervals, columns=['start', 'end', 'start_h', 'end_h', 'score', 'score_h', 'G4', 'pattern', 'seq_h', 'll1', 'll2', 'll3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "42d5f024-c1f1-451d-ad15-903e00602e9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>score</th>\n",
       "      <th>score_h</th>\n",
       "      <th>G4</th>\n",
       "      <th>pattern</th>\n",
       "      <th>seq_h</th>\n",
       "      <th>ll1</th>\n",
       "      <th>ll2</th>\n",
       "      <th>ll3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>64</td>\n",
       "      <td>-1.250</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>33</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>55</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>88</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>22</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>15526</td>\n",
       "      <td>15556</td>\n",
       "      <td>15502</td>\n",
       "      <td>15556</td>\n",
       "      <td>83</td>\n",
       "      <td>-1.450</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]</td>\n",
       "      <td>CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>29</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>51</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>G4</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>44</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>G4</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>55</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>UG4</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    start    end  start_h  end_h  score  score_h   G4  \\\n",
       "1     295    315      285    332     64   -1.250   G4   \n",
       "2     353    387      348    378     33   -1.100  UG4   \n",
       "3     458    506      418    592     55   -1.150   G4   \n",
       "4     534    571      418    592     88   -1.150   G4   \n",
       "6     957   1002      954    984     22   -1.390   G4   \n",
       "..    ...    ...      ...    ...    ...      ...  ...   \n",
       "80  15526  15556    15502  15556     83   -1.450   G4   \n",
       "81  15624  15668    15636  15683     29   -0.896  UG4   \n",
       "82  16159  16193    16167  16211     51   -1.160   G4   \n",
       "83  16260  16296    16239  16301     44   -0.968   G4   \n",
       "84  16353  16380    16344  16382     55   -1.210  UG4   \n",
       "\n",
       "                                              pattern  \\\n",
       "1                     [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "2     [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "3   [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "4      [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "6   [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "..                                                ...   \n",
       "80          [CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]   \n",
       "81  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   \n",
       "82      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   \n",
       "83      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   \n",
       "84               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   \n",
       "\n",
       "                                                seq_h  ll1  ll2  ll3  \n",
       "1    CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC    3    1    0  \n",
       "2                     CCAAACCCCAAAAACAAAGAACCCTAACACC    5    3    2  \n",
       "3   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...    1   24    1  \n",
       "4   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...    3   12    7  \n",
       "6                     CACCCCCTCCCCAATAAAGCTAAAACTCACC    1   15   14  \n",
       "..                                                ...  ...  ...  ...  \n",
       "80  CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...    7    1    1  \n",
       "81   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC    7   10    2  \n",
       "82      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC    3    8    1  \n",
       "83  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...    3   21    1  \n",
       "84            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC    8    8    0  \n",
       "\n",
       "[70 rows x 12 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interv[df_interv['G4'] != '-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "c9615e5b-0e7e-457d-8489-cde9415facf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf = df_interv[df_interv['G4'] != '-'][['start','end','score','pattern', 'll1', 'll2', 'll3']].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "85c84cec-84f1-4587-bb8c-6539e3c95f35",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf = g4pqsconf.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "e64d580f-6d12-4258-aa7e-e286e4d9415e",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf['G4_id'] = range(1, g4pqsconf.shape[0]+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4b0e477e-da2a-4702-a564-7c277b6418a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCACCAAACCCCCCCtCCCCC'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf['pattern'][0].replace('[','').replace(']','').replace('(','').replace(')','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "d3601612-3835-40ca-9907-46b432a5be93",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf['sequence'] = g4pqsconf['pattern'].str.replace('[','') \\\n",
    "    .str.replace(']','')\\\n",
    "    .str.replace('(','')\\\n",
    "    .str.replace(')','')\\\n",
    "    .str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "f2b01188-1609-4f71-a24b-1a49f6970e54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>ll1</th>\n",
       "      <th>ll2</th>\n",
       "      <th>ll3</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACCAAACCCCCCCTCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>22</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>15526</td>\n",
       "      <td>15556</td>\n",
       "      <td>83</td>\n",
       "      <td>[CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>CCCCTTAAACACCCCTCCCCACATCAAGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>29</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>51</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>CATAAAAACCCAATCCACATCAAAACCCCCTCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>44</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>CCCTCACCCACTAGGATACCAACAAACCTACCCACCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>55</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>CCCTTCTCGTCCCCATGGATGACCCCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    start    end  score                                            pattern  \\\n",
       "0     295    315     64                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "1     353    387     33    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "2     458    506     55  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "3     534    571     88     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "4     957   1002     22  [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "..    ...    ...    ...                                                ...   \n",
       "62  15526  15556     83          [CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]   \n",
       "63  15624  15668     29  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   \n",
       "64  16159  16193     51      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   \n",
       "65  16260  16296     44      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   \n",
       "66  16353  16380     55               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   \n",
       "\n",
       "    ll1  ll2  ll3  G4_id                                           sequence  \n",
       "0     3    1    0      1                              CCACCAAACCCCCCCTCCCCC  \n",
       "1     5    3    2      2                CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC  \n",
       "2     1   24    1      3  CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC  \n",
       "3     3   12    7      4             CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC  \n",
       "4     1   15   14      5     CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC  \n",
       "..  ...  ...  ...    ...                                                ...  \n",
       "62    7    1    1     63                    CCCCTTAAACACCCCTCCCCACATCAAGCCC  \n",
       "63    7   10    2     64      CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC  \n",
       "64    3    8    1     65                CATAAAAACCCAATCCACATCAAAACCCCCTCCCC  \n",
       "65    3   21    1     66              CCCTCACCCACTAGGATACCAACAAACCTACCCACCC  \n",
       "66    8    8    0     67                       CCCTTCTCGTCCCCATGGATGACCCCCC  \n",
       "\n",
       "[67 rows x 9 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca02407e-4b79-4dde-bce7-255336051240",
   "metadata": {},
   "source": [
    "### Statistic for loops in G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "e827d850-c650-47fb-9546-35e6f6f1688c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\seaborn\\_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\seaborn\\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x1e0bfa90b20>"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeoAAAIRCAYAAACF0P59AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAxZUlEQVR4nO3de3RU5b3G8WeEMCSQhGsyieQmjVxEQIUCUUoQEwXFS7RK8QKtVJCLRnSBEY8MVhOlixQVxINVxKWobQ9YKopEENQCngAiEJFiDYJKjAlIAglB4D1/2MxhnHBJSGa/kO9nrb3Kfvee/f5m910+2bfZLmOMEQAAsNI5ThcAAACOj6AGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIyghqNeeukluVwu39S8eXN5PB4NHDhQOTk5Ki4uDviM1+uVy+WqVT8VFRXyer1auXJlrT5XU1+JiYm65pprarWdk1mwYIFmzpxZ4zKXyyWv11uv/dW35cuXq1evXmrRooVcLpfefPPNGtfbsWOHXC6XXnrppaDWdzzffvutvF6vNm7cGLBs5MiRatmyZZ23nZqaqtTU1LoXB/wHQQ0rzJs3T2vWrFFeXp5mz56tnj176sknn1SXLl303nvv+a07atQorVmzplbbr6io0LRp02od1HXpqy5OFNRr1qzRqFGjGryGujLG6Oabb1ZISIgWL16sNWvWaMCAAU6XdUq+/fZbTZs2rcagBmzR1OkCAEnq1q2bevXq5Zu/8cYbdd999+myyy5TRkaGtm/frujoaElShw4d1KFDhwatp6KiQmFhYUHp62T69u3raP8n8+2332rPnj264YYbNGjQIKfLAc46HFHDWvHx8ZoxY4bKy8v13//93772mk5Hr1ixQqmpqWrbtq1CQ0MVHx+vG2+8URUVFdqxY4fat28vSZo2bZrvNPvIkSP9trdhwwbddNNNat26tTp27HjcvqotWrRI3bt3V/PmzXXeeefp6aef9ltefVp/x44dfu0rV66Uy+XyHd2npqZqyZIl+uqrr/wuA1Sr6dT3li1bdN1116l169Zq3ry5evbsqfnz59fYz2uvvaYpU6YoNjZWERERuuKKK7Rt27bj7/hjfPTRRxo0aJDCw8MVFhamlJQULVmyxLfc6/X6/pCZPHmyXC6XEhMTT2nbx9q+fbuGDx+uqKgoud1udenSRbNnz67z9zHGKDs7WwkJCWrevLl69eqlvLw8v9PRK1euVO/evSVJv/3tb337/ef7+osvvtCQIUPUsmVLxcXF6f7771dVVVWtv6Mk7dmzR2PHjtW5556rZs2a6bzzztOUKVMCtnfw4EFlZWUpKSlJzZo107nnnqtx48bphx9+8Fuv+jLMycYizmwcUcNqQ4YMUZMmTfTBBx8cd50dO3bo6quvVv/+/fXiiy+qVatW+uabb7R06VIdOnRIMTExWrp0qa666irdeeedvtPI1eFdLSMjQ8OGDdOYMWN04MCBE9a1ceNGZWZmyuv1yuPx6NVXX9W9996rQ4cO6YEHHqjVd3z22Wd111136d///rcWLVp00vW3bdumlJQURUVF6emnn1bbtm31yiuvaOTIkfruu+80adIkv/UfeughXXrppfrzn/+ssrIyTZ48WUOHDtXWrVvVpEmT4/azatUqpaWlqXv37nrhhRfkdrv17LPPaujQoXrttdd0yy23aNSoUerRo4cyMjI0YcIEDR8+XG63u1bf/7PPPlNKSorvDzOPx6N3331X99xzj0pKSjR16tRaf58pU6YoJydHd911lzIyMrRr1y6NGjVKP/74o84//3xJ0sUXX6x58+bpt7/9rR5++GFdffXVkuR3BuXHH3/UtddeqzvvvFP333+/PvjgA/3hD39QZGSkHnnkkVp9z4MHD2rgwIH697//rWnTpql79+768MMPlZOTo40bN/r+ADLG6Prrr9fy5cuVlZWl/v37a9OmTZo6darWrFmjNWvW+O3j+hyLsJQBHDRv3jwjyeTn5x93nejoaNOlSxff/NSpU82xQ/dvf/ubkWQ2btx43G18//33RpKZOnVqwLLq7T3yyCPHXXashIQE43K5AvpLS0szERER5sCBA37frbCw0G+9999/30gy77//vq/t6quvNgkJCTXW/vO6hw0bZtxut9m5c6ffeoMHDzZhYWHmhx9+8OtnyJAhfuv95S9/MZLMmjVrauyvWt++fU1UVJQpLy/3tR0+fNh069bNdOjQwRw9etQYY0xhYaGRZP74xz+ecHvHrjtv3jxf25VXXmk6dOhg9u3b57fu+PHjTfPmzc2ePXtq9X327Nlj3G63ueWWW/zWW7NmjZFkBgwY4GvLz88PqKfaiBEjjCTzl7/8xa99yJAhplOnTif9rgMGDPDr67nnnqtxe08++aSRZJYtW2aMMWbp0qVGkpk+fbrfem+88YaRZObOnetrO9WxiDMbp75hPXOSV6b37NlTzZo101133aX58+fryy+/rFM/N9544ymve8EFF6hHjx5+bcOHD1dZWZk2bNhQp/5P1YoVKzRo0CDFxcX5tY8cOVIVFRUBN79de+21fvPdu3eXJH311VfH7ePAgQP6+OOPddNNN/nd+dykSRPdfvvt+vrrr0/59PmJHDx4UMuXL9cNN9ygsLAwHT582DcNGTJEBw8e1Nq1a2v1fdauXauqqirdfPPNfuv17du31qflXS6Xhg4dGtDfifbd8axYsUItWrTQTTfd5NdefQlm+fLlvvWOba/261//Wi1atPCtV83JsYjgIKhhtQMHDqi0tFSxsbHHXadjx4567733FBUVpXHjxqljx47q2LGjnnrqqVr1FRMTc8rrejye47aVlpbWqt/aKi0trbHW6n308/7btm3rN1992rSysvK4fezdu1fGmFr1UxelpaU6fPiwnnnmGYWEhPhNQ4YMkSSVlJT4feZk36e6ruqbD49VU9uJhIWFqXnz5gH9HTx4sFbbqa7L4/EE3PMQFRWlpk2b+uouLS1V06ZNAy7NuFwueTyegP3u5FhEcHCNGlZbsmSJjhw5ctLnUfv376/+/fvryJEjWrdunZ555hllZmYqOjpaw4YNO6W+avNsdlFR0XHbqoOk+j/wP79R6OfBU1tt27bV7t27A9q//fZbSVK7du1Oa/uS1Lp1a51zzjlB6af6KH3cuHE1rpOUlFSrbVbv/++++y5gWVFRUZ1udqsPbdu21ccffyxjjN9YKy4u1uHDh337s23btjp8+LC+//57v7A2xqioqMh3A1y1UxmLOLNxRA1r7dy5Uw888IAiIyM1evToU/pMkyZN1KdPH98dw9Wn/k7lKLI2CgoK9Omnn/q1LViwQOHh4br44oslyRcImzZt8ltv8eLFAdtzu92nXNugQYO0YsUKX2BWe/nllxUWFlYvj3O1aNFCffr00cKFC/3qOnr0qF555RV16NDBd1PW6QgLC9PAgQP1ySefqHv37urVq1fAVNuw6dOnj9xut9544w2/9rVr1wacsq7vcXEigwYN0v79+wN+DObll1/2LT/2f1955RW/9f7nf/5HBw4cCHgE7lTGIs5sHFHDClu2bPFdmywuLtaHH36oefPmqUmTJlq0aFHAacBjPffcc1qxYoWuvvpqxcfH6+DBg3rxxRclSVdccYUkKTw8XAkJCfr73/+uQYMGqU2bNmrXrl2dj65iY2N17bXXyuv1KiYmRq+88ory8vL05JNPKiwsTJLUu3dvderUSQ888IAOHz6s1q1ba9GiRfroo48CtnfhhRdq4cKFmjNnji655BKdc845fs+VH2vq1Kl66623NHDgQD3yyCNq06aNXn31VS1ZskTTp09XZGRknb7Tz+Xk5CgtLU0DBw7UAw88oGbNmunZZ5/Vli1b9Nprr9X61+GO56mnntJll12m/v376+6771ZiYqLKy8v1xRdf6B//+Ifvmu2patOmjSZOnKicnBy1bt1aN9xwg77++mtNmzZNMTExOuec/z8+6dixo0JDQ/Xqq6+qS5cuatmypWJjY094qaWu7rjjDs2ePVsjRozQjh07dOGFF+qjjz5Sdna2hgwZ4huraWlpuvLKKzV58mSVlZXp0ksv9d31fdFFF+n222/32+6pjEWc4Zy9lw2NXfWd0dVTs2bNTFRUlBkwYIDJzs42xcXFAZ/5+Z3Ya9asMTfccINJSEgwbrfbtG3b1gwYMMAsXrzY73Pvvfeeueiii4zb7TaSzIgRI/y29/3335+0L2N+utP26quvNn/729/MBRdcYJo1a2YSExNNbm5uwOf/9a9/mfT0dBMREWHat29vJkyYYJYsWRJw1/eePXvMTTfdZFq1amVcLpdfn6rhbvXNmzeboUOHmsjISNOsWTPTo0ePgDuXq++S/utf/+rXXtOd18fz4Ycfmssvv9y0aNHChIaGmr59+5p//OMfNW6vrnd9V7f/7ne/M+eee64JCQkx7du3NykpKeaxxx6r0/c5evSoeeyxx0yHDh1Ms2bNTPfu3c1bb71levToYW644Qa/z7/22mumc+fOJiQkxG9fjxgxwrRo0SLgO9Q0Jmry87u+jTGmtLTUjBkzxsTExJimTZuahIQEk5WVZQ4ePOi3XmVlpZk8ebJJSEgwISEhJiYmxtx9991m7969fuvVZizizOUy5iS31ALAWaCwsFCdO3fW1KlT9dBDDzldTr1ITExUt27d9NZbbzldChoQp74BnHU+/fRTvfbaa0pJSVFERIS2bdum6dOnKyIiQnfeeafT5QG1QlADOOu0aNFC69at0wsvvKAffvhBkZGRSk1N1eOPP17rR7QAp3HqGwAAi/F4FgAAFiOoAQCw2Fkf1MYYlZWVnfT3ogEAsNFZH9Tl5eWKjIxUeXm506UAAFBrZ31QAwBwJiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALNbU6QLONDt37lRJSUnQ+mvXrp3i4+OD1h8AwC4EdS3s3LlTnTt3UWVlRdD6DA0N0+efbyWsAaCRIqhroaSkRJWVFerzu6mKiEls8P7Kdu/Qxy9OU0lJCUENAI0UQV0HETGJahPfyekyAACNADeTAQBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFHA/qb775Rrfddpvatm2rsLAw9ezZU+vXr/ctN8bI6/UqNjZWoaGhSk1NVUFBgYMVAwAQPI4G9d69e3XppZcqJCRE77zzjj777DPNmDFDrVq18q0zffp05ebmatasWcrPz5fH41FaWprKy8udKxwAgCBx9JfJnnzyScXFxWnevHm+tsTERN+/jTGaOXOmpkyZooyMDEnS/PnzFR0drQULFmj06NEB26yqqlJVVZVvvqysrOG+AAAADczRI+rFixerV69e+vWvf62oqChddNFFev75533LCwsLVVRUpPT0dF+b2+3WgAEDtHr16hq3mZOTo8jISN8UFxfX4N8DAICG4mhQf/nll5ozZ46Sk5P17rvvasyYMbrnnnv08ssvS5KKiookSdHR0X6fi46O9i37uaysLO3bt8837dq1q2G/BAAADcjRU99Hjx5Vr169lJ2dLUm66KKLVFBQoDlz5uiOO+7wredyufw+Z4wJaKvmdrvldrsbrmgAAILI0SPqmJgYde3a1a+tS5cu2rlzpyTJ4/FIUsDRc3FxccBRNgAAZyNHg/rSSy/Vtm3b/Nr+9a9/KSEhQZKUlJQkj8ejvLw83/JDhw5p1apVSklJCWqtAAA4wdFT3/fdd59SUlKUnZ2tm2++Wf/7v/+ruXPnau7cuZJ+OuWdmZmp7OxsJScnKzk5WdnZ2QoLC9Pw4cOdLB0AgKBwNKh79+6tRYsWKSsrS48++qiSkpI0c+ZM3Xrrrb51Jk2apMrKSo0dO1Z79+5Vnz59tGzZMoWHhztYOQAAweFoUEvSNddco2uuuea4y10ul7xer7xeb/CKAgDAEo7/hCgAADg+ghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMUeD2uv1yuVy+U0ej8e33Bgjr9er2NhYhYaGKjU1VQUFBQ5WDABAcDl+RH3BBRdo9+7dvmnz5s2+ZdOnT1dubq5mzZql/Px8eTwepaWlqby83MGKAQAIHseDumnTpvJ4PL6pffv2kn46mp45c6amTJmijIwMdevWTfPnz1dFRYUWLFjgcNUAAASH40G9fft2xcbGKikpScOGDdOXX34pSSosLFRRUZHS09N967rdbg0YMECrV68+7vaqqqpUVlbmNwEAcKZyNKj79Omjl19+We+++66ef/55FRUVKSUlRaWlpSoqKpIkRUdH+30mOjrat6wmOTk5ioyM9E1xcXEN+h0AAGhIjgb14MGDdeONN+rCCy/UFVdcoSVLlkiS5s+f71vH5XL5fcYYE9B2rKysLO3bt8837dq1q2GKBwAgCBw/9X2sFi1a6MILL9T27dt9d3///Oi5uLg44Cj7WG63WxEREX4TAABnKquCuqqqSlu3blVMTIySkpLk8XiUl5fnW37o0CGtWrVKKSkpDlYJAEDwNHWy8wceeEBDhw5VfHy8iouL9dhjj6msrEwjRoyQy+VSZmamsrOzlZycrOTkZGVnZyssLEzDhw93smwAAILG0aD++uuv9Zvf/EYlJSVq3769+vbtq7Vr1yohIUGSNGnSJFVWVmrs2LHau3ev+vTpo2XLlik8PNzJsgEACBpHg/r1118/4XKXyyWv1yuv1xucggAAsIxV16gBAIA/ghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLWRPUOTk5crlcyszM9LUZY+T1ehUbG6vQ0FClpqaqoKDAuSIBAAgyK4I6Pz9fc+fOVffu3f3ap0+frtzcXM2aNUv5+fnyeDxKS0tTeXm5Q5UCABBcjgf1/v37deutt+r5559X69atfe3GGM2cOVNTpkxRRkaGunXrpvnz56uiokILFixwsGIAAIKnqdMFjBs3TldffbWuuOIKPfbYY772wsJCFRUVKT093dfmdrs1YMAArV69WqNHj65xe1VVVaqqqvLNl5WVNVzxQbJ169ag9NOuXTvFx8cHpS8AwKlxNKhff/11bdiwQfn5+QHLioqKJEnR0dF+7dHR0frqq6+Ou82cnBxNmzatfgt1SOW+Ukku3XbbbUHpLzQ0TJ9/vpWwBgCLOBbUu3bt0r333qtly5apefPmx13P5XL5zRtjAtqOlZWVpYkTJ/rmy8rKFBcXd/oFO+DHinJJRj2HT1b7pM4N2lfZ7h36+MVpKikpIagBwCKOBfX69etVXFysSy65xNd25MgRffDBB5o1a5a2bdsm6acj65iYGN86xcXFAUfZx3K73XK73Q1XuANaRsWrTXwnp8sAADjAsZvJBg0apM2bN2vjxo2+qVevXrr11lu1ceNGnXfeefJ4PMrLy/N95tChQ1q1apVSUlKcKhsAgKBy7Ig6PDxc3bp182tr0aKF2rZt62vPzMxUdna2kpOTlZycrOzsbIWFhWn48OFOlAwAQNA5ftf3iUyaNEmVlZUaO3as9u7dqz59+mjZsmUKDw93ujQAAILCqqBeuXKl37zL5ZLX65XX63WkHgAAnOb4D54AAIDjI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALBYnYL6vPPOU2lpaUD7Dz/8oPPOO++0iwIAAD+pU1Dv2LFDR44cCWivqqrSN998c9pFAQCAnzStzcqLFy/2/fvdd99VZGSkb/7IkSNavny5EhMT6604AAAau1oF9fXXXy9JcrlcGjFihN+ykJAQJSYmasaMGfVWHAAAjV2tgvro0aOSpKSkJOXn56tdu3YNUhQAAPhJrYK6WmFhYX3XAQAAalCnoJak5cuXa/ny5SouLvYdaVd78cUXT7swAABQx6CeNm2aHn30UfXq1UsxMTFyuVz1XRcAAFAdg/q5557TSy+9pNtvv72+6wEAAMeo03PUhw4dUkpKSn3XAgAAfqZOQT1q1CgtWLCgvmsBAAA/U6dT3wcPHtTcuXP13nvvqXv37goJCfFbnpubWy/FAQDQ2NUpqDdt2qSePXtKkrZs2eK3jBvLAACoP3UK6vfff7++6wAAADXgNZcAAFisTkfUAwcOPOEp7hUrVtS5IAAA8P/qFNTV16er/fjjj9q4caO2bNkS8LIOAABQd3UK6j/96U81tnu9Xu3fv/+0CgIAAP+vXq9R33bbbfzONwAA9aheg3rNmjVq3rx5fW4SAIBGrU6nvjMyMvzmjTHavXu31q1bp//6r/+ql8IAAEAdgzoyMtJv/pxzzlGnTp306KOPKj09vV4KAwAAdQzqefPm1XcdAACgBnUK6mrr16/X1q1b5XK51LVrV1100UX1VRcAAFAdg7q4uFjDhg3TypUr1apVKxljtG/fPg0cOFCvv/662rdvX991AgDQKNXpru8JEyaorKxMBQUF2rNnj/bu3astW7aorKxM99xzT33XCABAo1WnI+qlS5fqvffeU5cuXXxtXbt21ezZs7mZDACAelSnI+qjR48GvINakkJCQnT06NHTLgoAAPykTkF9+eWX695779W3337ra/vmm2903333adCgQfVWHAAAjV2dgnrWrFkqLy9XYmKiOnbsqF/84hdKSkpSeXm5nnnmmfquEQCARqtO16jj4uK0YcMG5eXl6fPPP5cxRl27dtUVV1xR3/UBANCo1eqIesWKFeratavKysokSWlpaZowYYLuuece9e7dWxdccIE+/PDDU97enDlz1L17d0VERCgiIkL9+vXTO++841tujJHX61VsbKxCQ0OVmpqqgoKC2pQMAMAZrVZBPXPmTP3+979XREREwLLIyEiNHj1aubm5p7y9Dh066IknntC6deu0bt06XX755bruuut8YTx9+nTl5uZq1qxZys/Pl8fjUVpamsrLy2tTNgAAZ6xaBfWnn36qq6666rjL09PTtX79+lPe3tChQzVkyBCdf/75Ov/88/X444+rZcuWWrt2rYwxmjlzpqZMmaKMjAx169ZN8+fPV0VFhRYsWFCbsgEAOGPVKqi/++67Gh/Lqta0aVN9//33dSrkyJEjev3113XgwAH169dPhYWFKioq8nsu2+12a8CAAVq9evVxt1NVVaWysjK/CQCAM1Wtgvrcc8/V5s2bj7t806ZNiomJqVUBmzdvVsuWLeV2uzVmzBgtWrRIXbt2VVFRkSQpOjrab/3o6Gjfsprk5OQoMjLSN8XFxdWqHgAAbFKroB4yZIgeeeQRHTx4MGBZZWWlpk6dqmuuuaZWBXTq1EkbN27U2rVrdffdd2vEiBH67LPPfMtdLpff+saYgLZjZWVlad++fb5p165dtaoHAACb1OrxrIcfflgLFy7U+eefr/Hjx6tTp05yuVzaunWrZs+erSNHjmjKlCm1KqBZs2b6xS9+IUnq1auX8vPz9dRTT2ny5MmSpKKiIr+j9OLi4oCj7GO53W653e5a1QAAgK1qFdTR0dFavXq17r77bmVlZckYI+mno94rr7xSzz777AlD9FQYY1RVVaWkpCR5PB7l5eX5Xp956NAhrVq1Sk8++eRp9QEAwJmi1j94kpCQoLffflt79+7VF198IWOMkpOT1bp161p3/tBDD2nw4MGKi4tTeXm5Xn/9da1cuVJLly6Vy+VSZmamsrOzlZycrOTkZGVnZyssLEzDhw+vdV8AAJyJ6vTLZJLUunVr9e7d+7Q6/+6773T77bdr9+7dioyMVPfu3bV06VKlpaVJkiZNmqTKykqNHTtWe/fuVZ8+fbRs2TKFh4efVr8AAJwp6hzU9eGFF1444XKXyyWv1yuv1xucggAAsEydXsoBAACCg6AGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFmvqdAGwy9atW4PWV7t27RQfHx+0/gDgTERQQ5JUua9Ukku33XZb0PoMDQ3T559vJawB4AQIakiSfqwol2TUc/hktU/q3OD9le3eoY9fnKaSkhKCGgBOgKCGn5ZR8WoT38npMgAA/8HNZAAAWIygBgDAYgQ1AAAW4xo1HBWsx8F4FAzAmYqghiOC/TgYj4IBOFMR1HBEMB8H41EwAGcyghqO4nEwADgxbiYDAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxRwN6pycHPXu3Vvh4eGKiorS9ddfr23btvmtY4yR1+tVbGysQkNDlZqaqoKCAocqBgAguBwN6lWrVmncuHFau3at8vLydPjwYaWnp+vAgQO+daZPn67c3FzNmjVL+fn58ng8SktLU3l5uYOVAwAQHE2d7Hzp0qV+8/PmzVNUVJTWr1+vX/3qVzLGaObMmZoyZYoyMjIkSfPnz1d0dLQWLFig0aNHO1E2AABBY9U16n379kmS2rRpI0kqLCxUUVGR0tPTfeu43W4NGDBAq1evrnEbVVVVKisr85sAADhTWRPUxhhNnDhRl112mbp16yZJKioqkiRFR0f7rRsdHe1b9nM5OTmKjIz0TXFxcQ1bOAAADciaoB4/frw2bdqk1157LWCZy+XymzfGBLRVy8rK0r59+3zTrl27GqReAACCwdFr1NUmTJigxYsX64MPPlCHDh187R6PR9JPR9YxMTG+9uLi4oCj7Gput1tut7thCwYAIEgcPaI2xmj8+PFauHChVqxYoaSkJL/lSUlJ8ng8ysvL87UdOnRIq1atUkpKSrDLBQAg6Bw9oh43bpwWLFigv//97woPD/ddd46MjFRoaKhcLpcyMzOVnZ2t5ORkJScnKzs7W2FhYRo+fLiTpQMAEBSOBvWcOXMkSampqX7t8+bN08iRIyVJkyZNUmVlpcaOHau9e/eqT58+WrZsmcLDw4NcLQAAwedoUBtjTrqOy+WS1+uV1+tt+IIAALCMNXd9AwCAQAQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWa+p0AUCwbN26NWh9tWvXTvHx8UHrD8DZi6DGWa9yX6kkl2677bag9RkaGqbPP99KWAM4bQQ1zno/VpRLMuo5fLLaJ3Vu8P7Kdu/Qxy9OU0lJCUEN4LQR1Gg0WkbFq018J6fLAIBa4WYyAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDFCGoAACxGUAMAYDGCGgAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWMzRoP7ggw80dOhQxcbGyuVy6c033/RbboyR1+tVbGysQkNDlZqaqoKCAmeKBQDAAY4G9YEDB9SjRw/NmjWrxuXTp09Xbm6uZs2apfz8fHk8HqWlpam8vDzIlQIA4IymTnY+ePBgDR48uMZlxhjNnDlTU6ZMUUZGhiRp/vz5io6O1oIFCzR69OgaP1dVVaWqqirffFlZWf0XDgBAkFh7jbqwsFBFRUVKT0/3tbndbg0YMECrV68+7udycnIUGRnpm+Li4oJRLgAADcLaoC4qKpIkRUdH+7VHR0f7ltUkKytL+/bt8027du1q0DoBAGhIjp76PhUul8tv3hgT0HYst9stt9vd0GUBABAU1h5RezweSQo4ei4uLg44ygYA4GxlbVAnJSXJ4/EoLy/P13bo0CGtWrVKKSkpDlYGAEDwOHrqe//+/friiy9884WFhdq4caPatGmj+Ph4ZWZmKjs7W8nJyUpOTlZ2drbCwsI0fPhwB6sGACB4HA3qdevWaeDAgb75iRMnSpJGjBihl156SZMmTVJlZaXGjh2rvXv3qk+fPlq2bJnCw8OdKhk4ZVu3bg1KP+3atVN8fHxQ+gIQfI4GdWpqqowxx13ucrnk9Xrl9XqDVxRwmir3lUpy6bbbbgtKf6GhYfr8862ENXCWsv6ub+BM82NFuSSjnsMnq31S5wbtq2z3Dn384jSVlJQQ1MBZiqAGGkjLqHi1ie/kdBkAznDW3vUNAAAIagAArEZQAwBgMa5RA2eBYD0KJvE4GBBsBDVwBgv2o2ASj4MBwUZQA2ewYD4KJvE4GOAEgho4C/AoGHD24mYyAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMUIagAALEZQAwBgMX5CFECtBettXbypCyCoAdRCsN/WxZu6AIIaQC0E821dvKkL+AlBDaDWeFsXEDzcTAYAgMUIagAALEZQAwBgMYIaAACLEdQAAFiMoAYAwGIENQAAFiOoAQCwGEENAIDF+GUyAFbjBSBo7AhqAFbiBSDATwhqAFbiBSDATwhqAFbjBSBo7LiZDAAAixHUAABYjKAGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIyfEAWA/wjWm7okqaqqSm63Oyh98WawMxtBDaDRC/abuiRJLpdkTFC64s1gZzaCGkCjF8w3dUnS7s1rtGXxXN4MhlNCUAPAfwTrTV1lu3cEtT+c2biZDAAAixHUAABYjKAGAMBiXKMGANSrnTt3qqSkJCh9BfMxN8mZR90IagBAvdm5c6c6d+6iysqK4HQYxMfcJGcedSOoAQD1pqSkRJWVFerzu6mKiEls0L6C+Zib5NyjbgQ1AKDeRcQkNvijZ43lMbcz4mayZ599VklJSWrevLkuueQSffjhh06XBABAUFgf1G+88YYyMzM1ZcoUffLJJ+rfv78GDx6snTt3Ol0aAAANzvqgzs3N1Z133qlRo0apS5cumjlzpuLi4jRnzhynSwMAoMFZfY360KFDWr9+vR588EG/9vT0dK1evbrGz1RVVamqqso3v2/fPklSWVnZadezf/9+SdKer7bpcFXlaW/vZMp2fyVJ2vfNdoU0dZ01fQW7P77bmdnf2dpXsPsrK/rp7OP69et9/w1rSNu2bZMUnP9OBv3/t//sy/3799dLpkhSeHi4XK6T1G4s9s033xhJ5p///Kdf++OPP27OP//8Gj8zdepUI4mJiYmJicn6ad++fSfNQquPqKv9/K8NY8xx/wLJysrSxIkTffNHjx7Vnj171LZt25P/1XISZWVliouL065duxQREXFa2zpbsE8CsU/8sT8CsU8CNdZ9Eh4eftJ1rA7qdu3aqUmTJioqKvJrLy4uVnR0dI2fcbvdAb9S06pVq3qtKyIiolENpFPBPgnEPvHH/gjEPgnEPglk9c1kzZo10yWXXKK8vDy/9ry8PKWkpDhUFQAAwWP1EbUkTZw4Ubfffrt69eqlfv36ae7cudq5c6fGjBnjdGkAADQ464P6lltuUWlpqR599FHt3r1b3bp109tvv62EhISg1+J2uzV16tSg/gC87dgngdgn/tgfgdgngdgnx+cyJoi/Zg4AAGrF6mvUAAA0dgQ1AAAWI6gBALAYQQ0AgMUI6lrgdZv/z+v1yuVy+U0ej8fpsoLmgw8+0NChQxUbGyuXy6U333zTb7kxRl6vV7GxsQoNDVVqaqoKCgqcKTZITrZPRo4cGTBm+vbt60yxQZCTk6PevXsrPDxcUVFRuv76632/g12tsY2TU9knjW2cnAqC+hTxus1AF1xwgXbv3u2bNm/e7HRJQXPgwAH16NFDs2bNqnH59OnTlZubq1mzZik/P18ej0dpaWkqLy8PcqXBc7J9IklXXXWV35h5++23g1hhcK1atUrjxo3T2rVrlZeXp8OHDys9PV0HDhzwrdPYxsmp7BOpcY2TU3K6L85oLH75y1+aMWPG+LV17tzZPPjggw5V5KypU6eaHj16OF2GFSSZRYsW+eaPHj1qPB6PeeKJJ3xtBw8eNJGRkea5555zoMLg+/k+McaYESNGmOuuu86RemxQXFxsJJlVq1YZYxgnxgTuE2MYJzXhiPoUVL9uMz093a/9RK/bbAy2b9+u2NhYJSUladiwYfryyy+dLskKhYWFKioq8hsvbrdbAwYMaNTjRZJWrlypqKgonX/++fr973+v4uJip0sKmupX7rZp00YS40QK3CfVGvM4qQlBfQpKSkp05MiRgBeBREdHB7wwpLHo06ePXn75Zb377rt6/vnnVVRUpJSUFJWWljpdmuOqxwTjxd/gwYP16quvasWKFZoxY4by8/N1+eWX+70//mxljNHEiRN12WWXqVu3bpIYJzXtE6lxj5Pjsf4nRG1Sm9dtnu0GDx7s+/eFF16ofv36qWPHjpo/f77fa0YbM8aLv1tuucX3727duqlXr15KSEjQkiVLlJGR4WBlDW/8+PHatGmTPvroo4BljXWcHG+fNOZxcjwcUZ+Curxus7Fp0aKFLrzwQm3fvt3pUhxXffc74+XEYmJilJCQcNaPmQkTJmjx4sV6//331aFDB197Yx4nx9snNWks4+RECOpTwOs2T66qqkpbt25VTEyM06U4LikpSR6Px2+8HDp0SKtWrWK8HKO0tFS7du06a8eMMUbjx4/XwoULtWLFCiUlJfktb4zj5GT7pCZn+zg5JQ7eyHZGef31101ISIh54YUXzGeffWYyMzNNixYtzI4dO5wuzRH333+/Wblypfnyyy/N2rVrzTXXXGPCw8Mbzf4oLy83n3zyifnkk0+MJJObm2s++eQT89VXXxljjHniiSdMZGSkWbhwodm8ebP5zW9+Y2JiYkxZWZnDlTecE+2T8vJyc//995vVq1ebwsJC8/7775t+/fqZc88996zdJ3fffbeJjIw0K1euNLt37/ZNFRUVvnUa2zg52T5pjOPkVBDUtTB79myTkJBgmjVrZi6++GK/Rwoam1tuucXExMSYkJAQExsbazIyMkxBQYHTZQXN+++/byQFTCNGjDDG/PTozdSpU43H4zFut9v86le/Mps3b3a26AZ2on1SUVFh0tPTTfv27U1ISIiJj483I0aMMDt37nS67AZT076QZObNm+dbp7GNk5Ptk8Y4Tk4Fr7kEAMBiXKMGAMBiBDUAABYjqAEAsBhBDQCAxQhqAAAsRlADAGAxghoAAIsR1AAAWIygBuAnNTVVmZmZTpcB4D8IagAALEZQAwBgMYIaQIDDhw9r/PjxatWqldq2bauHH35Y1a8FSExM1B/+8AcNHz5cLVu2VGxsrJ555hmHKwbOXgQ1gADz589X06ZN9fHHH+vpp5/Wn/70J/35z3/2Lf/jH/+o7t27a8OGDcrKytJ9990X8L52APWDt2cB8JOamqri4mIVFBTI5XJJkh588EEtXrxYn332mRITE9WlSxe98847vs8MGzZMZWVlevvtt50qGzhrcUQNIEDfvn19IS1J/fr10/bt23XkyBHf/LH69eunrVu3BrVGoLEgqAHUi2ODHUD9IagBBFi7dm3AfHJyspo0aXLc5Z07dw5afUBj0tTpAgDYZ9euXZo4caJGjx6tDRs26JlnntGMGTN8y//5z39q+vTpuv7665WXl6e//vWvWrJkiYMVA2cvghpAgDvuuEOVlZX65S9/qSZNmmjChAm66667fMvvv/9+rV+/XtOmTVN4eLhmzJihK6+80sGKgbMXd30DqJXExERlZmbyM6NAkHCNGgAAixHUAABYjFPfAABYjCNqAAAsRlADAGAxghoAAIsR1AAAWIygBgDAYgQ1AAAWI6gBALAYQQ0AgMX+D8qydx6VccDXAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 500x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(pd.concat([g4pqsconf['ll1'],g4pqsconf['ll2'],g4pqsconf['ll3']])) \\\n",
    "    .set(title='Distribution of length loop', xlabel='bp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "b87df288-b7e0-45c3-871e-e21560ad5924",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf['llall'] = g4pqsconf['ll1'] + g4pqsconf['ll2'] + g4pqsconf['ll3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "0f924298-03b4-4f79-8c11-ba49bf9e6b5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14.0"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf['llall'].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "82ae7b2e-60b3-4205-a512-bbcc6782a028",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    67.000000\n",
       "mean     15.089552\n",
       "std       7.879262\n",
       "min       3.000000\n",
       "25%      10.000000\n",
       "50%      14.000000\n",
       "75%      21.000000\n",
       "max      32.000000\n",
       "Name: llall, dtype: float64"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf['llall'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "8d975bf4-5874-4bee-a038-53d7dd56be8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\seaborn\\_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "C:\\Users\\777\\anaconda3\\lib\\site-packages\\seaborn\\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x1e0c4e7d040>"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeoAAAH+CAYAAABTKk23AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAt3UlEQVR4nO3de1zUdb7H8feIMIoCiYpAcsvyft00L+mKopaaXaxTrmth1m61ail1TpGV0EXSjmZl6rE2s0fptm1alt3Ma23Y8ZqXzPSRSlu6pBaDoqjwPX94mBoBFRzm90Vez8djHo/mN7+Z+czPb76cC4zLGGMEAACsVMvpAQAAQPkINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINSrk1Vdflcvl8p7q1Kmj6Oho9enTR1lZWcrNzS11nYyMDLlcrgrdT0FBgTIyMrRy5coKXa+s+0pMTNQ111xTods5m/nz52v69OllXuZyuZSRkeHX+/O3ZcuWqXPnzqpXr55cLpfeeeedMvfbs2ePXC6XXn311YDOV54ff/xRGRkZ2rRpU6nLRo4cqfr16wd+qACqDmsL/keoUSlz585Vdna2li5dqhdffFEdO3bU5MmT1apVK3366ac++955553Kzs6u0O0XFBQoMzOzwqGuzH1VxplCnZ2drTvvvLPKZ6gsY4xuvvlmBQcHa/HixcrOzlbv3r2dHuuc/Pjjj8rMzCwz1MCFqrbTA6B6atu2rTp37uw9f+ONN2r8+PHq2bOnhg4dqp07d6pJkyaSpKZNm6pp06ZVOk9BQYFCQ0MDcl9n061bN0fv/2x+/PFHHTp0SDfccINSUlKcHgfAWfCMGn4THx+vqVOnKj8/X//zP//j3V7Wy9HLly9XcnKyGjZsqLp16yo+Pl433nijCgoKtGfPHjVu3FiSlJmZ6X2ZfeTIkT63t2HDBt10001q0KCBmjVrVu59lVi0aJHat2+vOnXq6JJLLtHzzz/vc3nJy/p79uzx2b5y5Uq5XC7vs/vk5GQtWbJEe/fu9XkboERZL09u3bpV1113nRo0aKA6deqoY8eOmjdvXpn3s2DBAk2YMEGxsbEKDw9Xv379tGPHjvIP/G98/vnnSklJUVhYmEJDQ9WjRw8tWbLEe3lGRob3HzIPPvigXC6XEhMTz+m2f2vnzp0aPny4oqKi5Ha71apVK7344ouVfjzGGE2aNEkJCQmqU6eOOnfurKVLlyo5OVnJycne2+vSpYsk6fbbb/ce99OP9a5duzRo0CDVr19fcXFxuv/++1VYWHjWx3SmNfnbx3P6qzxlvT1Q8jL8N998o6uuukr16tVTTEyMnn76aUnSmjVr1LNnT9WrV0/NmzcvtRYq4lzWliTl5ORoxIgRPn9mU6dOVXFxcanHMmXKFD311FOKj4/3/nksW7as0jPi/BBq+NWgQYMUFBSk1atXl7vPnj17NHjwYIWEhOiVV17RRx99pKefflr16tXT8ePHFRMTo48++kiSdMcddyg7O1vZ2dl69NFHfW5n6NChuvTSS/XWW29p9uzZZ5xr06ZNGjdunMaPH69FixapR48euu+++/Tf//3fFX6MM2fO1JVXXqno6GjvbGd6uX3Hjh3q0aOHtm3bpueff14LFy5U69atNXLkSE2ZMqXU/g8//LD27t2rl19+WXPmzNHOnTs1ZMgQFRUVnXGuVatWqW/fvsrLy9Nf//pXLViwQGFhYRoyZIjefPNNSafeGli4cKEkaezYscrOztaiRYsq9Pi//vprdenSRVu3btXUqVP1/vvva/Dgwbr33nuVmZlZqcczYcIETZgwQVdffbXeffdd3X333brzzjv17bffevf53e9+p7lz50qSHnnkEe9x/+3bDCdOnNC1116rlJQUvfvuuxo1apSeffZZTZ48+YyP6WxrsjJOnDihoUOHavDgwXr33Xc1cOBApaen6+GHH1ZqaqpGjRqlRYsWqUWLFho5cqTWr19f4fs417X1008/qUePHvrkk0/0xBNPaPHixerXr58eeOABjRkzptTtzpgxQx999JGmT5+u119/XbVq1dLAgQMD8rYSymCACpg7d66RZNauXVvuPk2aNDGtWrXynp84caL57VL7xz/+YSSZTZs2lXsbP/30k5FkJk6cWOqyktt77LHHyr3stxISEozL5Sp1f/379zfh4eHmyJEjPo9t9+7dPvutWLHCSDIrVqzwbhs8eLBJSEgoc/bT5x42bJhxu90mJyfHZ7+BAwea0NBQ88svv/jcz6BBg3z2+/vf/24kmezs7DLvr0S3bt1MVFSUyc/P9247efKkadu2rWnatKkpLi42xhize/duI8k888wzZ7y93+47d+5c77arrrrKNG3a1OTl5fnsO2bMGFOnTh1z6NChCj2eQ4cOGbfbbW655Raf/bKzs40k07t3b++2tWvXlpqnRGpqqpFk/v73v/tsHzRokGnRosUZH+e5rMmy1oExZR+jklnefvtt77YTJ06Yxo0bG0lmw4YN3u0HDx40QUFBJi0t7YwzGlP5tfXQQw8ZSebLL7/02e+ee+4xLpfL7Nixw+exxMbGmqNHj3r383g8JjIy0vTr1++sM8L/eEYNvzNn+Yrzjh07KiQkRH/+8581b948fffdd5W6nxtvvPGc923Tpo06dOjgs2348OHyeDzasGFDpe7/XC1fvlwpKSmKi4vz2T5y5EgVFBSUepZy7bXX+pxv3769JGnv3r3l3seRI0f05Zdf6qabbvL55HNQUJBuvfVW/etf/zrnl8/P5NixY1q2bJluuOEGhYaG6uTJk97ToEGDdOzYMa1Zs6ZCj2fNmjUqLCzUzTff7LNft27dKvyyvMvl0pAhQ0rd35mOneS/NXn6LIMGDfKer127ti699FLFxMSoU6dO3u2RkZGKioo664xlOde1tXz5crVu3VpXXHFFqf2MMVq+fLnP9qFDh6pOnTre8yWvzKxevfqsr+zA/wg1/OrIkSM6ePCgYmNjy92nWbNm+vTTTxUVFaXRo0erWbNmatasmZ577rkK3VdMTMw57xsdHV3utoMHD1bofivq4MGDZc5acoxOv/+GDRv6nHe73ZKko0ePlnsfP//8s4wxFbqfyjh48KBOnjypF154QcHBwT6nkigdOHDA5zpnezwlc5V8+PC3ytp2JqGhoT6BKbm/Y8eOnfF6/lqTZ5slJCREkZGRpfYNCQk564xlOde1VdE1WN7/L8ePH9fhw4crPCfOD5/6hl8tWbJERUVF3g8AladXr17q1auXioqKtG7dOr3wwgsaN26cmjRpomHDhp3TfVXkZ7P3799f7raSkJT8pXr6B49OD09FNWzYUPv27Su1/ccff5QkNWrU6LxuX5IaNGigWrVqBeR+Sp6ljx49usx9kpKSKnSbJcf/3//+d6nL9u/fX6kPu1XG2dZkVa2P83Gua6uia7C8/19CQkIu+J9VtxHPqOE3OTk5euCBBxQREaG77rrrnK4TFBSkrl27ej8xXPIy9Lk8i6yIbdu26auvvvLZNn/+fIWFhel3v/udJHmDsHnzZp/9Fi9eXOr23G73Oc+WkpKi5cuXe/9SLPHaa68pNDTULz/OVa9ePXXt2lULFy70mau4uFivv/66mjZtqubNm5/3/YSGhqpPnz7auHGj2rdvr86dO5c6nf4M+my6du0qt9vt/cBbiTVr1pR6Odjf66Is5a3JiqyPQDnXtZWSkqKvv/661Ns8r732mlwul/r06eOzfeHChT7P8PPz8/Xee++pV69eCgoKqqJHg/LwjBqVsnXrVu97k7m5ufrss880d+5cBQUFadGiRd4fryrL7NmztXz5cg0ePFjx8fE6duyYXnnlFUlSv379JJ16TywhIUHvvvuuUlJSFBkZqUaNGlX62VVsbKyuvfZaZWRkKCYmRq+//rqWLl2qyZMnKzQ0VJLUpUsXtWjRQg888IBOnjypBg0aaNGiRfr8889L3V67du20cOFCzZo1S5dffrlq1arl83PlvzVx4kS9//776tOnjx577DFFRkbqjTfe0JIlSzRlyhRFRERU6jGdLisrS/3791efPn30wAMPKCQkRDNnztTWrVu1YMGCCv92uPI899xz6tmzp3r16qV77rlHiYmJys/P165du/Tee++Ver/zbCIjI5WWlqasrCw1aNBAN9xwg/71r38pMzNTMTExqlXr1+cTzZo1U926dfXGG2+oVatWql+/vmJjY8/4Vsu5OJc1GR0drX79+nnnTEhI0LJly7yfonfCua6t8ePH67XXXtPgwYP1+OOPKyEhQUuWLNHMmTN1zz33lPpHXFBQkPr376+0tDQVFxdr8uTJ8ng8ZX6qHwHg8IfZUM2UfDK65BQSEmKioqJM7969zaRJk0xubm6p65z+Sezs7Gxzww03mISEBON2u03Dhg1N7969zeLFi32u9+mnn5pOnToZt9ttJJnU1FSf2/vpp5/Oel/GnPrU9+DBg80//vEP06ZNGxMSEmISExPNtGnTSl3/22+/NQMGDDDh4eGmcePGZuzYsWbJkiWlPu176NAhc9NNN5mLLrrIuFwun/tUGZ9W37JlixkyZIiJiIgwISEhpkOHDqU+uVzyqeK33nrLZ3tZnyouz2effWb69u1r6tWrZ+rWrWu6detm3nvvvTJvr7Kf+i7ZPmrUKHPxxReb4OBg07hxY9OjRw/z5JNPVurxFBcXmyeffNI0bdrUhISEmPbt25v333/fdOjQwdxwww0+11+wYIFp2bKlCQ4O9jnWqamppl69eqUeQ1lr4nTnuib37dtnbrrpJhMZGWkiIiLMiBEjzLp168r81HdZs/Tu3du0adOm1PaSNXo2lV1bxhizd+9eM3z4cNOwYUMTHBxsWrRoYZ555hlTVFTk3afkz2by5MkmMzPT++fRqVMn8/HHH591PlQNlzFn+YguADhg9+7datmypSZOnKiHH37Y6XFqhD179igpKUnPPPOMHnjgAafHwf/jpW8Ajvvqq6+0YMEC9ejRQ+Hh4dqxY4emTJmi8PBw3XHHHU6PBziKUANwXL169bRu3Tr99a9/1S+//KKIiAglJyfrqaeeqvCPaAEXGl76BgDAYvx4FgAAFiPUAABYjFADAGCxCz7Uxhh5PJ6zflEEAAA2uuBDnZ+fr4iICOXn5zs9CgAAFXbBhxoAgOqMUAMAYDFCDQCAxQg1AAAWI9QAAFiMUAMAYDFCDQCAxQg1AAAWI9QAAFiMUAMAYDFCDQCAxQg1AAAWI9QAAFiMUAMAYDFCDQCAxRwN9erVqzVkyBDFxsbK5XLpnXfe8V524sQJPfjgg2rXrp3q1aun2NhY3Xbbbfrxxx+dGxgAgABzNNRHjhxRhw4dNGPGjFKXFRQUaMOGDXr00Ue1YcMGLVy4UN9++62uvfZaByYFAMAZLmOMcXoISXK5XFq0aJGuv/76cvdZu3atrrjiCu3du1fx8fHndLsej0cRERHKy8tTeHi4n6YFACAwajs9QEXk5eXJ5XLpoosuKnefwsJCFRYWes97PB6/zpCTk6MDBw749TYDqVGjRuf8jxz4X3VfPxJrCAi0ahPqY8eO6aGHHtLw4cPP+Mw4KytLmZmZVTJDTk6OWrZspaNHC6rk9gOhbt1QffPNdv6idcCFsH4k1hAQaNUi1CdOnNCwYcNUXFysmTNnnnHf9PR0paWlec97PB7FxcX5ZY4DBw7o6NECdR01UeExiX65zUDy7NujL1/J1IEDB/hL1gHVff1IrCHACdaH+sSJE7r55pu1e/duLV++/KzvM7vdbrnd7iqdKTwmUZHxLar0PnDhYv0AqAirQ10S6Z07d2rFihVq2LCh0yMBABBQjob68OHD2rVrl/f87t27tWnTJkVGRio2NlY33XSTNmzYoPfff19FRUXav3+/JCkyMlIhISFOjQ0AQMA4Gup169apT58+3vMl7y2npqYqIyNDixcvliR17NjR53orVqxQcnJyoMYEAMAxjoY6OTlZZ/oxbkt+xBsAAMfwu74BALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALCYo6FevXq1hgwZotjYWLlcLr3zzjs+lxtjlJGRodjYWNWtW1fJycnatm2bM8MCAOAAR0N95MgRdejQQTNmzCjz8ilTpmjatGmaMWOG1q5dq+joaPXv31/5+fkBnhQAAGfUdvLOBw4cqIEDB5Z5mTFG06dP14QJEzR06FBJ0rx589SkSRPNnz9fd911VyBHBQDAEY6G+kx2796t/fv3a8CAAd5tbrdbvXv31hdffFFuqAsLC1VYWOg97/F4qnzW6mb79u1Oj1BphYWFcrvdTo9RKdX5uANwjrWh3r9/vySpSZMmPtubNGmivXv3lnu9rKwsZWZmVuls1dXRvIOSXBoxYoTTo1SeyyUZ4/QU5+VE4XGnRwBQjVgb6hIul8vnvDGm1LbfSk9PV1pamve8x+NRXFxclc1XnZwoyJdk1HH4g2qc1NLpcSps35ZsbV08p9rPf/LkSadHAVCNWBvq6OhoSaeeWcfExHi35+bmlnqW/Vtut7vavjQaKPWj4hUZ38LpMSrMs2+PpOo/PwBUhLU/R52UlKTo6GgtXbrUu+348eNatWqVevTo4eBkAAAEjqPPqA8fPqxdu3Z5z+/evVubNm1SZGSk4uPjNW7cOE2aNEmXXXaZLrvsMk2aNEmhoaEaPny4g1MDABA4joZ63bp16tOnj/d8yXvLqampevXVV/Vf//VfOnr0qP7yl7/o559/VteuXfXJJ58oLCzMqZEBAAgoR0OdnJwsc4ZP8LpcLmVkZCgjIyNwQwEAYBFr36MGAACEGgAAqxFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGKEGgAAixFqAAAsRqgBALAYoQYAwGJWh/rkyZN65JFHlJSUpLp16+qSSy7R448/ruLiYqdHAwAgIGo7PcCZTJ48WbNnz9a8efPUpk0brVu3TrfffrsiIiJ03333OT0eAABVzupQZ2dn67rrrtPgwYMlSYmJiVqwYIHWrVvn8GQAAASG1aHu2bOnZs+erW+//VbNmzfXV199pc8//1zTp08v9zqFhYUqLCz0nvd4PAGYFAACIycnRwcOHHB6jEpr1KiR4uPjnR6jWrE61A8++KDy8vLUsmVLBQUFqaioSE899ZT+8Ic/lHudrKwsZWZmBnBKAAiMnJwctWzZSkePFjg9SqXVrRuqb77ZTqwrwOpQv/nmm3r99dc1f/58tWnTRps2bdK4ceMUGxur1NTUMq+Tnp6utLQ073mPx6O4uLhAjQwAVebAgQM6erRAXUdNVHhMotPjVJhn3x59+UqmDhw4QKgrwOpQ/+d//qceeughDRs2TJLUrl077d27V1lZWeWG2u12y+12B3JMAAio8JhERca3cHoMBIjVP55VUFCgWrV8RwwKCuLHswAANYbVz6iHDBmip556SvHx8WrTpo02btyoadOmadSoUU6PBgBAQFgd6hdeeEGPPvqo/vKXvyg3N1exsbG666679Nhjjzk9GgAAAWF1qMPCwjR9+vQz/jgWAAAXMqvfowYAoKYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFKhXqSy65RAcPHiy1/ZdfftEll1xy3kMBAIBTKhXqPXv2qKioqNT2wsJC/fDDD+c9FAAAOKV2RXZevHix978//vhjRUREeM8XFRVp2bJlSkxM9NtwAADUdBUK9fXXXy9JcrlcSk1N9bksODhYiYmJmjp1qt+GAwCgpqtQqIuLiyVJSUlJWrt2rRo1alQlQwEAgFMqFOoSu3fv9vccAACgDJUKtSQtW7ZMy5YtU25urveZdolXXnnlvAcDAACVDHVmZqYef/xxde7cWTExMXK5XP6eCwAAqJKhnj17tl599VXdeuut/p4HAAD8RqV+jvr48ePq0aOHv2cBAACnqVSo77zzTs2fP9/fswAAgNNU6qXvY8eOac6cOfr000/Vvn17BQcH+1w+bdo0vwwHAEBNV6lQb968WR07dpQkbd261ecyPlgGAID/VCrUK1as8PccAACgDHzNJQAAFqvUM+o+ffqc8SXu5cuXV3ogAADwq0qFuuT96RInTpzQpk2btHXr1lJf1gEAACqvUqF+9tlny9yekZGhw4cPn9dAAADgV359j3rEiBH8nm8AAPzIr6HOzs5WnTp1/HmTAADUaJV66Xvo0KE+540x2rdvn9atW6dHH33UL4MBAIBKhjoiIsLnfK1atdSiRQs9/vjjGjBggF8GAwAAlQz13Llz/T0HAAAoQ6VCXWL9+vXavn27XC6XWrdurU6dOvlrLgAAoEp+mCw3N1d9+/ZVly5ddO+992rMmDG6/PLLlZKSop9++smvA/7www8aMWKEGjZsqNDQUHXs2FHr16/3630AAGCrSoV67Nix8ng82rZtmw4dOqSff/5ZW7dulcfj0b333uu34X7++WddeeWVCg4O1ocffqivv/5aU6dO1UUXXeS3+wAAwGaVeun7o48+0qeffqpWrVp5t7Vu3VovvviiXz9MNnnyZMXFxfm8J56YmOi32wcAwHaVekZdXFxc6juoJSk4OFjFxcXnPVSJxYsXq3PnzvqP//gPRUVFqVOnTnrppZf8dvsAANiuUs+o+/btq/vuu08LFixQbGyspFPvJY8fP14pKSl+G+67777TrFmzlJaWpocfflj/+7//q3vvvVdut1u33XZbmdcpLCxUYWGh97zH4/HbPABO2b59u9MjVFqjRo0UHx/v9BjAOatUqGfMmKHrrrtOiYmJiouLk8vlUk5Ojtq1a6fXX3/db8MVFxerc+fOmjRpkiSpU6dO2rZtm2bNmlVuqLOyspSZmem3GQD86mjeQUkujRgxwulRKq1u3VB98812Yo1qo1KhjouL04YNG7R06VJ98803MsaodevW6tevn1+Hi4mJUevWrX22tWrVSm+//Xa510lPT1daWpr3vMfjUVxcnF/nAmqqEwX5kow6Dn9QjZNaOj1OhXn27dGXr2TqwIEDhBrVRoVCvXz5co0ZM0Zr1qxReHi4+vfvr/79+0uS8vLy1KZNG82ePVu9evXyy3BXXnmlduzY4bPt22+/VUJCQrnXcbvdcrvdfrl/AGWrHxWvyPgWTo8B1AgV+jDZ9OnT9ac//Unh4eGlLouIiNBdd92ladOm+W248ePHa82aNZo0aZJ27dql+fPna86cORo9erTf7gMAAJtVKNRfffWVrr766nIvHzBggF9/GUmXLl20aNEiLViwQG3bttUTTzyh6dOn649//KPf7gMAAJtV6KXvf//732X+WJb3xmrX9vtvJrvmmmt0zTXX+PU2AQCoLir0jPriiy/Wli1byr188+bNiomJOe+hAADAKRUK9aBBg/TYY4/p2LFjpS47evSoJk6cyLNfAAD8qEIvfT/yyCNauHChmjdvrjFjxqhFixZyuVzavn27XnzxRRUVFWnChAlVNSsAADVOhULdpEkTffHFF7rnnnuUnp4uY4wkyeVy6aqrrtLMmTPVpEmTKhkUAICaqMK/8CQhIUEffPCBfv75Z+3atUvGGF122WVq0KBBVcwHAECNVqnfTCZJDRo0UJcuXfw5CwAAOE2lvj0LAAAEBqEGAMBihBoAAIsRagAALEaoAQCwGKEGAMBihBoAAIsRagAALEaoAQCwGKEGAMBihBoAAIsRagAALEaoAQCwGKEGAMBihBoAAIsRagAALEaoAQCwGKEGAMBihBoAAIsRagAALFbb6QEAINC2b9/u9AiVUl3nxvkh1ABqjKN5ByW5NGLECKdHOS8nCo87PQICiFADqDFOFORLMuo4/EE1Tmrp9DgVtm9LtrYunqOTJ086PQoCiFADqHHqR8UrMr6F02NUmGffHqdHgAP4MBkAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgMUINAIDFCDUAABYj1AAAWIxQAwBgsWoV6qysLLlcLo0bN87pUQAACIhqE+q1a9dqzpw5at++vdOjAAAQMNUi1IcPH9Yf//hHvfTSS2rQoIHT4wAAEDC1nR7gXIwePVqDBw9Wv3799OSTT55x38LCQhUWFnrPezyeqh4PAFAB27dvd3qESmvUqJHi4+MDep/Wh/pvf/ubNmzYoLVr157T/llZWcrMzKziqQAAFXU076Akl0aMGOH0KJVWt26ovvlme0BjbXWov//+e91333365JNPVKdOnXO6Tnp6utLS0rznPR6P4uLiqmpEAMA5OlGQL8mo4/AH1TippdPjVJhn3x59+UqmDhw4QKhLrF+/Xrm5ubr88su924qKirR69WrNmDFDhYWFCgoK8rmO2+2W2+0O9KgAgHNUPypekfEtnB6j2rA61CkpKdqyZYvPtttvv10tW7bUgw8+WCrSAABcaKwOdVhYmNq2beuzrV69emrYsGGp7QAAXIiqxY9nAQBQU1n9jLosK1eudHoEAAAChmfUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAAFiPUAABYzOpQZ2VlqUuXLgoLC1NUVJSuv/567dixw+mxAAAIGKtDvWrVKo0ePVpr1qzR0qVLdfLkSQ0YMEBHjhxxejQAAAKittMDnMlHH33kc37u3LmKiorS+vXr9fvf/96hqQAACByrQ326vLw8SVJkZGS5+xQWFqqwsNB73uPxVPlcAABUFatf+v4tY4zS0tLUs2dPtW3bttz9srKyFBER4T3FxcUFcEoAAPyr2oR6zJgx2rx5sxYsWHDG/dLT05WXl+c9ff/99wGaEAAA/6sWL32PHTtWixcv1urVq9W0adMz7ut2u+V2uwM0GQAAVcvqUBtjNHbsWC1atEgrV65UUlKS0yMBABBQVod69OjRmj9/vt59912FhYVp//79kqSIiAjVrVvX4ekAAKh6Vr9HPWvWLOXl5Sk5OVkxMTHe05tvvun0aAAABITVz6iNMU6PAACAo6x+Rg0AQE1HqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxGqAEAsBihBgDAYoQaAACLEWoAACxWLUI9c+ZMJSUlqU6dOrr88sv12WefOT0SAAABYX2o33zzTY0bN04TJkzQxo0b1atXLw0cOFA5OTlOjwYAQJWzPtTTpk3THXfcoTvvvFOtWrXS9OnTFRcXp1mzZjk9GgAAVa620wOcyfHjx7V+/Xo99NBDPtsHDBigL774oszrFBYWqrCw0Hs+Ly9PkuTxeM57nsOHD0uSDu3doZOFR8/79gLNs2+vJCnvh50Kru1yeJqKY37nVffHwPzOqvbz7z/1Su7hw4f90hRJCgsLk8t1lmNhLPbDDz8YSeaf//ynz/annnrKNG/evMzrTJw40UjixIkTJ06crD/l5eWdtYVWP6Mucfq/Nowx5f4LJD09XWlpad7zxcXFOnTokBo2bHj2f7VYxOPxKC4uTt9//73Cw8OdHscxHIdTOA6/4licwnE4pbofh7CwsLPuY3WoGzVqpKCgIO3fv99ne25urpo0aVLmddxut9xut8+2iy66qKpGrHLh4eHVcvH5G8fhFI7DrzgWp3AcTrmQj4PVHyYLCQnR5ZdfrqVLl/psX7p0qXr06OHQVAAABI7Vz6glKS0tTbfeeqs6d+6s7t27a86cOcrJydHdd9/t9GgAAFQ560N9yy236ODBg3r88ce1b98+tW3bVh988IESEhKcHq1Kud1uTZw4sdTL+DUNx+EUjsOvOBancBxOqQnHwWWMMU4PAQAAymb1e9QAANR0hBoAAIsRagAALEaoAQCwGKG2SEZGhlwul88pOjra6bECYvXq1RoyZIhiY2Plcrn0zjvv+FxujFFGRoZiY2NVt25dJScna9u2bc4MW4XOdhxGjhxZao1069bNmWGrUFZWlrp06aKwsDBFRUXp+uuv144dO3z2qQlr4lyOQ01YE7NmzVL79u29v9Ske/fu+vDDD72XX+hrgVBbpk2bNtq3b5/3tGXLFqdHCogjR46oQ4cOmjFjRpmXT5kyRdOmTdOMGTO0du1aRUdHq3///srPzw/wpFXrbMdBkq6++mqfNfLBBx8EcMLAWLVqlUaPHq01a9Zo6dKlOnnypAYMGKAjR45496kJa+JcjoN04a+Jpk2b6umnn9a6deu0bt069e3bV9ddd503xhf8WjjP782AH02cONF06NDB6TEcJ8ksWrTIe764uNhER0ebp59+2rvt2LFjJiIiwsyePduBCQPj9ONgjDGpqanmuuuuc2QeJ+Xm5hpJZtWqVcaYmrsmTj8OxtTcNdGgQQPz8ssv14i1wDNqy+zcuVOxsbFKSkrSsGHD9N133zk9kuN2796t/fv3a8CAAd5tbrdbvXv3LvfrTi9kK1euVFRUlJo3b64//elPys3NdXqkKlfydbWRkZGSau6aOP04lKhJa6KoqEh/+9vfdOTIEXXv3r1GrAVCbZGuXbvqtdde08cff6yXXnpJ+/fvV48ePXTw4EGnR3NUyZeynP5FLE2aNCn1hS0XuoEDB+qNN97Q8uXLNXXqVK1du1Z9+/b1+Q72C40xRmlpaerZs6fatm0rqWauibKOg1Rz1sSWLVtUv359ud1u3X333Vq0aJFat25dI9aC9b9CtCYZOHCg97/btWun7t27q1mzZpo3b57PV3fWVBX5utML1S233OL977Zt26pz585KSEjQkiVLNHToUAcnqzpjxozR5s2b9fnnn5e6rCatifKOQ01ZEy1atNCmTZv0yy+/6O2331ZqaqpWrVrlvfxCXgs8o7ZYvXr11K5dO+3cudPpURxV8sn3inzdaU0RExOjhISEC3aNjB07VosXL9aKFSvUtGlT7/aatibKOw5luVDXREhIiC699FJ17txZWVlZ6tChg5577rkasRYItcUKCwu1fft2xcTEOD2Ko5KSkhQdHe3zdafHjx/XqlWravzXnR48eFDff//9BbdGjDEaM2aMFi5cqOXLlyspKcnn8pqyJs52HMpyoa6J0xljVFhYWDPWgnOfY8Pp7r//frNy5Urz3XffmTVr1phrrrnGhIWFmT179jg9WpXLz883GzduNBs3bjSSzLRp08zGjRvN3r17jTHGPP300yYiIsIsXLjQbNmyxfzhD38wMTExxuPxODy5f53pOOTn55v777/ffPHFF2b37t1mxYoVpnv37ubiiy++4I7DPffcYyIiIszKlSvNvn37vKeCggLvPjVhTZztONSUNZGenm5Wr15tdu/ebTZv3mwefvhhU6tWLfPJJ58YYy78tUCoLXLLLbeYmJgYExwcbGJjY83QoUPNtm3bnB4rIFasWGEklTqlpqYaY079OM7EiRNNdHS0cbvd5ve//73ZsmWLs0NXgTMdh4KCAjNgwADTuHFjExwcbOLj401qaqrJyclxemy/K+sYSDJz58717lMT1sTZjkNNWROjRo0yCQkJJiQkxDRu3NikpKR4I23Mhb8W+JpLAAAsxnvUAABYjFADAGAxQg0AgMUINQAAFiPUAABYjFADAGAxQg0AgMUINQAfycnJGjdunNNjAPh/hBoAAIsRagAALEaoAZRy8uRJjRkzRhdddJEaNmyoRx55RCW/bTgxMVFPPPGEhg8frvr16ys2NlYvvPCCwxMDFy5CDaCUefPmqXbt2vryyy/1/PPP69lnn9XLL7/svfyZZ55R+/bttWHDBqWnp2v8+PE+XzMIwH/4Ug4APpKTk5Wbm6tt27bJ5XJJkh566CEtXrxYX3/9tRITE9WqVSt9+OGH3usMGzZMHo9HH3zwgVNjAxcsnlEDKKVbt27eSEtS9+7dtXPnThUVFXnP/1b37t21ffv2gM4I1BSEGoBf/DbsAPyHUAMoZc2aNaXOX3bZZQoKCir38pYtWwZsPqAmqe30AADs8/333ystLU133XWXNmzYoBdeeEFTp071Xv7Pf/5TU6ZM0fXXX6+lS5fqrbfe0pIlSxycGLhwEWoApdx22206evSorrjiCgUFBWns2LH685//7L38/vvv1/r165WZmamwsDBNnTpVV111lYMTAxcuPvUNoEISExM1btw4fs0oECC8Rw0AgMUINQAAFuOlbwAALMYzagAALEaoAQCwGKEGAMBihBoAAIsRagAALEaoAQCwGKEGAMBihBoAAIsRagAALPZ/QkyyKxkgW8EAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 500x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(g4pqsconf['llall']).set(title='Distribution of length sum loop', xlabel='bp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11df4589-90e3-4088-b9d1-b7bf20bb0f85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a3cb3c-5c99-4fc4-9390-cff64060882b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381b5a29-7a6b-4f11-95cb-62cd4bb88e71",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bd6f18e2-7ad4-4cc6-88e8-47f48288ac5f",
   "metadata": {},
   "source": [
    "## NEW Add category to nucteotids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "56c23157-fc51-4fa7-807b-0d817676ed69",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_positions = []\n",
    "for seq, start, G4_id in zip(g4pqsconf['sequence'], g4pqsconf['start'], g4pqsconf['G4_id']):\n",
    "    for (i, j), (x, y) in zip(enumerate(seq, start=start), enumerate(seq)):\n",
    "        row = [G4_id,x,i,j]\n",
    "        nuc_positions.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "7955148f-f5f8-4e5b-a116-ee4086e56db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_positions = pd.DataFrame(nuc_positions, columns=['G4_id', 'pos_in_seq', 'pos_in_ref', 'nuc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "5fe5a503-0c21-48d1-8897-ca9c96391b95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>pos_in_seq</th>\n",
       "      <th>pos_in_ref</th>\n",
       "      <th>nuc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>296</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>297</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>298</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>299</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418</th>\n",
       "      <td>67</td>\n",
       "      <td>23</td>\n",
       "      <td>16376</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2419</th>\n",
       "      <td>67</td>\n",
       "      <td>24</td>\n",
       "      <td>16377</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>67</td>\n",
       "      <td>25</td>\n",
       "      <td>16378</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2421</th>\n",
       "      <td>67</td>\n",
       "      <td>26</td>\n",
       "      <td>16379</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2422</th>\n",
       "      <td>67</td>\n",
       "      <td>27</td>\n",
       "      <td>16380</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2423 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id  pos_in_seq  pos_in_ref nuc\n",
       "0         1           0         295   C\n",
       "1         1           1         296   C\n",
       "2         1           2         297   A\n",
       "3         1           3         298   C\n",
       "4         1           4         299   C\n",
       "...     ...         ...         ...  ..\n",
       "2418     67          23       16376   C\n",
       "2419     67          24       16377   C\n",
       "2420     67          25       16378   C\n",
       "2421     67          26       16379   C\n",
       "2422     67          27       16380   C\n",
       "\n",
       "[2423 rows x 4 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuc_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "c0db05cf-bb35-452b-9923-3ebc441b55c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sequence = '[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]'  #for example\n",
    "def categorize_sequence(sequence):  \n",
    "    pattern = r'\\[[A-Z]+\\]|\\([A-Z]+\\)|\\[[A-Z]+|[A-Z]+\\]|[A-Z]+|[a-z]+'\n",
    "    \n",
    "    matches = re.findall(pattern, sequence)\n",
    "    categories = []\n",
    "    \n",
    "    for match in matches:\n",
    "        # print(match)\n",
    "        if match.startswith('[') or match.endswith(']'):\n",
    "            categories.extend(['Tetrad'] * len(match))\n",
    "        elif match.startswith('(') and match.endswith(')'):\n",
    "            categories.extend(['Bulges'] * len(match))\n",
    "        elif match.isupper():\n",
    "            categories.extend(['Loop'] * len(match))\n",
    "        elif match.islower():\n",
    "            categories.extend(['Mismatch'] * len(match))\n",
    "    \n",
    "    char_result = [char for char in sequence]\n",
    "\n",
    "    return char_result, categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "05cc6fe0-61d4-4276-b296-0fa44e1c2c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# categorize_sequence('[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "4be1a3fc-4522-4f59-9d08-0d3bce0f539f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category = pd.DataFrame.from_records(g4pqsconf['pattern'].apply(categorize_sequence), \n",
    "                          columns=['char', 'category']) \\\n",
    "    .explode(['char', 'category']) \\\n",
    "    .reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "0252024c-937b-49d6-84e8-0981178a8155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>char</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(</td>\n",
       "      <td>Bulges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>Bulges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3160</th>\n",
       "      <td>[</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161</th>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162</th>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3163</th>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3164</th>\n",
       "      <td>]</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3165 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     char category\n",
       "0       [   Tetrad\n",
       "1       C   Tetrad\n",
       "2       C   Tetrad\n",
       "3       (   Bulges\n",
       "4       A   Bulges\n",
       "...   ...      ...\n",
       "3160    [   Tetrad\n",
       "3161    C   Tetrad\n",
       "3162    C   Tetrad\n",
       "3163    C   Tetrad\n",
       "3164    ]   Tetrad\n",
       "\n",
       "[3165 rows x 2 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "adc63309-b346-4e81-bcca-6fc3f23fc7c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category = df_category[~df_category['char'].isin(['[', ']', '(', ')'])].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "2c8574ee-9868-4c7b-be47-c7dc01e3a0b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_category = df_category.rename(columns={'char':'nuc'})\n",
    "df_category['char'] = df_category['char'].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "9ebb82c3-2a21-4b2a-8074-7ec5995bae57",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category = nuc_positions.join(df_category).drop('char', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "f3457d54-ad52-4b08-851d-ff4b29bcf325",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>pos_in_seq</th>\n",
       "      <th>pos_in_ref</th>\n",
       "      <th>nuc</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>296</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>297</td>\n",
       "      <td>A</td>\n",
       "      <td>Bulges</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>298</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>299</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418</th>\n",
       "      <td>67</td>\n",
       "      <td>23</td>\n",
       "      <td>16376</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2419</th>\n",
       "      <td>67</td>\n",
       "      <td>24</td>\n",
       "      <td>16377</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>67</td>\n",
       "      <td>25</td>\n",
       "      <td>16378</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2421</th>\n",
       "      <td>67</td>\n",
       "      <td>26</td>\n",
       "      <td>16379</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2422</th>\n",
       "      <td>67</td>\n",
       "      <td>27</td>\n",
       "      <td>16380</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2423 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id  pos_in_seq  pos_in_ref nuc category\n",
       "0         1           0         295   C   Tetrad\n",
       "1         1           1         296   C   Tetrad\n",
       "2         1           2         297   A   Bulges\n",
       "3         1           3         298   C   Tetrad\n",
       "4         1           4         299   C   Tetrad\n",
       "...     ...         ...         ...  ..      ...\n",
       "2418     67          23       16376   C   Tetrad\n",
       "2419     67          24       16377   C   Tetrad\n",
       "2420     67          25       16378   C   Tetrad\n",
       "2421     67          26       16379   C   Tetrad\n",
       "2422     67          27       16380   C   Tetrad\n",
       "\n",
       "[2423 rows x 5 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1d65fa85-7bae-4cfa-9fb0-eaea6a3fde6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_category.to_csv('../data/processed/df_category_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0cba513-d59f-4d79-b155-f1d79cecae66",
   "metadata": {},
   "source": [
    "#### Choose only loops from each G4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "9795965f-704b-4f0a-ab98-923583600a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category = pd.read_csv('../data/processed/df_category_pqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "9a0b1ab1-d975-483b-90aa-99e58fcf226e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>pos_in_seq</th>\n",
       "      <th>pos_in_ref</th>\n",
       "      <th>nuc</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>300</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>301</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>302</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>307</td>\n",
       "      <td>C</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>357</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2412</th>\n",
       "      <td>67</td>\n",
       "      <td>17</td>\n",
       "      <td>16370</td>\n",
       "      <td>G</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413</th>\n",
       "      <td>67</td>\n",
       "      <td>18</td>\n",
       "      <td>16371</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2414</th>\n",
       "      <td>67</td>\n",
       "      <td>19</td>\n",
       "      <td>16372</td>\n",
       "      <td>T</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2415</th>\n",
       "      <td>67</td>\n",
       "      <td>20</td>\n",
       "      <td>16373</td>\n",
       "      <td>G</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2416</th>\n",
       "      <td>67</td>\n",
       "      <td>21</td>\n",
       "      <td>16374</td>\n",
       "      <td>A</td>\n",
       "      <td>Loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1011 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id  pos_in_seq  pos_in_ref nuc category\n",
       "5         1           5         300   A     Loop\n",
       "6         1           6         301   A     Loop\n",
       "7         1           7         302   A     Loop\n",
       "12        1          12         307   C     Loop\n",
       "25        2           4         357   A     Loop\n",
       "...     ...         ...         ...  ..      ...\n",
       "2412     67          17       16370   G     Loop\n",
       "2413     67          18       16371   A     Loop\n",
       "2414     67          19       16372   T     Loop\n",
       "2415     67          20       16373   G     Loop\n",
       "2416     67          21       16374   A     Loop\n",
       "\n",
       "[1011 rows x 5 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_category[df_category['category'] == 'Loop']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "23250286-8adb-44c6-83b3-c78a4b6dc85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_loops = []\n",
    "for G4_id in df_category[df_category['category'] == 'Loop']['G4_id'].unique():\n",
    "    row = []\n",
    "    positions = df_category[(df_category['category'] == 'Loop') & (df_category['G4_id'] == G4_id)]['pos_in_ref'].values.tolist()\n",
    "    seq = ''.join(df_category[(df_category['category'] == 'Loop') & (df_category['G4_id'] == G4_id)]['nuc'].values)\n",
    "\n",
    "    row.append(G4_id)\n",
    "    row.append('loops')\n",
    "    row.append(positions)\n",
    "    row.append(seq)\n",
    "    data_loops.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "73ce6988-a912-45ed-b587-b7c873895a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_loops\n",
    "data_loops = pd.DataFrame(data=data_loops, columns = ['G4_id', 'side', 'positions', 'sequence'])\n",
    "nuc_count = data_loops['sequence'].apply(lambda x: Counter(x))\n",
    "nuc_count = pd.DataFrame.from_records(nuc_count).sort_index(axis=1)\n",
    "data_loops = data_loops.join(nuc_count).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "759830ad-ca33-49c1-8ad9-5d8e0d7d6280",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>positions</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>loops</td>\n",
       "      <td>[300, 301, 302, 307]</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>loops</td>\n",
       "      <td>[357, 358, 359, 360, 361, 372, 373, 374, 379, ...</td>\n",
       "      <td>AAAAATAAAG</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>loops</td>\n",
       "      <td>[464, 470, 471, 472, 473, 474, 475, 476, 477, ...</td>\n",
       "      <td>AATACTACTAATCTCATCAATACAAG</td>\n",
       "      <td>12</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>loops</td>\n",
       "      <td>[538, 539, 540, 545, 546, 547, 548, 549, 550, ...</td>\n",
       "      <td>ATAGAACCAACCAAAAAAGACA</td>\n",
       "      <td>14</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>loops</td>\n",
       "      <td>[961, 966, 967, 968, 969, 970, 971, 972, 973, ...</td>\n",
       "      <td>TAATAAAGCTAAAACTTGAGTTGTAAAAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>63</td>\n",
       "      <td>loops</td>\n",
       "      <td>[15530, 15531, 15532, 15533, 15534, 15535, 155...</td>\n",
       "      <td>TTAAACATA</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>64</td>\n",
       "      <td>loops</td>\n",
       "      <td>[15633, 15634, 15635, 15636, 15637, 15638, 156...</td>\n",
       "      <td>TATCCATTAGCAATAATAT</td>\n",
       "      <td>8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>65</td>\n",
       "      <td>loops</td>\n",
       "      <td>[16170, 16171, 16172, 16177, 16178, 16179, 161...</td>\n",
       "      <td>AATATCAAAACT</td>\n",
       "      <td>7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>66</td>\n",
       "      <td>loops</td>\n",
       "      <td>[16263, 16264, 16265, 16269, 16270, 16271, 162...</td>\n",
       "      <td>TCAACTAGGATACCAACAAACCTAA</td>\n",
       "      <td>12</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>67</td>\n",
       "      <td>loops</td>\n",
       "      <td>[16356, 16357, 16358, 16359, 16360, 16361, 163...</td>\n",
       "      <td>TTCTCGTCATGGATGA</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id   side                                          positions  \\\n",
       "0       1  loops                               [300, 301, 302, 307]   \n",
       "1       2  loops  [357, 358, 359, 360, 361, 372, 373, 374, 379, ...   \n",
       "2       3  loops  [464, 470, 471, 472, 473, 474, 475, 476, 477, ...   \n",
       "3       4  loops  [538, 539, 540, 545, 546, 547, 548, 549, 550, ...   \n",
       "4       5  loops  [961, 966, 967, 968, 969, 970, 971, 972, 973, ...   \n",
       "..    ...    ...                                                ...   \n",
       "62     63  loops  [15530, 15531, 15532, 15533, 15534, 15535, 155...   \n",
       "63     64  loops  [15633, 15634, 15635, 15636, 15637, 15638, 156...   \n",
       "64     65  loops  [16170, 16171, 16172, 16177, 16178, 16179, 161...   \n",
       "65     66  loops  [16263, 16264, 16265, 16269, 16270, 16271, 162...   \n",
       "66     67  loops  [16356, 16357, 16358, 16359, 16360, 16361, 163...   \n",
       "\n",
       "                          sequence   A    C    G    T  \n",
       "0                             AAAC   3  1.0  0.0  0.0  \n",
       "1                       AAAAATAAAG   8  0.0  1.0  1.0  \n",
       "2       AATACTACTAATCTCATCAATACAAG  12  6.0  1.0  7.0  \n",
       "3           ATAGAACCAACCAAAAAAGACA  14  5.0  2.0  1.0  \n",
       "4   TAATAAAGCTAAAACTTGAGTTGTAAAAAA  16  2.0  4.0  8.0  \n",
       "..                             ...  ..  ...  ...  ...  \n",
       "62                       TTAAACATA   5  1.0  0.0  3.0  \n",
       "63             TATCCATTAGCAATAATAT   8  3.0  1.0  7.0  \n",
       "64                    AATATCAAAACT   7  2.0  0.0  3.0  \n",
       "65       TCAACTAGGATACCAACAAACCTAA  12  7.0  2.0  4.0  \n",
       "66                TTCTCGTCATGGATGA   3  3.0  4.0  6.0  \n",
       "\n",
       "[67 rows x 8 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "cb159650-350b-454d-87ba-52c9e3114ef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_loops.to_csv('../data/processed/data_loops_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e30ab88-66fe-42d4-9782-203264d269f6",
   "metadata": {},
   "source": [
    "## Step 2\n",
    "### Add columns: \n",
    "- 'bp_untill_next_G4' and 'bp_untill_previous_G4'\n",
    "- 'up_win' and 'down_win "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6f37550-8c22-43fb-8b39-311763b0cefd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f30b178d-e771-4e53-8fa9-c3e8ee52f1c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "refG4 = pd.read_excel('../data/processed/refG4.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "571255a8-f1b3-42d2-acff-528e9d0f97b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = refG4[refG4['confirmed'] == 1]\n",
    "g4conf = g4conf.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b97baa6a-b99e-4d5a-8209-ed5db7663bc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>confirmed</th>\n",
       "      <th>G4</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.25</td>\n",
       "      <td>-2.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>185</td>\n",
       "      <td>432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.10</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>1</td>\n",
       "      <td>UG4</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>248</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.15</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>785</td>\n",
       "      <td>808</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.21</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCTTAGCCTAGCCACACCCCCAC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.39</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>1</td>\n",
       "      <td>G4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            seqnames  start  end  width  \\\n",
       "0  NC_012920.1 Homo sapiens mitochondrion, comple...    285  332     48   \n",
       "1  NC_012920.1 Homo sapiens mitochondrion, comple...    348  378     31   \n",
       "2  NC_012920.1 Homo sapiens mitochondrion, comple...    418  592    175   \n",
       "3  NC_012920.1 Homo sapiens mitochondrion, comple...    785  808     24   \n",
       "4  NC_012920.1 Homo sapiens mitochondrion, comple...    954  984     31   \n",
       "\n",
       "  strand  score  max_score  threshold  window  \\\n",
       "0      -  -1.25      -2.24          1      25   \n",
       "1      -  -1.10      -1.16          1      25   \n",
       "2      -  -1.15      -1.88          1      25   \n",
       "3      -  -1.21      -1.16          1      25   \n",
       "4      -  -1.39      -1.48          1      25   \n",
       "\n",
       "                                            sequence  confirmed   G4  \\\n",
       "0   CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC          1   G4   \n",
       "1                    CCAAACCCCAAAAACAAAGAACCCTAACACC          1  UG4   \n",
       "2  CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...          1   G4   \n",
       "3                           CGCTTAGCCTAGCCACACCCCCAC          1   G4   \n",
       "4                    CACCCCCTCCCCAATAAAGCTAAAACTCACC          1   G4   \n",
       "\n",
       "                         note  start-100bp  end+100bp  \n",
       "0            more than africa          185        432  \n",
       "1            less than africa          248        478  \n",
       "2  5 overlapping g4 in africa          318        692  \n",
       "3                         NaN          685        908  \n",
       "4                         NaN          854       1084  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "de87fae6-c002-4148-b019-731261035e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = add_columns(g4conf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "dbe53171-6128-4cfb-b51b-dc2249d1c18d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4conf.to_csv('../data/processed/g4conf.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2874dcb7-8a9a-417e-8d86-30b6cebed367",
   "metadata": {},
   "source": [
    "### NEW pqsfinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e227ea1e-d172-494d-bab1-562562714d15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACCAAACCCCCCCTCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>22</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end  score                                            pattern  \\\n",
       "0    295   315     64                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "1    353   387     33    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "2    458   506     55  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "3    534   571     88     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "4    957  1002     22  [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "\n",
       "   G4_id                                           sequence  \n",
       "0      1                              CCACCAAACCCCCCCTCCCCC  \n",
       "1      2                CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC  \n",
       "2      3  CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC  \n",
       "3      4             CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC  \n",
       "4      5     CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2db98ee4-7976-4808-baa3-285c671ec1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf = add_columns(g4pqsconf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "60947fd7-0292-40ce-9369-7b7bf2e7d21e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACCAAACCCCCCCTCCCCC</td>\n",
       "      <td>38</td>\n",
       "      <td>484</td>\n",
       "      <td>195</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC</td>\n",
       "      <td>71</td>\n",
       "      <td>38</td>\n",
       "      <td>315</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC</td>\n",
       "      <td>28</td>\n",
       "      <td>71</td>\n",
       "      <td>387</td>\n",
       "      <td>534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC</td>\n",
       "      <td>386</td>\n",
       "      <td>28</td>\n",
       "      <td>506</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>22</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC</td>\n",
       "      <td>680</td>\n",
       "      <td>386</td>\n",
       "      <td>857</td>\n",
       "      <td>1102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end  score                                            pattern  \\\n",
       "0    295   315     64                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "1    353   387     33    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "2    458   506     55  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "3    534   571     88     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "4    957  1002     22  [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "\n",
       "   G4_id                                           sequence  \\\n",
       "0      1                              CCACCAAACCCCCCCTCCCCC   \n",
       "1      2                CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC   \n",
       "2      3  CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC   \n",
       "3      4             CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC   \n",
       "4      5     CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC   \n",
       "\n",
       "   bp_untill_next_G4  bp_untill_previous_G4  up_win  down_win  \n",
       "0                 38                    484     195       353  \n",
       "1                 71                     38     315       458  \n",
       "2                 28                     71     387       534  \n",
       "3                386                     28     506       671  \n",
       "4                680                    386     857      1102  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d59b43bb-e471-47ba-b3ae-d5cd21528226",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4pqsconf.to_csv('../data/processed/g4pqsconf.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81a6b18c-8247-414a-b565-2c0a72537d6b",
   "metadata": {},
   "source": [
    "#### **Add +-15bp**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5021b88f-fe51-4f0b-839b-6545b3d386ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACCAAACCCCCCCTCCCCC</td>\n",
       "      <td>38</td>\n",
       "      <td>484</td>\n",
       "      <td>195</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC</td>\n",
       "      <td>71</td>\n",
       "      <td>38</td>\n",
       "      <td>315</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC</td>\n",
       "      <td>28</td>\n",
       "      <td>71</td>\n",
       "      <td>387</td>\n",
       "      <td>534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC</td>\n",
       "      <td>386</td>\n",
       "      <td>28</td>\n",
       "      <td>506</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>22</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC</td>\n",
       "      <td>680</td>\n",
       "      <td>386</td>\n",
       "      <td>857</td>\n",
       "      <td>1102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>15526</td>\n",
       "      <td>15556</td>\n",
       "      <td>83</td>\n",
       "      <td>[CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]</td>\n",
       "      <td>63</td>\n",
       "      <td>CCCCTTAAACACCCCTCCCCACATCAAGCCC</td>\n",
       "      <td>68</td>\n",
       "      <td>8</td>\n",
       "      <td>15518</td>\n",
       "      <td>15624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>29</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>64</td>\n",
       "      <td>CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC</td>\n",
       "      <td>491</td>\n",
       "      <td>68</td>\n",
       "      <td>15556</td>\n",
       "      <td>15768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>51</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>65</td>\n",
       "      <td>CATAAAAACCCAATCCACATCAAAACCCCCTCCCC</td>\n",
       "      <td>67</td>\n",
       "      <td>491</td>\n",
       "      <td>16059</td>\n",
       "      <td>16260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>44</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>66</td>\n",
       "      <td>CCCTCACCCACTAGGATACCAACAAACCTACCCACCC</td>\n",
       "      <td>57</td>\n",
       "      <td>67</td>\n",
       "      <td>16193</td>\n",
       "      <td>16353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>55</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>67</td>\n",
       "      <td>CCCTTCTCGTCCCCATGGATGACCCCCC</td>\n",
       "      <td>484</td>\n",
       "      <td>57</td>\n",
       "      <td>16296</td>\n",
       "      <td>16480</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    start    end  score                                            pattern  \\\n",
       "0     295    315     64                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "1     353    387     33    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "2     458    506     55  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "3     534    571     88     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "4     957   1002     22  [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "..    ...    ...    ...                                                ...   \n",
       "62  15526  15556     83          [CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]   \n",
       "63  15624  15668     29  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   \n",
       "64  16159  16193     51      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   \n",
       "65  16260  16296     44      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   \n",
       "66  16353  16380     55               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   \n",
       "\n",
       "    G4_id                                           sequence  \\\n",
       "0       1                              CCACCAAACCCCCCCTCCCCC   \n",
       "1       2                CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC   \n",
       "2       3  CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC   \n",
       "3       4             CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC   \n",
       "4       5     CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC   \n",
       "..    ...                                                ...   \n",
       "62     63                    CCCCTTAAACACCCCTCCCCACATCAAGCCC   \n",
       "63     64      CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC   \n",
       "64     65                CATAAAAACCCAATCCACATCAAAACCCCCTCCCC   \n",
       "65     66              CCCTCACCCACTAGGATACCAACAAACCTACCCACCC   \n",
       "66     67                       CCCTTCTCGTCCCCATGGATGACCCCCC   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  up_win  down_win  \n",
       "0                  38                    484     195       353  \n",
       "1                  71                     38     315       458  \n",
       "2                  28                     71     387       534  \n",
       "3                 386                     28     506       671  \n",
       "4                 680                    386     857      1102  \n",
       "..                ...                    ...     ...       ...  \n",
       "62                 68                      8   15518     15624  \n",
       "63                491                     68   15556     15768  \n",
       "64                 67                    491   16059     16260  \n",
       "65                 57                     67   16193     16353  \n",
       "66                484                     57   16296     16480  \n",
       "\n",
       "[67 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "df4ad4cd-a529-473e-ae8b-e31bd5a69233",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4pqsconf = add_columns_15(g4pqsconf).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a2badc80-01bb-4e13-b997-b94d4cad629f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "      <th>15_up</th>\n",
       "      <th>15_down</th>\n",
       "      <th>30_up</th>\n",
       "      <th>30_down</th>\n",
       "      <th>45_up</th>\n",
       "      <th>45_down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>295</td>\n",
       "      <td>315</td>\n",
       "      <td>64</td>\n",
       "      <td>[CC(A)CC]AAA[CCCC]C[CCtC][CCCC]</td>\n",
       "      <td>1</td>\n",
       "      <td>CCACCAAACCCCCCCTCCCCC</td>\n",
       "      <td>38</td>\n",
       "      <td>484</td>\n",
       "      <td>195</td>\n",
       "      <td>353</td>\n",
       "      <td>280.0</td>\n",
       "      <td>330.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>345.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>353</td>\n",
       "      <td>387</td>\n",
       "      <td>33</td>\n",
       "      <td>[CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]</td>\n",
       "      <td>2</td>\n",
       "      <td>CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC</td>\n",
       "      <td>71</td>\n",
       "      <td>38</td>\n",
       "      <td>315</td>\n",
       "      <td>458</td>\n",
       "      <td>338.0</td>\n",
       "      <td>402.0</td>\n",
       "      <td>323.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>432.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>458</td>\n",
       "      <td>506</td>\n",
       "      <td>55</td>\n",
       "      <td>[CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...</td>\n",
       "      <td>3</td>\n",
       "      <td>CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC</td>\n",
       "      <td>28</td>\n",
       "      <td>71</td>\n",
       "      <td>387</td>\n",
       "      <td>534</td>\n",
       "      <td>443.0</td>\n",
       "      <td>521.0</td>\n",
       "      <td>428.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>534</td>\n",
       "      <td>571</td>\n",
       "      <td>88</td>\n",
       "      <td>[CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC</td>\n",
       "      <td>386</td>\n",
       "      <td>28</td>\n",
       "      <td>506</td>\n",
       "      <td>671</td>\n",
       "      <td>519.0</td>\n",
       "      <td>586.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>601.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>616.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>957</td>\n",
       "      <td>1002</td>\n",
       "      <td>22</td>\n",
       "      <td>[CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...</td>\n",
       "      <td>5</td>\n",
       "      <td>CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC</td>\n",
       "      <td>680</td>\n",
       "      <td>386</td>\n",
       "      <td>857</td>\n",
       "      <td>1102</td>\n",
       "      <td>942.0</td>\n",
       "      <td>1017.0</td>\n",
       "      <td>927.0</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>912.0</td>\n",
       "      <td>1047.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>15526</td>\n",
       "      <td>15556</td>\n",
       "      <td>83</td>\n",
       "      <td>[CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]</td>\n",
       "      <td>63</td>\n",
       "      <td>CCCCTTAAACACCCCTCCCCACATCAAGCCC</td>\n",
       "      <td>68</td>\n",
       "      <td>8</td>\n",
       "      <td>15518</td>\n",
       "      <td>15624</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15571.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15586.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15601.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>15624</td>\n",
       "      <td>15668</td>\n",
       "      <td>29</td>\n",
       "      <td>[CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...</td>\n",
       "      <td>64</td>\n",
       "      <td>CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC</td>\n",
       "      <td>491</td>\n",
       "      <td>68</td>\n",
       "      <td>15556</td>\n",
       "      <td>15768</td>\n",
       "      <td>15609.0</td>\n",
       "      <td>15683.0</td>\n",
       "      <td>15594.0</td>\n",
       "      <td>15698.0</td>\n",
       "      <td>15579.0</td>\n",
       "      <td>15713.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>16159</td>\n",
       "      <td>16193</td>\n",
       "      <td>51</td>\n",
       "      <td>[C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]</td>\n",
       "      <td>65</td>\n",
       "      <td>CATAAAAACCCAATCCACATCAAAACCCCCTCCCC</td>\n",
       "      <td>67</td>\n",
       "      <td>491</td>\n",
       "      <td>16059</td>\n",
       "      <td>16260</td>\n",
       "      <td>16144.0</td>\n",
       "      <td>16208.0</td>\n",
       "      <td>16129.0</td>\n",
       "      <td>16223.0</td>\n",
       "      <td>16114.0</td>\n",
       "      <td>16238.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>16260</td>\n",
       "      <td>16296</td>\n",
       "      <td>44</td>\n",
       "      <td>[CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]</td>\n",
       "      <td>66</td>\n",
       "      <td>CCCTCACCCACTAGGATACCAACAAACCTACCCACCC</td>\n",
       "      <td>57</td>\n",
       "      <td>67</td>\n",
       "      <td>16193</td>\n",
       "      <td>16353</td>\n",
       "      <td>16245.0</td>\n",
       "      <td>16311.0</td>\n",
       "      <td>16230.0</td>\n",
       "      <td>16326.0</td>\n",
       "      <td>16215.0</td>\n",
       "      <td>16341.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>16353</td>\n",
       "      <td>16380</td>\n",
       "      <td>55</td>\n",
       "      <td>[CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]</td>\n",
       "      <td>67</td>\n",
       "      <td>CCCTTCTCGTCCCCATGGATGACCCCCC</td>\n",
       "      <td>484</td>\n",
       "      <td>57</td>\n",
       "      <td>16296</td>\n",
       "      <td>16480</td>\n",
       "      <td>16338.0</td>\n",
       "      <td>16395.0</td>\n",
       "      <td>16323.0</td>\n",
       "      <td>16410.0</td>\n",
       "      <td>16308.0</td>\n",
       "      <td>16425.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    start    end  score                                            pattern  \\\n",
       "0     295    315     64                    [CC(A)CC]AAA[CCCC]C[CCtC][CCCC]   \n",
       "1     353    387     33    [CCCC]AAAAA[C(AAAGAA)CCC]TAA[CaCC]AG[CC(TAA)CC]   \n",
       "2     458    506     55  [CC(T)CCC]A[CtCCC]ATACTACTAATCTCATCAATACAA[CCC...   \n",
       "3     534    571     88     [CCCC]ATA[CCCC]GAACCAACCAAA[CCCC]AAAGACA[CCCC]   \n",
       "4     957   1002     22  [CCCC]T[CCCC]AATAAAGCTAAAACT[CaCC]TGAGTTGTAAAA...   \n",
       "..    ...    ...    ...                                                ...   \n",
       "62  15526  15556     83          [CCCC]TTAAACA[CCCC]T[CCCC]A[C(ATCAAG)CCC]   \n",
       "63  15624  15668     29  [CCC(TATTA)C]TATCCAT[CC(TCAT)CC]TAGCAATAAT[CCC...   \n",
       "64  16159  16193     51      [C(ATAAAAA)CCC]AAT[CCaC]ATCAAAAC[CCCC]T[CCCC]   \n",
       "65  16260  16296     44      [CCC]TCA[CCC]ACTAGGATACCAACAAACCTA[CCC]A[CCC]   \n",
       "66  16353  16380     55               [CCC]TTCTCGTC[CCC]ATGGATGA[CCC][CCC]   \n",
       "\n",
       "    G4_id                                           sequence  \\\n",
       "0       1                              CCACCAAACCCCCCCTCCCCC   \n",
       "1       2                CCCCAAAAACAAAGAACCCTAACACCAGCCTAACC   \n",
       "2       3  CCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCC   \n",
       "3       4             CCCCATACCCCGAACCAACCAAACCCCAAAGACACCCC   \n",
       "4       5     CCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCC   \n",
       "..    ...                                                ...   \n",
       "62     63                    CCCCTTAAACACCCCTCCCCACATCAAGCCC   \n",
       "63     64      CCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCC   \n",
       "64     65                CATAAAAACCCAATCCACATCAAAACCCCCTCCCC   \n",
       "65     66              CCCTCACCCACTAGGATACCAACAAACCTACCCACCC   \n",
       "66     67                       CCCTTCTCGTCCCCATGGATGACCCCCC   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  up_win  down_win    15_up  \\\n",
       "0                  38                    484     195       353    280.0   \n",
       "1                  71                     38     315       458    338.0   \n",
       "2                  28                     71     387       534    443.0   \n",
       "3                 386                     28     506       671    519.0   \n",
       "4                 680                    386     857      1102    942.0   \n",
       "..                ...                    ...     ...       ...      ...   \n",
       "62                 68                      8   15518     15624      0.0   \n",
       "63                491                     68   15556     15768  15609.0   \n",
       "64                 67                    491   16059     16260  16144.0   \n",
       "65                 57                     67   16193     16353  16245.0   \n",
       "66                484                     57   16296     16480  16338.0   \n",
       "\n",
       "    15_down    30_up  30_down    45_up  45_down  \n",
       "0     330.0    265.0    345.0    250.0      0.0  \n",
       "1     402.0    323.0    417.0      0.0    432.0  \n",
       "2     521.0    428.0      0.0    413.0      0.0  \n",
       "3     586.0      0.0    601.0      0.0    616.0  \n",
       "4    1017.0    927.0   1032.0    912.0   1047.0  \n",
       "..      ...      ...      ...      ...      ...  \n",
       "62  15571.0      0.0  15586.0      0.0  15601.0  \n",
       "63  15683.0  15594.0  15698.0  15579.0  15713.0  \n",
       "64  16208.0  16129.0  16223.0  16114.0  16238.0  \n",
       "65  16311.0  16230.0  16326.0  16215.0  16341.0  \n",
       "66  16395.0  16323.0  16410.0  16308.0  16425.0  \n",
       "\n",
       "[67 rows x 16 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4pqsconf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "422e5626-3476-4383-82ee-da2291deca39",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_15_col(df:pd.DataFrame, column_15:str, start_or_end:str, up=True):\n",
    "    data_up = []\n",
    "    for up_15, start, G4_id in zip(df[column_15], df[start_or_end], df['G4_id']):\n",
    "        if up_15 > 0:\n",
    "            row = []\n",
    "            seq = ''\n",
    "            if up == True:\n",
    "                window = [*range(int(up_15), int(start))]\n",
    "            elif up == False:\n",
    "                window = [*range(int(start), int(up_15))]\n",
    "            \n",
    "            for pos in window:\n",
    "                if pos != 3107:\n",
    "                    seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "                else:\n",
    "                    seq += ''\n",
    "            row.append(G4_id)\n",
    "            row.append(column_15)\n",
    "            if up == True:\n",
    "                row.append(int(up_15))\n",
    "                row.append(int(start))\n",
    "            elif up == False:    \n",
    "                row.append(int(start))\n",
    "                row.append(int(up_15))\n",
    "            row.append(seq)\n",
    "    \n",
    "            data_up.append(row)\n",
    "        else:\n",
    "            continue\n",
    "    return data_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "223d73a8-518a-44f3-ae7e-bc130bebfb3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_15_col = make_15_col(g4pqsconf, '15_up', 'start', up=True) + \\\n",
    "make_15_col(g4pqsconf, '30_up', '15_up', up=True) + \\\n",
    "make_15_col(g4pqsconf, '45_up', '30_up', up=True) + \\\n",
    "make_15_col(g4pqsconf, '15_down', 'end', up=False) + \\\n",
    "make_15_col(g4pqsconf, '30_down', '15_down', up=False) + \\\n",
    "make_15_col(g4pqsconf, '45_down', '30_down', up=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "cc650887-b1ec-4916-8d89-deb6243666d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_15_col = pd.DataFrame(data=data_15_col, columns = ['G4_id', 'side', 'window_start', 'window_end', 'sequence'])\n",
    "nuc_count_15 = data_15_col['sequence'].apply(lambda x: Counter(x))\n",
    "nuc_count_15 = pd.DataFrame.from_records(nuc_count_15).sort_index(axis=1)\n",
    "data_15_col = data_15_col.join(nuc_count_15).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f6c82c20-13f2-443d-9376-4379a3702179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>15_up</td>\n",
       "      <td>280</td>\n",
       "      <td>295</td>\n",
       "      <td>CATAACAAAAAATTT</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>15_up</td>\n",
       "      <td>338</td>\n",
       "      <td>353</td>\n",
       "      <td>CACATCTCTGCCAAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>15_up</td>\n",
       "      <td>443</td>\n",
       "      <td>458</td>\n",
       "      <td>AACACATTATTTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>15_up</td>\n",
       "      <td>519</td>\n",
       "      <td>534</td>\n",
       "      <td>ACACACCGCTGCTAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>15_up</td>\n",
       "      <td>942</td>\n",
       "      <td>957</td>\n",
       "      <td>AGTGTTTTAGATCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>63</td>\n",
       "      <td>45_down</td>\n",
       "      <td>15586</td>\n",
       "      <td>15601</td>\n",
       "      <td>TCTCCGATCCGTCCC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>64</td>\n",
       "      <td>45_down</td>\n",
       "      <td>15698</td>\n",
       "      <td>15713</td>\n",
       "      <td>CGCCCACTAAGCCAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>65</td>\n",
       "      <td>45_down</td>\n",
       "      <td>16223</td>\n",
       "      <td>16238</td>\n",
       "      <td>CTCAACTATCACACA</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>66</td>\n",
       "      <td>45_down</td>\n",
       "      <td>16326</td>\n",
       "      <td>16341</td>\n",
       "      <td>ACCGTACATAGCACA</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>67</td>\n",
       "      <td>45_down</td>\n",
       "      <td>16410</td>\n",
       "      <td>16425</td>\n",
       "      <td>CCGTGAAATCAATAT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>358 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     G4_id     side  window_start  window_end         sequence    A    C    G  \\\n",
       "0        1    15_up           280         295  CATAACAAAAAATTT  9.0  2.0  0.0   \n",
       "1        2    15_up           338         353  CACATCTCTGCCAAA  5.0  6.0  1.0   \n",
       "2        3    15_up           443         458  AACACATTATTTTCC  5.0  4.0  0.0   \n",
       "3        4    15_up           519         534  ACACACCGCTGCTAA  5.0  6.0  2.0   \n",
       "4        5    15_up           942         957  AGTGTTTTAGATCAC  4.0  2.0  3.0   \n",
       "..     ...      ...           ...         ...              ...  ...  ...  ...   \n",
       "353     63  45_down         15586       15601  TCTCCGATCCGTCCC  1.0  8.0  2.0   \n",
       "354     64  45_down         15698       15713  CGCCCACTAAGCCAA  5.0  7.0  2.0   \n",
       "355     65  45_down         16223       16238  CTCAACTATCACACA  6.0  6.0  0.0   \n",
       "356     66  45_down         16326       16341  ACCGTACATAGCACA  6.0  5.0  2.0   \n",
       "357     67  45_down         16410       16425  CCGTGAAATCAATAT  6.0  3.0  2.0   \n",
       "\n",
       "     T  \n",
       "0    4  \n",
       "1    3  \n",
       "2    6  \n",
       "3    2  \n",
       "4    6  \n",
       "..  ..  \n",
       "353  4  \n",
       "354  1  \n",
       "355  3  \n",
       "356  2  \n",
       "357  4  \n",
       "\n",
       "[358 rows x 9 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_15_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4610cac3-2e2f-44d4-adcf-10ad9b4abfe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_15_col.to_csv('../data/processed/data_15_col_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13fd773b-32d6-4c6e-bda3-45af15b750f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_15_col = pd.read_csv('../data/processed/data_15_col_pqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a566f7e1-61b0-46e8-a1d9-c821a286b817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>15_up</td>\n",
       "      <td>338</td>\n",
       "      <td>353</td>\n",
       "      <td>CACATCTCTGCCAAA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>2</td>\n",
       "      <td>30_up</td>\n",
       "      <td>323</td>\n",
       "      <td>338</td>\n",
       "      <td>GCCACAGCACTTAAA</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>2</td>\n",
       "      <td>15_down</td>\n",
       "      <td>387</td>\n",
       "      <td>402</td>\n",
       "      <td>CAGATTTCAAATTTT</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>2</td>\n",
       "      <td>30_down</td>\n",
       "      <td>402</td>\n",
       "      <td>417</td>\n",
       "      <td>ATCTTTTGGCGGTAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>2</td>\n",
       "      <td>45_down</td>\n",
       "      <td>417</td>\n",
       "      <td>432</td>\n",
       "      <td>GCACTTTTAACAGTC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     G4_id     side  window_start  window_end         sequence    A    C    G  \\\n",
       "1        2    15_up           338         353  CACATCTCTGCCAAA  5.0  6.0  1.0   \n",
       "66       2    30_up           323         338  GCCACAGCACTTAAA  6.0  5.0  2.0   \n",
       "180      2  15_down           387         402  CAGATTTCAAATTTT  5.0  2.0  1.0   \n",
       "245      2  30_down           402         417  ATCTTTTGGCGGTAT  2.0  2.0  4.0   \n",
       "303      2  45_down           417         432  GCACTTTTAACAGTC  4.0  4.0  2.0   \n",
       "\n",
       "     T  \n",
       "1    3  \n",
       "66   2  \n",
       "180  7  \n",
       "245  7  \n",
       "303  5  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_15_col[data_15_col['G4_id'] == 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "154b760d-666f-4ff7-be60-c93ff4d7513d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step 3  \n",
    "make windows for sliding window around G4 (make df_win)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce234396-3f83-42fd-8860-c88ca99caea4",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### OLD  \n",
    "\n",
    "New interval size:\n",
    "- Take all confirmed G4   \n",
    "- **Dont sort G4 by size between G4 >= 50 bp**  \n",
    "- Obtain different interval size (0-100bp) between G4  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e084d609-6f9e-4e99-9514-354e2f751a31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# G4_id = 38 hasnt downstream windows\n",
    "# G4_id = 39 hasnt upsteram windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d3c32786-4097-4917-bc67-17455f8e7e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4conf[(g4conf['G4_id'] == 38) | (g4conf['G4_id'] == 39)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "de0772f5-ddb9-44b5-9de2-bd4a0229c7a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>91</td>\n",
       "      <td>185</td>\n",
       "      <td>194</td>\n",
       "      <td>GCGAACATAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>90</td>\n",
       "      <td>186</td>\n",
       "      <td>195</td>\n",
       "      <td>CGAACATACT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>89</td>\n",
       "      <td>187</td>\n",
       "      <td>196</td>\n",
       "      <td>GAACATACTT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>88</td>\n",
       "      <td>188</td>\n",
       "      <td>197</td>\n",
       "      <td>AACATACTTA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>87</td>\n",
       "      <td>189</td>\n",
       "      <td>198</td>\n",
       "      <td>ACATACTTAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id side  number_of_win  window_start  window_end    sequence    A    C  \\\n",
       "0      1   up             91           185         194  GCGAACATAC  4.0  3.0   \n",
       "1      1   up             90           186         195  CGAACATACT  4.0  3.0   \n",
       "2      1   up             89           187         196  GAACATACTT  4.0  2.0   \n",
       "3      1   up             88           188         197  AACATACTTA  5.0  2.0   \n",
       "4      1   up             87           189         198  ACATACTTAC  4.0  3.0   \n",
       "\n",
       "     G    T  \n",
       "0  2.0  1.0  \n",
       "1  1.0  2.0  \n",
       "2  1.0  3.0  \n",
       "3  0.0  3.0  \n",
       "4  0.0  3.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win = make_win_df(make_win_up(g4conf), make_win_down(g4conf))\n",
    "df_win.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b5b63c49-b04d-42f0-803f-16598e995f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win.to_csv('../data/processed/df_win.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78f5f2a-7b79-4043-aa17-e44b5ac0accf",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### NEW pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "16545ac6-9c5f-48cd-8604-0e599d201848",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_pqs = make_win_df(make_win_up(g4pqsconf), make_win_down(g4pqsconf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "24f5a999-ac7e-48df-a16c-17e5552db03b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>91</td>\n",
       "      <td>195</td>\n",
       "      <td>204</td>\n",
       "      <td>TTACTAAAGT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>90</td>\n",
       "      <td>196</td>\n",
       "      <td>205</td>\n",
       "      <td>TACTAAAGTG</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>89</td>\n",
       "      <td>197</td>\n",
       "      <td>206</td>\n",
       "      <td>ACTAAAGTGT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>88</td>\n",
       "      <td>198</td>\n",
       "      <td>207</td>\n",
       "      <td>CTAAAGTGTG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>up</td>\n",
       "      <td>87</td>\n",
       "      <td>199</td>\n",
       "      <td>208</td>\n",
       "      <td>TAAAGTGTGT</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id side  number_of_win  window_start  window_end    sequence    A    C  \\\n",
       "0      1   up             91           195         204  TTACTAAAGT  4.0  1.0   \n",
       "1      1   up             90           196         205  TACTAAAGTG  4.0  1.0   \n",
       "2      1   up             89           197         206  ACTAAAGTGT  4.0  1.0   \n",
       "3      1   up             88           198         207  CTAAAGTGTG  3.0  1.0   \n",
       "4      1   up             87           199         208  TAAAGTGTGT  3.0  0.0   \n",
       "\n",
       "     G    T  \n",
       "0  1.0  4.0  \n",
       "1  2.0  3.0  \n",
       "2  2.0  3.0  \n",
       "3  3.0  3.0  \n",
       "4  3.0  4.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_pqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0cba2eed-e30c-4763-b6c5-0853eb07d951",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_pqs.to_csv('../data/processed/df_win_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afef4e00-7932-4f3a-a496-c3b6e0cc77b9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation around G4 for Cancer "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b1ad36c-f441-43cc-85cf-e7e8754b63d6",
   "metadata": {},
   "source": [
    "### Observed mutations Cancer dataset:\n",
    "\n",
    "df_win_cancer - observed mutations from Cancer dataset with nucleotide frequencies from df_win\n",
    "  \n",
    "  Yuan, Y., Ju, Y.S., Kim, Y. et al. Comprehensive molecular characterization of mitochondrial genomes in human cancers. Nat Genet 52, 342–352 (2020). https://doi.org/10.1038/s41588-019-0557-x\n",
    "\n",
    "Add mutations from Cancer dataset:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fcb50c9-f2ae-492d-beb5-ec2b983bda6b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### OLD G4Hunter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "935484e8-29c6-4e38-a802-84153feca618",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win = pd.read_csv('../data/processed/df_win.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5a8fb746-4231-4e96-a4a8-125e55d1f093",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer = collect_mut_cancer(df_cancers, df_win)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "21fe33c6-e94b-4b43-878e-8823fbb60def",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer.to_csv('../data/processed/df_win_cancer.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f8f41b0-6084-4d67-80a4-933e5665fef3",
   "metadata": {},
   "source": [
    "### NEW pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "68e12c2b-a5d9-4343-9b7e-aea5b49ed961",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win_pqs = pd.read_csv('../data/processed/df_win_pqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b840074d-932a-400a-a0d6-755eac7772b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_pqs = collect_mut_cancer(df_cancers, df_win_pqs, regime='2col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "410a2482-c240-4d71-8aa6-1535256c3584",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_pqs.to_csv('../data/processed/df_win_cancer_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e72b704d-e1ba-444f-97fd-4155cc97d104",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Data preparation around G4 for Human Global Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5857365f-77a5-444f-ae4d-c4d77dbc5d0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt = pd.read_table('../data/raw/fulltreeCodons.csv', sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c7883c2f-3160-4909-aed1-05286d08bc5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/raw/hgt_alignment.blast') as fin:\n",
    "    while True:\n",
    "        line = fin.readline()\n",
    "        if 'Strand=Plus/Plus' in line:\n",
    "            fin.readline()\n",
    "            break\n",
    "    qseq = ''\n",
    "    rods = ''\n",
    "    sseq = ''\n",
    "    query, sbjct = 'Query', 'Sbjct'\n",
    "    i = 0\n",
    "    size = 60\n",
    "    start = 14\n",
    "    while query.startswith('Query') and sbjct.startswith('Sbjct'):\n",
    "        i += 1\n",
    "        query = fin.readline().strip()\n",
    "        cur_rods = fin.readline()\n",
    "        sbjct = fin.readline().strip()\n",
    "\n",
    "        if len(query[start:start+size]) == size:\n",
    "            qseq += query[start:start+size]\n",
    "            sseq += sbjct[start:start+size]\n",
    "            rods += cur_rods[start:start+size]\n",
    "        else:\n",
    "            pos = re.search('Query\\s{2}(\\d+)\\s+[\\w-]+\\s+(\\d+)', query).groups()\n",
    "            size = int(pos[1]) - int(pos[0]) + 1\n",
    "            \n",
    "            qseq += query[start:start+size]\n",
    "            sseq += sbjct[start:start+size]\n",
    "            rods += cur_rods[start:start+size]\n",
    "            break\n",
    "            \n",
    "        fin.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "95af0cf5-e3f2-4ac7-85bd-be659f34fd79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sseq = 16568 qseq = 16568\n"
     ]
    }
   ],
   "source": [
    "print('sseq =', len(sseq), 'qseq =', len(qseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "06f4b065-c8ea-48e2-b49e-2312f36be982",
   "metadata": {},
   "outputs": [],
   "source": [
    "sseq = 'G' + sseq\n",
    "qseq = 'G' + qseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3b7d7077-ca3c-4c2c-b583-276373dd4d9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sseq = 16569 qseq = 16569\n"
     ]
    }
   ],
   "source": [
    "print('sseq =', len(sseq), 'qseq =', len(qseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a8b7ee0c-f238-4d38-8e77-58d16cf2958f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_hgt = {i: qseq[i] for i in range(len(qseq))}\n",
    "dict_ref = {i: sseq[i] for i in range(len(sseq))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "05e513a6-65b2-49df-8906-51fd00ce3dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = pd.DataFrame.from_dict(dict_ref, orient='index', columns=['nuc_ref']) \\\n",
    "    .reset_index() \\\n",
    "    .rename(columns={'index':'Pos'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4113e24e-739f-4efb-a05b-58d1b6af7953",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref['nuc_hgt'] = df_ali_hgt_to_ref['Pos'].map(dict_hgt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3ecc70dc-76b2-410f-9340-e1c726c71ab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref['Pos'] = range(1, len(df_ali_hgt_to_ref['Pos'])+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "fd09aabc-b2fa-464e-8d39-2c9d2fabcbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = df_ali_hgt_to_ref.replace({'c':'C', 'n':'N', 't':'T', 'g':'G', 'a':'A'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133a84d6-868b-422b-bb73-92a65f903216",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique = df_hgt[['ref_pos', 'nuc_ref_in_ali']].sort_values(by='ref_pos').drop_duplicates(subset=['ref_pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5b2678c0-a075-4664-bb83-ce3542c096ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique = df_unique.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f1dae751-7d37-4ee4-9c1d-07ed2f003e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique_sort = df_unique[~((df_unique['nuc_ref_in_ali'] == '-') | (df_unique['nuc_ref_in_ali'] == 'N'))][['ref_pos', 'nuc_ref_in_ali']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "884975a1-7706-4217-8bab-de6f5baf659a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_unique_sort = df_unique_sort.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "592b006a-8361-4657-997c-0d91b5fa2ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Detele all posirions with N, -, and first position\n",
    "a = df_ali_hgt_to_ref[~((df_ali_hgt_to_ref['nuc_hgt'] == 'N') | \n",
    "                        (df_ali_hgt_to_ref['nuc_hgt'] == '-') | \n",
    "                        (df_ali_hgt_to_ref['Pos'] == 1))].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "be87253e-954f-4475-a221-07be82cb4a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos_ref_to_hgt = pd.concat([a, df_unique_sort], axis=1).rename(columns={'ref_pos':'hgt_ref_pos'}).set_index('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "972e8755-03f2-44b4-adb2-12309cd25b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pos_ref_to_hgt.index.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7e9e6eac-96f0-4624-902e-e108e2c3a163",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref = df_ali_hgt_to_ref.join(df_pos_ref_to_hgt, how='left', rsuffix='2') \\\n",
    "    .drop(columns=['Pos2', 'nuc_ref2', 'nuc_hgt2', 'nuc_hgt']) \\\n",
    "    .astype({'hgt_ref_pos':'Int64'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0b77155c-c82c-4d30-8b00-89795bc6100f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add N in table\n",
    "df_ali_hgt_to_ref.iloc[3106, df_ali_hgt_to_ref.columns.get_loc('hgt_ref_pos')] = 3106\n",
    "df_ali_hgt_to_ref.iloc[3106, df_ali_hgt_to_ref.columns.get_loc('nuc_ref_in_ali')] = 'N'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "71fff04b-9d3a-48e0-8421-bb94aca06688",
   "metadata": {},
   "outputs": [],
   "source": [
    "for start, end, G4_id in zip(g4conf['start'], g4conf['end'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(start, end+1)]),'G4'] = 1 #is position G4? 1-yes, 0-no\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(start, end+1)]),'G4_id'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c9df48fd-023f-4672-a35c-4a9e9f9e9342",
   "metadata": {},
   "outputs": [],
   "source": [
    "for up_win, start, G4_id in zip(g4conf['up_win'], g4conf['start'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(up_win, start+1)]),'up_win'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "f6abbe4e-ed75-4093-8fa2-c2ac748d418e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for end, down_win, G4_id in zip(g4conf['end'], g4conf['down_win'], g4conf['G4_id']):\n",
    "    df_ali_hgt_to_ref.loc[df_ali_hgt_to_ref.Pos.isin([*range(end, down_win+1)]),'down_win'] = G4_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0ae3a737-8c80-4d9a-a351-3ac0816c3108",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ali_hgt_to_ref[['G4', 'G4_id', 'up_win', 'down_win']] = df_ali_hgt_to_ref[['G4',  'G4_id', \n",
    "                                                                              'up_win', 'down_win']].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "30cff4ed-0b9b-4c8d-ac76-79c87b33ed96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>Pos</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>G4</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref_pos  Pos  G4_id   G4  up_win  down_win\n",
       "0     <NA>    1    0.0  0.0     0.0       0.0\n",
       "1        1    2    0.0  0.0     0.0       0.0\n",
       "2        2    3    0.0  0.0     0.0       0.0\n",
       "3        3    4    0.0  0.0     0.0       0.0\n",
       "4        4    5    0.0  0.0     0.0       0.0"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ali_hgt_to_ref[['hgt_ref_pos', 'Pos', 'G4_id', 'G4', 'up_win', 'down_win']] \\\n",
    "    .rename(columns={'hgt_ref_pos':'ref_pos'}).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "19e55ecf-5372-4392-82a3-2889866402c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt_g4 = df_hgt \\\n",
    "    .merge(df_ali_hgt_to_ref[['hgt_ref_pos', 'Pos', 'G4_id', 'G4', 'up_win', 'down_win']] \\\n",
    "           .rename(columns={'hgt_ref_pos':'ref_pos'}), how='left') \\\n",
    "    .sort_values(by='ref_pos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b38c2f91-81f6-4339-9ec5-2f58bc3d65c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hgt_g4['Mut'] = df_hgt_g4.ancestor.str[2] + '>' + df_hgt_g4.descendant.str[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b56cd6d7-8493-4db9-8175-8a99fbc53970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_hgt_g4.to_csv('data/processed/df_hgt_g4.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7078b016-cbd2-4057-b122-5db717a75411",
   "metadata": {},
   "source": [
    "### Collect observed mutations from HGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d1e14f14-0750-4d91-92c8-0b80449d7504",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_13856\\3731285902.py:1: DtypeWarning: Columns (10,11,12,14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')\n"
     ]
    }
   ],
   "source": [
    "df_hgt_g4 = pd.read_csv('data/processed/df_hgt_g4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e61e6ff7-e62d-4f33-b59c-2cb67a86a7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract mutations without gaps (-) and other (R, M, N, ect) bases in mut position:\n",
    "df_hgt_g4 = df_hgt_g4[\n",
    "    (df_hgt_g4.ancestor.str.lower().str.fullmatch(\"[acgt]{5}\")) &\n",
    "    (df_hgt_g4.descendant.str.lower().str.fullmatch(\"[acgt]{5}\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "cf42ee54-8933-4a17-9547-c6846dfff594",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_up = []\n",
    "for up_win, Pos, Mut in zip(df_hgt_g4['up_win'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if up_win > 0:\n",
    "        row.append(up_win)\n",
    "        row.append('up')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_up.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "63a09fcb-4f8c-422c-9dff-12bfeffd72c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_down = []\n",
    "for down_win, Pos, Mut in zip(df_hgt_g4['down_win'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if down_win > 0:\n",
    "        row.append(down_win)\n",
    "        row.append('down')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_down.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8e4fe658-50cc-4fc9-b12b-f1b568b975db",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt = obs_mut_hgt_up + obs_mut_hgt_down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "40aba95a-b118-4632-ba7b-fe5de90f037e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_obs_mut_hgt = pd.DataFrame(data=obs_mut_hgt, columns=['G4_id', 'side', 'Pos', 'Mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "7c2726ca-0560-45ed-8bd3-3bd17c29e9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win = pd.read_csv('data/processed/df_win.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4a788a6f-9920-4132-811e-8cbb0957962a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_hgt_win = []\n",
    "for window_start, window_end in zip(df_win['window_start'], df_win['window_end']):\n",
    "    row = []\n",
    "    \n",
    "    mut = df_obs_mut_hgt[(df_obs_mut_hgt['Pos'] >= window_start) & (df_obs_mut_hgt['Pos'] <= window_end)].Mut.tolist()\n",
    "    \n",
    "    row.append(mut)\n",
    "    \n",
    "    Mut_hgt_win.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "85104fc9-f0fb-473a-bffb-b7afa2cbd888",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Mut_hgt_win = pd.DataFrame(data=Mut_hgt_win)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "84f9c9e7-fde8-4607-8e07-458dcf2d94f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutG4_hgt = pd.DataFrame.from_records(df_Mut_hgt_win[0].apply(lambda x: Counter(x))).fillna(0).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "f52fd3a1-ac53-46ea-9248-17057f847d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt = df_win.join(obsMutG4_hgt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3d5fcbb3-b07e-43a3-988b-2ef6b6db3dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt['number_of_Mut'] = df_win_hgt[['A>C','A>G','A>T','C>A','C>G','C>T','G>A','G>C','G>T','T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "2dbcc008-a0f0-47b4-b3f4-a482361d577e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_hgt.to_csv('data/processed/df_win_hgt.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409edb6e-a55c-4d47-aebd-61fad4ab0158",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# WITHIN G4 WITH SIDES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "690727aa-9531-4e4e-9fe9-384cbc5d3b01",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Data preparation within G4 with sides"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0e438f3-4909-40da-ba58-4b0f61440628",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### OLD G4Hunter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "60592576-eb4b-4bca-8cd0-be06b7c261d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4conf = pd.read_csv('../data/processed/g4conf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8534fe95-dbbe-4b9a-95b6-e9ad4a520632",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within_sides = make_win_df(make_win_start_within(g4conf), make_win_end_within(g4conf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f3d6801d-f02a-4f40-a5bc-e69f51154b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_within_sides.to_csv('../data/processed/df_win_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62aa7de7-9d49-4e51-b89e-6d4e8f718e4f",
   "metadata": {},
   "source": [
    "### NEW pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "07f49c7d-215a-40ea-8f64-61f8b3c21eb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_within_sides_pqs = make_win_df(make_win_start_within(g4pqsconf), make_win_end_within(g4pqsconf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "18915d8e-1f1a-441a-b85e-10abe6dac77c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_within_sides_pqs.to_csv('../data/processed/df_win_within_sides_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8201f772-c292-4b89-af76-2695b79b3dee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Within sides Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3750bed1-1ccb-4912-bb12-811cc7d16ecf",
   "metadata": {},
   "source": [
    "### OLD G4Hunter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1e19fc2d-fb5a-4997-b490-fe4e0e305d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win_within_sides = pd.read_csv('../data/processed/df_win_within_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3a315197-9b0c-45f6-81ea-f1431d7c3586",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within_sides = collect_mut_cancer(df_cancers, df_win_within_sides, regime='2col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "9eadd5ed-cea7-469d-851e-2ad4e081f2ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_within_sides.to_csv('../data/processed/df_win_cancer_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f087e68-cb9c-4bb2-ac03-7236c836b0de",
   "metadata": {},
   "source": [
    "### NEW pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "ebfdfd58-487b-4e8d-86c0-2012ab38fe0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win_within_sides_pqs = pd.read_csv('../data/processed/df_win_within_sides_pqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "29e2eaa2-eb41-4fcb-8e2d-abd44aaa16c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within_sides_pqs = collect_mut_cancer(df_cancers, df_win_within_sides_pqs, regime='2col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "59edd1b9-f044-4959-ba00-96adbe3284e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_within_sides_pqs.to_csv('../data/processed/df_win_cancer_within_sides_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "d053e418-eeee-4770-ba91-7a605bc5f829",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')\n",
    "df_win_within_sides_pqs = pd.read_csv('../data/processed/df_win_within_sides_pqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "11ed8ae9-3717-4ae7-b362-66cc8d88fdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_within_sides_pqs = collect_mut_cancer(df_cancers, df_win_within_sides_pqs, regime='2col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "693567b3-6a41-4015-8036-335c614b075b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>1</td>\n",
       "      <td>295</td>\n",
       "      <td>304</td>\n",
       "      <td>CCACCAAACC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>2</td>\n",
       "      <td>296</td>\n",
       "      <td>305</td>\n",
       "      <td>CACCAAACCC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>3</td>\n",
       "      <td>297</td>\n",
       "      <td>306</td>\n",
       "      <td>ACCAAACCCC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>4</td>\n",
       "      <td>298</td>\n",
       "      <td>307</td>\n",
       "      <td>CCAAACCCCC</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>5</td>\n",
       "      <td>299</td>\n",
       "      <td>308</td>\n",
       "      <td>CAAACCCCCC</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3434</th>\n",
       "      <td>67</td>\n",
       "      <td>within_end</td>\n",
       "      <td>13</td>\n",
       "      <td>16358</td>\n",
       "      <td>16367</td>\n",
       "      <td>TCGTCCCCAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3435</th>\n",
       "      <td>67</td>\n",
       "      <td>within_end</td>\n",
       "      <td>14</td>\n",
       "      <td>16357</td>\n",
       "      <td>16366</td>\n",
       "      <td>CTCGTCCCCA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436</th>\n",
       "      <td>67</td>\n",
       "      <td>within_end</td>\n",
       "      <td>15</td>\n",
       "      <td>16356</td>\n",
       "      <td>16365</td>\n",
       "      <td>TCTCGTCCCC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3437</th>\n",
       "      <td>67</td>\n",
       "      <td>within_end</td>\n",
       "      <td>16</td>\n",
       "      <td>16355</td>\n",
       "      <td>16364</td>\n",
       "      <td>TTCTCGTCCC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3438</th>\n",
       "      <td>67</td>\n",
       "      <td>within_end</td>\n",
       "      <td>17</td>\n",
       "      <td>16354</td>\n",
       "      <td>16363</td>\n",
       "      <td>CTTCTCGTCC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3439 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id          side  number_of_win  window_start  window_end  \\\n",
       "0         1  within_start              1           295         304   \n",
       "1         1  within_start              2           296         305   \n",
       "2         1  within_start              3           297         306   \n",
       "3         1  within_start              4           298         307   \n",
       "4         1  within_start              5           299         308   \n",
       "...     ...           ...            ...           ...         ...   \n",
       "3434     67    within_end             13         16358       16367   \n",
       "3435     67    within_end             14         16357       16366   \n",
       "3436     67    within_end             15         16356       16365   \n",
       "3437     67    within_end             16         16355       16364   \n",
       "3438     67    within_end             17         16354       16363   \n",
       "\n",
       "        sequence    A    C    G    T  ...  C>A  C>G   C>T  G>A  G>C  G>T  T>A  \\\n",
       "0     CCACCAAACC  4.0  6.0  0.0  0.0  ...  0.0  0.0   4.0  0.0  0.0  0.0  0.0   \n",
       "1     CACCAAACCC  4.0  6.0  0.0  0.0  ...  0.0  0.0   4.0  0.0  0.0  0.0  0.0   \n",
       "2     ACCAAACCCC  4.0  6.0  0.0  0.0  ...  0.0  0.0   1.0  0.0  0.0  0.0  0.0   \n",
       "3     CCAAACCCCC  3.0  7.0  0.0  0.0  ...  0.0  0.0   1.0  0.0  0.0  0.0  0.0   \n",
       "4     CAAACCCCCC  3.0  7.0  0.0  0.0  ...  5.0  0.0   1.0  0.0  0.0  0.0  0.0   \n",
       "...          ...  ...  ...  ...  ...  ...  ...  ...   ...  ...  ...  ...  ...   \n",
       "3434  TCGTCCCCAT  1.0  5.0  1.0  3.0  ...  2.0  0.0  10.0  1.0  0.0  0.0  1.0   \n",
       "3435  CTCGTCCCCA  1.0  6.0  1.0  2.0  ...  1.0  0.0   9.0  1.0  0.0  0.0  1.0   \n",
       "3436  TCTCGTCCCC  0.0  6.0  1.0  3.0  ...  1.0  0.0   2.0  1.0  0.0  0.0  1.0   \n",
       "3437  TTCTCGTCCC  0.0  5.0  1.0  4.0  ...  0.0  0.0   7.0  1.0  0.0  0.0  1.0   \n",
       "3438  CTTCTCGTCC  0.0  5.0  1.0  4.0  ...  0.0  0.0   8.0  1.0  0.0  0.0  1.0   \n",
       "\n",
       "      T>C  T>G  number_of_Mut  \n",
       "0     0.0  0.0            6.0  \n",
       "1     0.0  0.0            6.0  \n",
       "2     0.0  0.0            3.0  \n",
       "3     0.0  0.0            1.0  \n",
       "4     0.0  0.0            6.0  \n",
       "...   ...  ...            ...  \n",
       "3434  1.0  0.0           15.0  \n",
       "3435  3.0  0.0           15.0  \n",
       "3436  5.0  0.0           10.0  \n",
       "3437  5.0  0.0           14.0  \n",
       "3438  5.0  0.0           15.0  \n",
       "\n",
       "[3439 rows x 23 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_cancer_within_sides_pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7e638f6d-4ef9-4da9-ab66-810e92abf1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_within_sides_pqs.to_csv('../data/processed/df_win_cancer_within_sides_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e455a9ac-bbfc-4374-b3e3-2352c7bcb33d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Within sides HGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "24c1a84c-1131-481c-b0e3-20f560764b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\777\\AppData\\Local\\Temp\\ipykernel_13260\\1406226394.py:1: DtypeWarning: Columns (10,11,12,14,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_hgt_g4 = pd.read_csv('../data/processed/df_hgt_g4.csv')\n"
     ]
    }
   ],
   "source": [
    "df_hgt_g4 = pd.read_csv('../data/processed/df_hgt_g4.csv')\n",
    "df_win_within_sides = pd.read_csv('../data/processed/df_win_within_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "bb4f24bf-54ee-4c91-8969-0fc301989c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract mutations without gaps (-) and other (R, M, N, ect) bases in mut position:\n",
    "df_hgt_g4 = df_hgt_g4[\n",
    "    (df_hgt_g4.ancestor.str.lower().str.fullmatch(\"[acgt]{5}\")) &\n",
    "    (df_hgt_g4.descendant.str.lower().str.fullmatch(\"[acgt]{5}\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f13f219f-8e59-4811-9ba8-c65fa84ee6fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_mut_hgt_within_sides = []\n",
    "for G4, G4_id, Pos, Mut in zip(df_hgt_g4['G4'], df_hgt_g4['G4_id'], df_hgt_g4['Pos'], df_hgt_g4['Mut']):\n",
    "    row = []\n",
    "    \n",
    "    if G4 > 0:\n",
    "        row.append(G4_id)\n",
    "        row.append('within')\n",
    "        row.append(Pos)\n",
    "        row.append(Mut)\n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    obs_mut_hgt_within_sides.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "96ac0880-d4ef-4c5c-8972-deca9ab19fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_obs_mut_hgt_within_sides = pd.DataFrame(data=obs_mut_hgt_within_sides, columns=['G4_id', 'side', 'Pos', 'Mut'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a238997a-ddf4-4d5a-802c-5a9b2014a3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_hgt_win_within_sides = []\n",
    "for window_start, window_end in zip(df_win_within_sides['window_start'], df_win_within_sides['window_end']):\n",
    "    row = []\n",
    "    \n",
    "    mut = df_obs_mut_hgt_within_sides[(df_obs_mut_hgt_within_sides['Pos'] >= window_start) & \n",
    "        (df_obs_mut_hgt_within_sides['Pos'] <= window_end)]['Mut'].tolist()\n",
    "\n",
    "    row.append(mut)\n",
    "    \n",
    "    Mut_hgt_win_within_sides.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "e7d2b5b1-2536-4a36-a9cd-0b03fe9c624b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Mut_hgt_win_within_sides = pd.DataFrame(data=Mut_hgt_win_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ae247255-4877-4aaa-b560-30330023a8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutG4_hgt_within_sides = pd.DataFrame.from_records(df_Mut_hgt_win_within_sides[0].apply(lambda x: Counter(x))) \\\n",
    "    .fillna(0) \\\n",
    "    .sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "82bf9307-75fc-4f0c-a8a4-e36fc38917c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within_sides = df_win_within_sides.join(obsMutG4_hgt_within_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9a17414b-3b4f-491c-bb59-1eb5be0e6a4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_hgt_within_sides['number_of_Mut'] = df_win_hgt_within_sides[['A>C','A>G','A>T',\n",
    "                                                        'C>A','C>G','C>T',\n",
    "                                                        'G>A','G>C','G>T',\n",
    "                                                        'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "0cf8821a-4985-4111-a181-7c0281d9ac08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>1</td>\n",
       "      <td>285</td>\n",
       "      <td>294</td>\n",
       "      <td>CAAAAAATTT</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>2</td>\n",
       "      <td>286</td>\n",
       "      <td>295</td>\n",
       "      <td>AAAAAATTTC</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>3</td>\n",
       "      <td>287</td>\n",
       "      <td>296</td>\n",
       "      <td>AAAAATTTCC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>4</td>\n",
       "      <td>288</td>\n",
       "      <td>297</td>\n",
       "      <td>AAAATTTCCA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>within_start</td>\n",
       "      <td>5</td>\n",
       "      <td>289</td>\n",
       "      <td>298</td>\n",
       "      <td>AAATTTCCAC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>539.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id          side  number_of_win  window_start  window_end    sequence  \\\n",
       "0      1  within_start              1           285         294  CAAAAAATTT   \n",
       "1      1  within_start              2           286         295  AAAAAATTTC   \n",
       "2      1  within_start              3           287         296  AAAAATTTCC   \n",
       "3      1  within_start              4           288         297  AAAATTTCCA   \n",
       "4      1  within_start              5           289         298  AAATTTCCAC   \n",
       "\n",
       "     A    C    G    T  ...   C>A  C>G    C>T   G>A  G>C  G>T  T>A    T>C  T>G  \\\n",
       "0  6.0  1.0  0.0  3.0  ...   0.0  0.0   76.0   5.0  0.0  0.0  8.0   88.0  0.0   \n",
       "1  6.0  1.0  0.0  3.0  ...  26.0  0.0  128.0   5.0  0.0  0.0  9.0  140.0  0.0   \n",
       "2  5.0  2.0  0.0  3.0  ...  28.0  0.0  152.0   5.0  0.0  0.0  9.0  163.0  0.0   \n",
       "3  5.0  2.0  0.0  3.0  ...  34.0  0.0  152.0  72.0  0.0  0.0  9.0  163.0  0.0   \n",
       "4  4.0  3.0  0.0  3.0  ...  35.0  0.0  152.0  67.0  0.0  0.0  9.0  163.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0          189.0  \n",
       "1          346.0  \n",
       "2          397.0  \n",
       "3          548.0  \n",
       "4          539.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_hgt_within_sides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6908ca3d-6cb5-4e63-9762-6028447d53af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_hgt_within_sides.to_csv('../data/processed/df_win_hgt_within_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a1a2d65-23cd-4c62-8b07-93344a1c2520",
   "metadata": {},
   "source": [
    "# Data preparation within category, cancer. PQS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "0a3774f0-4b2f-4ed5-b8e2-87e3bb241bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv') \n",
    "df_category = pd.read_csv('../data/processed/df_category_pqs.csv')\n",
    "data_15_col = pd.read_csv('../data/processed/data_15_col_pqs.csv')\n",
    "data_loops = pd.read_csv('../data/processed/data_loops_pqs.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "2757dea5-b660-4960-8aa3-b04dbb05bad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category_cancers = collect_mut_cancer(df_cancers, df_category, regime='pos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3523023c-73a6-4094-bed7-df6bd8f2a5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_category_cancers.to_csv('../data/processed/df_category_cancers_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "048209cc-2332-4e8a-b83b-1f7df808c7de",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Statistic for categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "99420453-3775-47e7-9ed0-513d24a77d4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>pos_in_seq</th>\n",
       "      <th>pos_in_ref</th>\n",
       "      <th>nuc</th>\n",
       "      <th>category</th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>296</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>297</td>\n",
       "      <td>A</td>\n",
       "      <td>Bulges</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>298</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>299</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418</th>\n",
       "      <td>67</td>\n",
       "      <td>23</td>\n",
       "      <td>16376</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2419</th>\n",
       "      <td>67</td>\n",
       "      <td>24</td>\n",
       "      <td>16377</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>67</td>\n",
       "      <td>25</td>\n",
       "      <td>16378</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2421</th>\n",
       "      <td>67</td>\n",
       "      <td>26</td>\n",
       "      <td>16379</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2422</th>\n",
       "      <td>67</td>\n",
       "      <td>27</td>\n",
       "      <td>16380</td>\n",
       "      <td>C</td>\n",
       "      <td>Tetrad</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2423 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id  pos_in_seq  pos_in_ref nuc category  A>C  A>G  A>T  C>A  C>G  \\\n",
       "0         1           0         295   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "1         1           1         296   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "2         1           2         297   A   Bulges  0.0  2.0  0.0  0.0  0.0   \n",
       "3         1           3         298   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "4         1           4         299   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "...     ...         ...         ...  ..      ...  ...  ...  ...  ...  ...   \n",
       "2418     67          23       16376   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "2419     67          24       16377   C   Tetrad  0.0  0.0  0.0  1.0  0.0   \n",
       "2420     67          25       16378   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "2421     67          26       16379   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "2422     67          27       16380   C   Tetrad  0.0  0.0  0.0  0.0  0.0   \n",
       "\n",
       "      C>T  G>A  G>C  G>T  T>A  T>C  T>G  number_of_Mut  \n",
       "0     1.0  0.0  0.0  0.0  0.0  0.0  0.0            1.0  \n",
       "1     3.0  0.0  0.0  0.0  0.0  0.0  0.0            3.0  \n",
       "2     0.0  0.0  0.0  0.0  0.0  0.0  0.0            2.0  \n",
       "3     0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "4     0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "...   ...  ...  ...  ...  ...  ...  ...            ...  \n",
       "2418  0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "2419  0.0  0.0  0.0  0.0  0.0  0.0  0.0            1.0  \n",
       "2420  0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "2421  0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "2422  0.0  0.0  0.0  0.0  0.0  0.0  0.0            0.0  \n",
       "\n",
       "[2423 rows x 18 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_category_cancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "8907bd64-9c39-4569-9f37-f41cd18556c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "complementator = str.maketrans(\"ACGT\", \"TGCA\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "16f25fc9-779a-4d2d-8266-377fd51a3077",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = df_category_cancers.groupby(['G4_id','category']).agg({'A>C':'sum','A>G':'sum','A>T':'sum',\n",
    "                                                'C>A':'sum','C>G':'sum','C>T':'sum',\n",
    "                                                'G>A':'sum','G>C':'sum','G>T':'sum',\n",
    "                                                'T>A':'sum','T>C':'sum','T>G':'sum'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "d9fe8848-eee5-4a34-979d-116ed24e2159",
   "metadata": {},
   "outputs": [],
   "source": [
    "a.columns = a.columns.str.translate(complementator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "af4cdb24-fe03-4d76-b3e3-4a5d0d238b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>A&gt;C</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G4_id</th>\n",
       "      <th>category</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">1</th>\n",
       "      <th>Bulges</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Loop</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mismatch</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tetrad</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <th>Bulges</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <th>Tetrad</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">66</th>\n",
       "      <th>Loop</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tetrad</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">67</th>\n",
       "      <th>Loop</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tetrad</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                T>G   T>C  T>A  G>T  G>C   G>A  C>T  C>G  C>A  A>T  A>G  A>C\n",
       "G4_id category                                                              \n",
       "1     Bulges    0.0   2.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "      Loop      0.0   0.0  0.0  5.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "      Mismatch  0.0   0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "      Tetrad    0.0   0.0  0.0  0.0  0.0   6.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "2     Bulges    0.0   2.0  0.0  0.0  0.0   0.0  1.0  0.0  1.0  0.0  0.0  0.0\n",
       "...             ...   ...  ...  ...  ...   ...  ...  ...  ...  ...  ...  ...\n",
       "65    Tetrad    0.0   0.0  0.0  4.0  1.0  23.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "66    Loop      0.0  12.0  0.0  0.0  0.0   5.0  1.0  0.0  0.0  0.0  1.0  0.0\n",
       "      Tetrad    0.0   0.0  0.0  1.0  2.0  32.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "67    Loop      0.0   0.0  0.0  0.0  0.0   2.0  1.0  0.0  0.0  1.0  9.0  0.0\n",
       "      Tetrad    0.0   0.0  0.0  3.0  0.0  15.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "\n",
       "[203 rows x 12 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33d57aa2-e3f0-4675-86f3-fd26ecb9839e",
   "metadata": {},
   "source": [
    "### +-15 interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "72120141-b4af-41de-953f-8bda505dbeb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loop_cancer = collect_mut_cancer(df_cancers=df_cancers, df_win=data_loops, regime='1col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "3a52bdde-5be5-4e04-aa54-55ee9d6391f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_15_col_cancer = collect_mut_cancer(df_cancers=df_cancers, df_win=data_15_col, regime='2col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "81238ed1-f3ed-4d88-8c07-db4f14a4211e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_loop_cancer.to_csv('../data/processed/df_loop_cancer_pqs.csv', index=False)\n",
    "# df_15_col_cancer.to_csv('../data/processed/df_15_col_cancer_pqs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cebb1dd-3ab0-4e3b-8120-646927e9dae6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "42bda5c4-a695-40b8-bb99-f1f82f724f61",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# OTHER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f5ecbb-abea-496d-8954-4cd968d2a740",
   "metadata": {},
   "source": [
    "## Location of g4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9b25f5fd-29f8-465b-8078-ba0572244a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4_location = []\n",
    "for G4_id, start, end in zip(g4conf['G4_id'], g4conf['start'], g4conf['end']):\n",
    "    location = []\n",
    "    row = []\n",
    "    for i in range(start, end+1):\n",
    "        # row.append(G4_id)\n",
    "        # row.append(i)\n",
    "        location.append(ref_annot[ref_annot['Pos'] == i]['Type'].to_string(index=False))\n",
    "        \n",
    "    row.append(G4_id)\n",
    "    row.append('/'.join(list(set(location))))\n",
    "    g4_location.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5ac157a1-9c4c-4630-9830-76b013eb3f7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1, 'D-loop'],\n",
       " [2, 'D-loop'],\n",
       " [3, 'D-loop/tRNA'],\n",
       " [4, 'rRNA'],\n",
       " [5, 'rRNA'],\n",
       " [6, 'tRNA/rRNA'],\n",
       " [7, 'rRNA'],\n",
       " [8, 'rRNA'],\n",
       " [9, 'rRNA'],\n",
       " [10, 'rRNA'],\n",
       " [11, 'CDS'],\n",
       " [12, 'CDS'],\n",
       " [13, 'CDS'],\n",
       " [14, 'CDS'],\n",
       " [15, 'CDS'],\n",
       " [16, 'CDS'],\n",
       " [17, 'tRNA'],\n",
       " [18, 'tRNA/CDS'],\n",
       " [19, 'CDS'],\n",
       " [20, 'CDS'],\n",
       " [21, 'CDS'],\n",
       " [22, 'CDS'],\n",
       " [23, 'CDS'],\n",
       " [24, 'tRNA/NaN/CDS'],\n",
       " [25, 'CDS'],\n",
       " [26, 'CDS'],\n",
       " [27, 'CDS'],\n",
       " [28, 'CDS'],\n",
       " [29, 'CDS'],\n",
       " [30, 'CDS'],\n",
       " [31, 'CDS'],\n",
       " [32, 'NaN/CDS/tRNA'],\n",
       " [33, 'CDS'],\n",
       " [34, 'CDS'],\n",
       " [35, 'CDS'],\n",
       " [36, 'CDS'],\n",
       " [37, 'CDS'],\n",
       " [38, 'CDS'],\n",
       " [39, 'CDS'],\n",
       " [40, 'CDS'],\n",
       " [41, 'CDS'],\n",
       " [42, 'CDS'],\n",
       " [43, 'CDS'],\n",
       " [44, 'CDS'],\n",
       " [45, 'CDS'],\n",
       " [46, 'CDS'],\n",
       " [47, 'CDS'],\n",
       " [48, 'CDS'],\n",
       " [49, 'CDS'],\n",
       " [50, 'CDS'],\n",
       " [51, 'CDS'],\n",
       " [52, 'CDS'],\n",
       " [53, 'CDS'],\n",
       " [54, 'CDS'],\n",
       " [55, 'CDS'],\n",
       " [56, 'CDS'],\n",
       " [57, 'CDS'],\n",
       " [58, 'CDS'],\n",
       " [59, 'CDS'],\n",
       " [60, 'CDS'],\n",
       " [61, 'CDS'],\n",
       " [62, 'CDS'],\n",
       " [63, 'CDS'],\n",
       " [64, 'CDS'],\n",
       " [65, 'CDS'],\n",
       " [66, 'CDS'],\n",
       " [67, 'CDS'],\n",
       " [68, 'CDS'],\n",
       " [69, 'CDS'],\n",
       " [70, 'CDS'],\n",
       " [71, 'D-loop'],\n",
       " [72, 'D-loop'],\n",
       " [73, 'D-loop'],\n",
       " [74, 'D-loop']]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6996452e-ce34-4554-8198-91d265e23823",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_location = pd.DataFrame(data=g4_location, columns=['G4_id', 'location'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d8b008b9-8161-490c-8aa0-18949456d5fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>D-loop/tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>70</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>72</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>73</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>74</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id     location\n",
       "0       1       D-loop\n",
       "1       2       D-loop\n",
       "2       3  D-loop/tRNA\n",
       "3       4         rRNA\n",
       "4       5         rRNA\n",
       "..    ...          ...\n",
       "69     70          CDS\n",
       "70     71       D-loop\n",
       "71     72       D-loop\n",
       "72     73       D-loop\n",
       "73     74       D-loop\n",
       "\n",
       "[74 rows x 2 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3f80e031-6250-4ac5-bbbb-e99a7ad713a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_location = df_g4_location.set_index('G4_id')['location'].str.split('/', expand=True) \\\n",
    "    .stack() \\\n",
    "    .reset_index(name='location') \\\n",
    "    .drop('level_1', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "619c81a1-64cd-4aeb-8736-77b00a279e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>70</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>71</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>72</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>73</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>74</td>\n",
       "      <td>D-loop</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    G4_id location\n",
       "0       1   D-loop\n",
       "1       2   D-loop\n",
       "2       3   D-loop\n",
       "3       3     tRNA\n",
       "4       4     rRNA\n",
       "..    ...      ...\n",
       "76     70      CDS\n",
       "77     71   D-loop\n",
       "78     72   D-loop\n",
       "79     73   D-loop\n",
       "80     74   D-loop\n",
       "\n",
       "[81 rows x 2 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "238b70b6-bbb5-4d35-b151-fb6dd91d1c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_location.to_csv('../data/processed/df_g4_location.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e217401f-778f-471c-8944-d60363780fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqnames', 'start', 'end', 'width', 'strand', 'score', 'max_score',\n",
       "       'threshold', 'window', 'sequence', 'confirmed', 'G4', 'note',\n",
       "       'start-100bp', 'end+100bp', 'start-50bp', 'end+50bp',\n",
       "       'bp_untill_next_G4', 'bp_untill_previous_G4', 'G4_id', 'up_win',\n",
       "       'down_win'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "207182b6-3926-440c-beaa-e5d8debac76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_type = g4conf[['G4_id', 'G4']].rename(columns={'G4':'Type_of_G4'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1a3802d-9cb0-45ef-ae1b-667476e4815e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_type.to_csv('../data/processed/df_g4_type.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea9edee-8842-4dd8-afd3-ad23e066bb24",
   "metadata": {},
   "source": [
    "## G4 location *in vivo* (doi.org/10.1093/nar/gkad535)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "10eb5e60-ca71-4f9c-99cd-3165eb22bc2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "G4_peaks = [[1, 2425, 3986], [2, 4330, 4624], [3, 5570, 7168], [4, 7139, 7559], [5, 7800, 8276], [6, 8958, 9869], [7, 12762, 13771]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fa699b23-3911-42c9-a74a-ae2ac50bf4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_vivo_peaks = pd.DataFrame(G4_peaks, columns=['number', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8c79b7f6-7683-4f16-8f5c-68c50cc9a383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2425</td>\n",
       "      <td>3986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4330</td>\n",
       "      <td>4624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>5570</td>\n",
       "      <td>7168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>7139</td>\n",
       "      <td>7559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>7800</td>\n",
       "      <td>8276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>8958</td>\n",
       "      <td>9869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>12762</td>\n",
       "      <td>13771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   number  start    end\n",
       "0       1   2425   3986\n",
       "1       2   4330   4624\n",
       "2       3   5570   7168\n",
       "3       4   7139   7559\n",
       "4       5   7800   8276\n",
       "5       6   8958   9869\n",
       "6       7  12762  13771"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_vivo_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9864ebf4-92b7-4dff-b323-d16e64db9c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# g4conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6ef9e448-5e23-47e0-a09f-7d651fe49230",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4_vivo = []\n",
    "for start, end, G4_id in zip(g4conf['start'], g4conf['end'], g4conf['G4_id']):\n",
    "    for p_start, p_end, number in zip(df_g4_vivo_peaks['start'], df_g4_vivo_peaks['end'], df_g4_vivo_peaks['number']):\n",
    "        row = []\n",
    "        if start in range(p_start, p_end) or end in range(p_start, p_end):\n",
    "            row.append(number)\n",
    "            row.append(G4_id)\n",
    "            row.append(1)\n",
    "            g4_vivo.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c339a343-cc39-421a-9ff3-af75aea04821",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g4_vivo = pd.DataFrame(g4_vivo, columns=['peak_number', 'G4_id', 'in_vivo'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "21443f3b-5492-4c6f-a9a9-883b243ffed8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_number</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>in_vivo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>6</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>6</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>6</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>7</td>\n",
       "      <td>52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>7</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>7</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>7</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>7</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    peak_number  G4_id  in_vivo\n",
       "0             1      9        1\n",
       "1             1     10        1\n",
       "2             1     11        1\n",
       "3             1     12        1\n",
       "4             1     13        1\n",
       "5             2     17        1\n",
       "6             2     18        1\n",
       "7             3     24        1\n",
       "8             3     25        1\n",
       "9             3     26        1\n",
       "10            3     27        1\n",
       "11            3     28        1\n",
       "12            4     29        1\n",
       "13            5     30        1\n",
       "14            5     31        1\n",
       "15            5     32        1\n",
       "16            6     37        1\n",
       "17            6     38        1\n",
       "18            6     39        1\n",
       "19            6     40        1\n",
       "20            7     52        1\n",
       "21            7     53        1\n",
       "22            7     54        1\n",
       "23            7     55        1\n",
       "24            7     56        1\n",
       "25            7     57        1"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g4_vivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "95362125-0985-4a77-9152-8dbee0d20b3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_g4_vivo.to_csv('../data/processed/df_g4_vivo.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "193291d6-4be2-4d02-ae83-78c1cc1bb043",
   "metadata": {},
   "source": [
    "## Check decrease of С>T  \n",
    "Make new windows which cover all length G4 and +-25 bp \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f454cf2d-1bc9-4f35-8c67-3f29be88c1a2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "      <th>start-50bp</th>\n",
       "      <th>end+50bp</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>285</td>\n",
       "      <td>332</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.250</td>\n",
       "      <td>-2.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>185</td>\n",
       "      <td>432</td>\n",
       "      <td>235</td>\n",
       "      <td>382</td>\n",
       "      <td>16</td>\n",
       "      <td>294</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>348</td>\n",
       "      <td>378</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCAAACCCCAAAAACAAAGAACCCTAACACC</td>\n",
       "      <td>...</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>248</td>\n",
       "      <td>478</td>\n",
       "      <td>298</td>\n",
       "      <td>428</td>\n",
       "      <td>40</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>332</td>\n",
       "      <td>418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "      <td>368</td>\n",
       "      <td>642</td>\n",
       "      <td>193</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>378</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>785</td>\n",
       "      <td>808</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCTTAGCCTAGCCACACCCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "      <td>735</td>\n",
       "      <td>858</td>\n",
       "      <td>146</td>\n",
       "      <td>193</td>\n",
       "      <td>4</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "      <td>904</td>\n",
       "      <td>1034</td>\n",
       "      <td>683</td>\n",
       "      <td>146</td>\n",
       "      <td>5</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>15536</td>\n",
       "      <td>15783</td>\n",
       "      <td>15586</td>\n",
       "      <td>15733</td>\n",
       "      <td>484</td>\n",
       "      <td>80</td>\n",
       "      <td>70</td>\n",
       "      <td>15556</td>\n",
       "      <td>15783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16067</td>\n",
       "      <td>16311</td>\n",
       "      <td>16117</td>\n",
       "      <td>16261</td>\n",
       "      <td>28</td>\n",
       "      <td>484</td>\n",
       "      <td>71</td>\n",
       "      <td>16067</td>\n",
       "      <td>16239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>63</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16139</td>\n",
       "      <td>16401</td>\n",
       "      <td>16189</td>\n",
       "      <td>16351</td>\n",
       "      <td>43</td>\n",
       "      <td>28</td>\n",
       "      <td>72</td>\n",
       "      <td>16211</td>\n",
       "      <td>16344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16244</td>\n",
       "      <td>16482</td>\n",
       "      <td>16294</td>\n",
       "      <td>16432</td>\n",
       "      <td>145</td>\n",
       "      <td>43</td>\n",
       "      <td>73</td>\n",
       "      <td>16301</td>\n",
       "      <td>16482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16527</td>\n",
       "      <td>16560</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.853</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16427</td>\n",
       "      <td>91</td>\n",
       "      <td>16477</td>\n",
       "      <td>16610</td>\n",
       "      <td>294</td>\n",
       "      <td>145</td>\n",
       "      <td>74</td>\n",
       "      <td>16427</td>\n",
       "      <td>16660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start    end  width  \\\n",
       "0   NC_012920.1 Homo sapiens mitochondrion, comple...    285    332     48   \n",
       "1   NC_012920.1 Homo sapiens mitochondrion, comple...    348    378     31   \n",
       "2   NC_012920.1 Homo sapiens mitochondrion, comple...    418    592    175   \n",
       "3   NC_012920.1 Homo sapiens mitochondrion, comple...    785    808     24   \n",
       "4   NC_012920.1 Homo sapiens mitochondrion, comple...    954    984     31   \n",
       "..                                                ...    ...    ...    ...   \n",
       "69  NC_012920.1 Homo sapiens mitochondrion, comple...  15636  15683     48   \n",
       "70  NC_012920.1 Homo sapiens mitochondrion, comple...  16167  16211     45   \n",
       "71  NC_012920.1 Homo sapiens mitochondrion, comple...  16239  16301     63   \n",
       "72  NC_012920.1 Homo sapiens mitochondrion, comple...  16344  16382     39   \n",
       "73  NC_012920.1 Homo sapiens mitochondrion, comple...  16527  16560     34   \n",
       "\n",
       "   strand  score  max_score  threshold  window  \\\n",
       "0       - -1.250      -2.24          1      25   \n",
       "1       - -1.100      -1.16          1      25   \n",
       "2       - -1.150      -1.88          1      25   \n",
       "3       - -1.210      -1.16          1      25   \n",
       "4       - -1.390      -1.48          1      25   \n",
       "..    ...    ...        ...        ...     ...   \n",
       "69      - -0.896      -1.16          1      25   \n",
       "70      - -1.160      -1.80          1      25   \n",
       "71      - -0.968      -1.40          1      25   \n",
       "72      - -1.210      -1.48          1      25   \n",
       "73      - -0.853      -1.16          1      25   \n",
       "\n",
       "                                             sequence  ...  \\\n",
       "0    CAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCAC  ...   \n",
       "1                     CCAAACCCCAAAAACAAAGAACCCTAACACC  ...   \n",
       "2   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...  ...   \n",
       "3                            CGCTTAGCCTAGCCACACCCCCAC  ...   \n",
       "4                     CACCCCCTCCCCAATAAAGCTAAAACTCACC  ...   \n",
       "..                                                ...  ...   \n",
       "69   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC  ...   \n",
       "70      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC  ...   \n",
       "71  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...  ...   \n",
       "72            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC  ...   \n",
       "73                 CCTAAATAGCCCACACGTTCCCCTTAAATAAGAC  ...   \n",
       "\n",
       "                          note start-100bp end+100bp  start-50bp  end+50bp  \\\n",
       "0             more than africa         185       432         235       382   \n",
       "1             less than africa         248       478         298       428   \n",
       "2   5 overlapping g4 in africa         318       692         368       642   \n",
       "3                          NaN         685       908         735       858   \n",
       "4                          NaN         854      1084         904      1034   \n",
       "..                         ...         ...       ...         ...       ...   \n",
       "69            more than africa       15536     15783       15586     15733   \n",
       "70            more than africa       16067     16311       16117     16261   \n",
       "71            more than africa       16139     16401       16189     16351   \n",
       "72            more than africa       16244     16482       16294     16432   \n",
       "73            more than africa       16427        91       16477     16610   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  G4_id  up_win  down_win  \n",
       "0                  16                    294      1     185       348  \n",
       "1                  40                     16      2     332       418  \n",
       "2                 193                     40      3     378       692  \n",
       "3                 146                    193      4     685       908  \n",
       "4                 683                    146      5     854      1084  \n",
       "..                ...                    ...    ...     ...       ...  \n",
       "69                484                     80     70   15556     15783  \n",
       "70                 28                    484     71   16067     16239  \n",
       "71                 43                     28     72   16211     16344  \n",
       "72                145                     43     73   16301     16482  \n",
       "73                294                    145     74   16427     16660  \n",
       "\n",
       "[74 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g4conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "682becd2-66c4-48b5-b8a7-2585d1818f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_25bp = g4conf[(g4conf['bp_untill_previous_G4'] >= 25) & \n",
    "            (g4conf['bp_untill_next_G4'] >= 25) & \n",
    "            (g4conf['end'] + 25 <= 16569)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf69aff9-8a67-4b42-8b60-9e7439654dfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>max_score</th>\n",
       "      <th>threshold</th>\n",
       "      <th>window</th>\n",
       "      <th>sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>note</th>\n",
       "      <th>start-100bp</th>\n",
       "      <th>end+100bp</th>\n",
       "      <th>start-50bp</th>\n",
       "      <th>end+50bp</th>\n",
       "      <th>bp_untill_next_G4</th>\n",
       "      <th>bp_untill_previous_G4</th>\n",
       "      <th>G4_id</th>\n",
       "      <th>up_win</th>\n",
       "      <th>down_win</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>418</td>\n",
       "      <td>592</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.150</td>\n",
       "      <td>-1.88</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>318</td>\n",
       "      <td>692</td>\n",
       "      <td>368</td>\n",
       "      <td>642</td>\n",
       "      <td>193</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>378</td>\n",
       "      <td>692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>785</td>\n",
       "      <td>808</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCTTAGCCTAGCCACACCCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "      <td>735</td>\n",
       "      <td>858</td>\n",
       "      <td>146</td>\n",
       "      <td>193</td>\n",
       "      <td>4</td>\n",
       "      <td>685</td>\n",
       "      <td>908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>954</td>\n",
       "      <td>984</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.390</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACCCCCTCCCCAATAAAGCTAAAACTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "      <td>904</td>\n",
       "      <td>1034</td>\n",
       "      <td>683</td>\n",
       "      <td>146</td>\n",
       "      <td>5</td>\n",
       "      <td>854</td>\n",
       "      <td>1084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>1667</td>\n",
       "      <td>1707</td>\n",
       "      <td>41</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTGAGCTAAACCTAGCCCCAAACCCACTCCACCTTACTACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than in africa</td>\n",
       "      <td>1567</td>\n",
       "      <td>1807</td>\n",
       "      <td>1617</td>\n",
       "      <td>1757</td>\n",
       "      <td>350</td>\n",
       "      <td>683</td>\n",
       "      <td>6</td>\n",
       "      <td>1567</td>\n",
       "      <td>1807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>2057</td>\n",
       "      <td>2079</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.520</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCACAGAACCCTCTAAATCCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1957</td>\n",
       "      <td>2179</td>\n",
       "      <td>2007</td>\n",
       "      <td>2129</td>\n",
       "      <td>133</td>\n",
       "      <td>350</td>\n",
       "      <td>7</td>\n",
       "      <td>1957</td>\n",
       "      <td>2179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>2212</td>\n",
       "      <td>2236</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACACCCACTACCTAAAAAATCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2112</td>\n",
       "      <td>2336</td>\n",
       "      <td>2162</td>\n",
       "      <td>2286</td>\n",
       "      <td>912</td>\n",
       "      <td>133</td>\n",
       "      <td>8</td>\n",
       "      <td>2112</td>\n",
       "      <td>2336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3471</td>\n",
       "      <td>3604</td>\n",
       "      <td>134</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.170</td>\n",
       "      <td>-2.76</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>5 overlapping g4 in africa</td>\n",
       "      <td>3371</td>\n",
       "      <td>3704</td>\n",
       "      <td>3421</td>\n",
       "      <td>3654</td>\n",
       "      <td>270</td>\n",
       "      <td>249</td>\n",
       "      <td>11</td>\n",
       "      <td>3371</td>\n",
       "      <td>3704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3874</td>\n",
       "      <td>3909</td>\n",
       "      <td>36</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.917</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3774</td>\n",
       "      <td>4009</td>\n",
       "      <td>3824</td>\n",
       "      <td>3959</td>\n",
       "      <td>42</td>\n",
       "      <td>270</td>\n",
       "      <td>12</td>\n",
       "      <td>3774</td>\n",
       "      <td>3951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>3951</td>\n",
       "      <td>3984</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.882</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>3851</td>\n",
       "      <td>4084</td>\n",
       "      <td>3901</td>\n",
       "      <td>4034</td>\n",
       "      <td>102</td>\n",
       "      <td>42</td>\n",
       "      <td>13</td>\n",
       "      <td>3909</td>\n",
       "      <td>4084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4228</td>\n",
       "      <td>4260</td>\n",
       "      <td>33</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCATTACAATCTCCAGCATTCCCCCTCAAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4128</td>\n",
       "      <td>4360</td>\n",
       "      <td>4178</td>\n",
       "      <td>4310</td>\n",
       "      <td>112</td>\n",
       "      <td>67</td>\n",
       "      <td>16</td>\n",
       "      <td>4161</td>\n",
       "      <td>4360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4372</td>\n",
       "      <td>4398</td>\n",
       "      <td>27</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.220</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCCGTGCCACCTATCACACCCCATCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4272</td>\n",
       "      <td>4498</td>\n",
       "      <td>4322</td>\n",
       "      <td>4448</td>\n",
       "      <td>58</td>\n",
       "      <td>112</td>\n",
       "      <td>17</td>\n",
       "      <td>4272</td>\n",
       "      <td>4456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4456</td>\n",
       "      <td>4505</td>\n",
       "      <td>50</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.100</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTTCCCGTACTAATTAATCCCCTGGCCCAACCCGTCATCTACTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in africa</td>\n",
       "      <td>4356</td>\n",
       "      <td>4605</td>\n",
       "      <td>4406</td>\n",
       "      <td>4555</td>\n",
       "      <td>370</td>\n",
       "      <td>58</td>\n",
       "      <td>18</td>\n",
       "      <td>4398</td>\n",
       "      <td>4605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>4875</td>\n",
       "      <td>4899</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.080</td>\n",
       "      <td>-1.08</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGCCCCCATCTCAATCATATACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4775</td>\n",
       "      <td>4999</td>\n",
       "      <td>4825</td>\n",
       "      <td>4949</td>\n",
       "      <td>288</td>\n",
       "      <td>370</td>\n",
       "      <td>19</td>\n",
       "      <td>4775</td>\n",
       "      <td>4999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5187</td>\n",
       "      <td>5255</td>\n",
       "      <td>69</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.957</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGC...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>5087</td>\n",
       "      <td>5355</td>\n",
       "      <td>5137</td>\n",
       "      <td>5305</td>\n",
       "      <td>33</td>\n",
       "      <td>288</td>\n",
       "      <td>20</td>\n",
       "      <td>5087</td>\n",
       "      <td>5288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5424</td>\n",
       "      <td>5499</td>\n",
       "      <td>76</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.200</td>\n",
       "      <td>-1.92</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>5324</td>\n",
       "      <td>5599</td>\n",
       "      <td>5374</td>\n",
       "      <td>5549</td>\n",
       "      <td>376</td>\n",
       "      <td>32</td>\n",
       "      <td>23</td>\n",
       "      <td>5392</td>\n",
       "      <td>5599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>5875</td>\n",
       "      <td>5916</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.881</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACTCAGCCATTTTACCTCACCCCCACTGATGTTCGCCGACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>5775</td>\n",
       "      <td>6016</td>\n",
       "      <td>5825</td>\n",
       "      <td>5966</td>\n",
       "      <td>297</td>\n",
       "      <td>376</td>\n",
       "      <td>24</td>\n",
       "      <td>5775</td>\n",
       "      <td>6016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6213</td>\n",
       "      <td>6242</td>\n",
       "      <td>30</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.900</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCTTACCTCCCTCTCTCCTACTCCTGCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>6113</td>\n",
       "      <td>6342</td>\n",
       "      <td>6163</td>\n",
       "      <td>6292</td>\n",
       "      <td>168</td>\n",
       "      <td>297</td>\n",
       "      <td>25</td>\n",
       "      <td>6113</td>\n",
       "      <td>6342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6410</td>\n",
       "      <td>6464</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.050</td>\n",
       "      <td>-1.44</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>6310</td>\n",
       "      <td>6564</td>\n",
       "      <td>6360</td>\n",
       "      <td>6514</td>\n",
       "      <td>81</td>\n",
       "      <td>168</td>\n",
       "      <td>26</td>\n",
       "      <td>6310</td>\n",
       "      <td>6545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6545</td>\n",
       "      <td>6572</td>\n",
       "      <td>28</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTCAACACCACCTTCTTCGACCCCGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>less than africa</td>\n",
       "      <td>6445</td>\n",
       "      <td>6672</td>\n",
       "      <td>6495</td>\n",
       "      <td>6622</td>\n",
       "      <td>254</td>\n",
       "      <td>81</td>\n",
       "      <td>27</td>\n",
       "      <td>6464</td>\n",
       "      <td>6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>6826</td>\n",
       "      <td>6851</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.040</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTACCATAATCATCGCTATCCCCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6726</td>\n",
       "      <td>6951</td>\n",
       "      <td>6776</td>\n",
       "      <td>6901</td>\n",
       "      <td>535</td>\n",
       "      <td>254</td>\n",
       "      <td>28</td>\n",
       "      <td>6726</td>\n",
       "      <td>6951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7386</td>\n",
       "      <td>7427</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>7286</td>\n",
       "      <td>7527</td>\n",
       "      <td>7336</td>\n",
       "      <td>7477</td>\n",
       "      <td>371</td>\n",
       "      <td>535</td>\n",
       "      <td>29</td>\n",
       "      <td>7286</td>\n",
       "      <td>7527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7798</td>\n",
       "      <td>7840</td>\n",
       "      <td>43</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.977</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7698</td>\n",
       "      <td>7940</td>\n",
       "      <td>7748</td>\n",
       "      <td>7890</td>\n",
       "      <td>96</td>\n",
       "      <td>371</td>\n",
       "      <td>30</td>\n",
       "      <td>7698</td>\n",
       "      <td>7936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>7936</td>\n",
       "      <td>7974</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.923</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7836</td>\n",
       "      <td>8074</td>\n",
       "      <td>7886</td>\n",
       "      <td>8024</td>\n",
       "      <td>278</td>\n",
       "      <td>96</td>\n",
       "      <td>31</td>\n",
       "      <td>7840</td>\n",
       "      <td>8074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8252</td>\n",
       "      <td>8305</td>\n",
       "      <td>54</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.220</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCGTATTTACCCTATAGCACCCCCTCTACCCCCTCTAGAGCCCAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>8152</td>\n",
       "      <td>8405</td>\n",
       "      <td>8202</td>\n",
       "      <td>8355</td>\n",
       "      <td>150</td>\n",
       "      <td>278</td>\n",
       "      <td>32</td>\n",
       "      <td>8152</td>\n",
       "      <td>8405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8455</td>\n",
       "      <td>8488</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.090</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACAAACTACCACCTACCTCCCTCACCAAAGCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8355</td>\n",
       "      <td>8588</td>\n",
       "      <td>8405</td>\n",
       "      <td>8538</td>\n",
       "      <td>56</td>\n",
       "      <td>150</td>\n",
       "      <td>33</td>\n",
       "      <td>8355</td>\n",
       "      <td>8544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8791</td>\n",
       "      <td>8815</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTCACTCATTTACACCAACCACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8691</td>\n",
       "      <td>8915</td>\n",
       "      <td>8741</td>\n",
       "      <td>8865</td>\n",
       "      <td>102</td>\n",
       "      <td>148</td>\n",
       "      <td>36</td>\n",
       "      <td>8691</td>\n",
       "      <td>8915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>8917</td>\n",
       "      <td>8958</td>\n",
       "      <td>42</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.881</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>8817</td>\n",
       "      <td>9058</td>\n",
       "      <td>8867</td>\n",
       "      <td>9008</td>\n",
       "      <td>285</td>\n",
       "      <td>102</td>\n",
       "      <td>37</td>\n",
       "      <td>8817</td>\n",
       "      <td>9058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>9505</td>\n",
       "      <td>9607</td>\n",
       "      <td>103</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.020</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>9405</td>\n",
       "      <td>9707</td>\n",
       "      <td>9455</td>\n",
       "      <td>9657</td>\n",
       "      <td>568</td>\n",
       "      <td>212</td>\n",
       "      <td>40</td>\n",
       "      <td>9405</td>\n",
       "      <td>9707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10175</td>\n",
       "      <td>10214</td>\n",
       "      <td>40</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.200</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>10075</td>\n",
       "      <td>10314</td>\n",
       "      <td>10125</td>\n",
       "      <td>10264</td>\n",
       "      <td>43</td>\n",
       "      <td>568</td>\n",
       "      <td>41</td>\n",
       "      <td>10075</td>\n",
       "      <td>10257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10257</td>\n",
       "      <td>10303</td>\n",
       "      <td>47</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.894</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>10157</td>\n",
       "      <td>10403</td>\n",
       "      <td>10207</td>\n",
       "      <td>10353</td>\n",
       "      <td>300</td>\n",
       "      <td>43</td>\n",
       "      <td>42</td>\n",
       "      <td>10214</td>\n",
       "      <td>10403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10603</td>\n",
       "      <td>10637</td>\n",
       "      <td>35</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCTCATAACCCTCAACACCCACTCCCTCTTAGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10503</td>\n",
       "      <td>10737</td>\n",
       "      <td>10553</td>\n",
       "      <td>10687</td>\n",
       "      <td>275</td>\n",
       "      <td>300</td>\n",
       "      <td>43</td>\n",
       "      <td>10503</td>\n",
       "      <td>10737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>10912</td>\n",
       "      <td>10990</td>\n",
       "      <td>79</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.330</td>\n",
       "      <td>-2.08</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 g4 in africa</td>\n",
       "      <td>10812</td>\n",
       "      <td>11090</td>\n",
       "      <td>10862</td>\n",
       "      <td>11040</td>\n",
       "      <td>132</td>\n",
       "      <td>275</td>\n",
       "      <td>44</td>\n",
       "      <td>10812</td>\n",
       "      <td>11090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11122</td>\n",
       "      <td>11162</td>\n",
       "      <td>41</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.854</td>\n",
       "      <td>-1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11022</td>\n",
       "      <td>11262</td>\n",
       "      <td>11072</td>\n",
       "      <td>11212</td>\n",
       "      <td>50</td>\n",
       "      <td>132</td>\n",
       "      <td>45</td>\n",
       "      <td>11022</td>\n",
       "      <td>11212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11212</td>\n",
       "      <td>11249</td>\n",
       "      <td>38</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.947</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11112</td>\n",
       "      <td>11349</td>\n",
       "      <td>11162</td>\n",
       "      <td>11299</td>\n",
       "      <td>243</td>\n",
       "      <td>50</td>\n",
       "      <td>46</td>\n",
       "      <td>11162</td>\n",
       "      <td>11349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11492</td>\n",
       "      <td>11555</td>\n",
       "      <td>64</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACC...</td>\n",
       "      <td>...</td>\n",
       "      <td>4 overlapping g4 in africa</td>\n",
       "      <td>11392</td>\n",
       "      <td>11655</td>\n",
       "      <td>11442</td>\n",
       "      <td>11605</td>\n",
       "      <td>100</td>\n",
       "      <td>243</td>\n",
       "      <td>47</td>\n",
       "      <td>11392</td>\n",
       "      <td>11655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>11655</td>\n",
       "      <td>11692</td>\n",
       "      <td>38</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.816</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11555</td>\n",
       "      <td>11792</td>\n",
       "      <td>11605</td>\n",
       "      <td>11742</td>\n",
       "      <td>371</td>\n",
       "      <td>100</td>\n",
       "      <td>48</td>\n",
       "      <td>11555</td>\n",
       "      <td>11792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12063</td>\n",
       "      <td>12124</td>\n",
       "      <td>62</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.60</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>11963</td>\n",
       "      <td>12224</td>\n",
       "      <td>12013</td>\n",
       "      <td>12174</td>\n",
       "      <td>222</td>\n",
       "      <td>371</td>\n",
       "      <td>49</td>\n",
       "      <td>11963</td>\n",
       "      <td>12224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12346</td>\n",
       "      <td>12450</td>\n",
       "      <td>105</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.76</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>12246</td>\n",
       "      <td>12550</td>\n",
       "      <td>12296</td>\n",
       "      <td>12500</td>\n",
       "      <td>92</td>\n",
       "      <td>222</td>\n",
       "      <td>50</td>\n",
       "      <td>12246</td>\n",
       "      <td>12542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>12542</td>\n",
       "      <td>12573</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.060</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCACAACCCAAACAACCCAGCTCTCCCTAAGC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>12442</td>\n",
       "      <td>12673</td>\n",
       "      <td>12492</td>\n",
       "      <td>12623</td>\n",
       "      <td>378</td>\n",
       "      <td>92</td>\n",
       "      <td>51</td>\n",
       "      <td>12450</td>\n",
       "      <td>12673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13107</td>\n",
       "      <td>13150</td>\n",
       "      <td>44</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.909</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCAC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>13007</td>\n",
       "      <td>13250</td>\n",
       "      <td>13057</td>\n",
       "      <td>13200</td>\n",
       "      <td>280</td>\n",
       "      <td>25</td>\n",
       "      <td>54</td>\n",
       "      <td>13082</td>\n",
       "      <td>13250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13430</td>\n",
       "      <td>13458</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.030</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATACCTCTCACTTCAACCTCCCTCACC</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>13330</td>\n",
       "      <td>13558</td>\n",
       "      <td>13380</td>\n",
       "      <td>13508</td>\n",
       "      <td>167</td>\n",
       "      <td>280</td>\n",
       "      <td>55</td>\n",
       "      <td>13330</td>\n",
       "      <td>13558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13625</td>\n",
       "      <td>13705</td>\n",
       "      <td>81</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.050</td>\n",
       "      <td>-1.68</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13525</td>\n",
       "      <td>13805</td>\n",
       "      <td>13575</td>\n",
       "      <td>13755</td>\n",
       "      <td>30</td>\n",
       "      <td>167</td>\n",
       "      <td>56</td>\n",
       "      <td>13525</td>\n",
       "      <td>13735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13735</td>\n",
       "      <td>13809</td>\n",
       "      <td>75</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.170</td>\n",
       "      <td>-1.84</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13635</td>\n",
       "      <td>13909</td>\n",
       "      <td>13685</td>\n",
       "      <td>13859</td>\n",
       "      <td>150</td>\n",
       "      <td>30</td>\n",
       "      <td>57</td>\n",
       "      <td>13705</td>\n",
       "      <td>13909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>13959</td>\n",
       "      <td>14004</td>\n",
       "      <td>46</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.935</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>13859</td>\n",
       "      <td>14104</td>\n",
       "      <td>13909</td>\n",
       "      <td>14054</td>\n",
       "      <td>45</td>\n",
       "      <td>150</td>\n",
       "      <td>58</td>\n",
       "      <td>13859</td>\n",
       "      <td>14049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14049</td>\n",
       "      <td>14074</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-1.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTCCACCTCCATCATCACCTCAACCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13949</td>\n",
       "      <td>14174</td>\n",
       "      <td>13999</td>\n",
       "      <td>14124</td>\n",
       "      <td>35</td>\n",
       "      <td>45</td>\n",
       "      <td>59</td>\n",
       "      <td>14004</td>\n",
       "      <td>14109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14109</td>\n",
       "      <td>14169</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.951</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAACCTATT...</td>\n",
       "      <td>...</td>\n",
       "      <td>2 overlapping g4 in africa</td>\n",
       "      <td>14009</td>\n",
       "      <td>14269</td>\n",
       "      <td>14059</td>\n",
       "      <td>14219</td>\n",
       "      <td>56</td>\n",
       "      <td>35</td>\n",
       "      <td>60</td>\n",
       "      <td>14074</td>\n",
       "      <td>14225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14225</td>\n",
       "      <td>14292</td>\n",
       "      <td>68</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-1.52</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CGCCCATAATCATACAAAGCCCCCGCACCAATAGGATCCTCCCGAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in africa</td>\n",
       "      <td>14125</td>\n",
       "      <td>14392</td>\n",
       "      <td>14175</td>\n",
       "      <td>14342</td>\n",
       "      <td>36</td>\n",
       "      <td>56</td>\n",
       "      <td>61</td>\n",
       "      <td>14169</td>\n",
       "      <td>14328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>14751</td>\n",
       "      <td>14832</td>\n",
       "      <td>82</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.110</td>\n",
       "      <td>-1.84</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>3 overlapping g4 in afica</td>\n",
       "      <td>14651</td>\n",
       "      <td>14932</td>\n",
       "      <td>14701</td>\n",
       "      <td>14882</td>\n",
       "      <td>528</td>\n",
       "      <td>98</td>\n",
       "      <td>67</td>\n",
       "      <td>14653</td>\n",
       "      <td>14932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15360</td>\n",
       "      <td>15391</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.120</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAAACAACCCCCTAGGAATCACCTCCCATTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15260</td>\n",
       "      <td>15491</td>\n",
       "      <td>15310</td>\n",
       "      <td>15441</td>\n",
       "      <td>111</td>\n",
       "      <td>528</td>\n",
       "      <td>68</td>\n",
       "      <td>15260</td>\n",
       "      <td>15491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15502</td>\n",
       "      <td>15556</td>\n",
       "      <td>55</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.450</td>\n",
       "      <td>-2.12</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15402</td>\n",
       "      <td>15656</td>\n",
       "      <td>15452</td>\n",
       "      <td>15606</td>\n",
       "      <td>80</td>\n",
       "      <td>111</td>\n",
       "      <td>69</td>\n",
       "      <td>15402</td>\n",
       "      <td>15636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>15636</td>\n",
       "      <td>15683</td>\n",
       "      <td>48</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.896</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>15536</td>\n",
       "      <td>15783</td>\n",
       "      <td>15586</td>\n",
       "      <td>15733</td>\n",
       "      <td>484</td>\n",
       "      <td>80</td>\n",
       "      <td>70</td>\n",
       "      <td>15556</td>\n",
       "      <td>15783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16167</td>\n",
       "      <td>16211</td>\n",
       "      <td>45</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.160</td>\n",
       "      <td>-1.80</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16067</td>\n",
       "      <td>16311</td>\n",
       "      <td>16117</td>\n",
       "      <td>16261</td>\n",
       "      <td>28</td>\n",
       "      <td>484</td>\n",
       "      <td>71</td>\n",
       "      <td>16067</td>\n",
       "      <td>16239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16239</td>\n",
       "      <td>16301</td>\n",
       "      <td>63</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.968</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16139</td>\n",
       "      <td>16401</td>\n",
       "      <td>16189</td>\n",
       "      <td>16351</td>\n",
       "      <td>43</td>\n",
       "      <td>28</td>\n",
       "      <td>72</td>\n",
       "      <td>16211</td>\n",
       "      <td>16344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_012920.1 Homo sapiens mitochondrion, comple...</td>\n",
       "      <td>16344</td>\n",
       "      <td>16382</td>\n",
       "      <td>39</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.210</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC</td>\n",
       "      <td>...</td>\n",
       "      <td>more than africa</td>\n",
       "      <td>16244</td>\n",
       "      <td>16482</td>\n",
       "      <td>16294</td>\n",
       "      <td>16432</td>\n",
       "      <td>145</td>\n",
       "      <td>43</td>\n",
       "      <td>73</td>\n",
       "      <td>16301</td>\n",
       "      <td>16482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqnames  start    end  width  \\\n",
       "2   NC_012920.1 Homo sapiens mitochondrion, comple...    418    592    175   \n",
       "3   NC_012920.1 Homo sapiens mitochondrion, comple...    785    808     24   \n",
       "4   NC_012920.1 Homo sapiens mitochondrion, comple...    954    984     31   \n",
       "5   NC_012920.1 Homo sapiens mitochondrion, comple...   1667   1707     41   \n",
       "6   NC_012920.1 Homo sapiens mitochondrion, comple...   2057   2079     23   \n",
       "7   NC_012920.1 Homo sapiens mitochondrion, comple...   2212   2236     25   \n",
       "10  NC_012920.1 Homo sapiens mitochondrion, comple...   3471   3604    134   \n",
       "11  NC_012920.1 Homo sapiens mitochondrion, comple...   3874   3909     36   \n",
       "12  NC_012920.1 Homo sapiens mitochondrion, comple...   3951   3984     34   \n",
       "15  NC_012920.1 Homo sapiens mitochondrion, comple...   4228   4260     33   \n",
       "16  NC_012920.1 Homo sapiens mitochondrion, comple...   4372   4398     27   \n",
       "17  NC_012920.1 Homo sapiens mitochondrion, comple...   4456   4505     50   \n",
       "18  NC_012920.1 Homo sapiens mitochondrion, comple...   4875   4899     25   \n",
       "19  NC_012920.1 Homo sapiens mitochondrion, comple...   5187   5255     69   \n",
       "22  NC_012920.1 Homo sapiens mitochondrion, comple...   5424   5499     76   \n",
       "23  NC_012920.1 Homo sapiens mitochondrion, comple...   5875   5916     42   \n",
       "24  NC_012920.1 Homo sapiens mitochondrion, comple...   6213   6242     30   \n",
       "25  NC_012920.1 Homo sapiens mitochondrion, comple...   6410   6464     55   \n",
       "26  NC_012920.1 Homo sapiens mitochondrion, comple...   6545   6572     28   \n",
       "27  NC_012920.1 Homo sapiens mitochondrion, comple...   6826   6851     26   \n",
       "28  NC_012920.1 Homo sapiens mitochondrion, comple...   7386   7427     42   \n",
       "29  NC_012920.1 Homo sapiens mitochondrion, comple...   7798   7840     43   \n",
       "30  NC_012920.1 Homo sapiens mitochondrion, comple...   7936   7974     39   \n",
       "31  NC_012920.1 Homo sapiens mitochondrion, comple...   8252   8305     54   \n",
       "32  NC_012920.1 Homo sapiens mitochondrion, comple...   8455   8488     34   \n",
       "35  NC_012920.1 Homo sapiens mitochondrion, comple...   8791   8815     25   \n",
       "36  NC_012920.1 Homo sapiens mitochondrion, comple...   8917   8958     42   \n",
       "39  NC_012920.1 Homo sapiens mitochondrion, comple...   9505   9607    103   \n",
       "40  NC_012920.1 Homo sapiens mitochondrion, comple...  10175  10214     40   \n",
       "41  NC_012920.1 Homo sapiens mitochondrion, comple...  10257  10303     47   \n",
       "42  NC_012920.1 Homo sapiens mitochondrion, comple...  10603  10637     35   \n",
       "43  NC_012920.1 Homo sapiens mitochondrion, comple...  10912  10990     79   \n",
       "44  NC_012920.1 Homo sapiens mitochondrion, comple...  11122  11162     41   \n",
       "45  NC_012920.1 Homo sapiens mitochondrion, comple...  11212  11249     38   \n",
       "46  NC_012920.1 Homo sapiens mitochondrion, comple...  11492  11555     64   \n",
       "47  NC_012920.1 Homo sapiens mitochondrion, comple...  11655  11692     38   \n",
       "48  NC_012920.1 Homo sapiens mitochondrion, comple...  12063  12124     62   \n",
       "49  NC_012920.1 Homo sapiens mitochondrion, comple...  12346  12450    105   \n",
       "50  NC_012920.1 Homo sapiens mitochondrion, comple...  12542  12573     32   \n",
       "53  NC_012920.1 Homo sapiens mitochondrion, comple...  13107  13150     44   \n",
       "54  NC_012920.1 Homo sapiens mitochondrion, comple...  13430  13458     29   \n",
       "55  NC_012920.1 Homo sapiens mitochondrion, comple...  13625  13705     81   \n",
       "56  NC_012920.1 Homo sapiens mitochondrion, comple...  13735  13809     75   \n",
       "57  NC_012920.1 Homo sapiens mitochondrion, comple...  13959  14004     46   \n",
       "58  NC_012920.1 Homo sapiens mitochondrion, comple...  14049  14074     26   \n",
       "59  NC_012920.1 Homo sapiens mitochondrion, comple...  14109  14169     61   \n",
       "60  NC_012920.1 Homo sapiens mitochondrion, comple...  14225  14292     68   \n",
       "66  NC_012920.1 Homo sapiens mitochondrion, comple...  14751  14832     82   \n",
       "67  NC_012920.1 Homo sapiens mitochondrion, comple...  15360  15391     32   \n",
       "68  NC_012920.1 Homo sapiens mitochondrion, comple...  15502  15556     55   \n",
       "69  NC_012920.1 Homo sapiens mitochondrion, comple...  15636  15683     48   \n",
       "70  NC_012920.1 Homo sapiens mitochondrion, comple...  16167  16211     45   \n",
       "71  NC_012920.1 Homo sapiens mitochondrion, comple...  16239  16301     63   \n",
       "72  NC_012920.1 Homo sapiens mitochondrion, comple...  16344  16382     39   \n",
       "\n",
       "   strand  score  max_score  threshold  window  \\\n",
       "2       - -1.150      -1.88          1      25   \n",
       "3       - -1.210      -1.16          1      25   \n",
       "4       - -1.390      -1.48          1      25   \n",
       "5       - -1.020      -1.48          1      25   \n",
       "6       - -1.520      -1.40          1      25   \n",
       "7       - -1.000      -1.00          1      25   \n",
       "10      - -1.170      -2.76          1      25   \n",
       "11      - -0.917      -1.28          1      25   \n",
       "12      - -0.882      -1.20          1      25   \n",
       "15      - -1.210      -1.28          1      25   \n",
       "16      - -1.220      -1.28          1      25   \n",
       "17      - -1.100      -1.36          1      25   \n",
       "18      - -1.080      -1.08          1      25   \n",
       "19      - -0.957      -1.24          1      25   \n",
       "22      - -1.200      -1.92          1      25   \n",
       "23      - -0.881      -1.24          1      25   \n",
       "24      - -0.900      -1.04          1      25   \n",
       "25      - -1.050      -1.44          1      25   \n",
       "26      - -1.210      -1.24          1      25   \n",
       "27      - -1.040      -1.04          1      25   \n",
       "28      - -1.020      -1.56          1      25   \n",
       "29      - -0.977      -1.32          1      25   \n",
       "30      - -0.923      -1.20          1      25   \n",
       "31      - -1.220      -2.00          1      25   \n",
       "32      - -1.090      -1.36          1      25   \n",
       "35      - -1.000      -1.00          1      25   \n",
       "36      - -0.881      -1.56          1      25   \n",
       "39      - -1.020      -1.96          1      25   \n",
       "40      - -1.200      -1.80          1      25   \n",
       "41      - -0.894      -1.36          1      25   \n",
       "42      - -1.060      -1.28          1      25   \n",
       "43      - -1.330      -2.08          1      25   \n",
       "44      - -0.854      -1.12          1      25   \n",
       "45      - -0.947      -1.24          1      25   \n",
       "46      - -1.000      -1.32          1      25   \n",
       "47      - -0.816      -1.20          1      25   \n",
       "48      - -1.210      -1.60          1      25   \n",
       "49      - -1.060      -1.76          1      25   \n",
       "50      - -1.060      -1.20          1      25   \n",
       "53      - -0.909      -1.24          1      25   \n",
       "54      - -1.030      -1.04          1      25   \n",
       "55      - -1.050      -1.68          1      25   \n",
       "56      - -1.170      -1.84          1      25   \n",
       "57      - -0.935      -1.28          1      25   \n",
       "58      - -1.120      -1.12          1      25   \n",
       "59      - -0.951      -1.20          1      25   \n",
       "60      - -1.070      -1.52          1      25   \n",
       "66      - -1.110      -1.84          1      25   \n",
       "67      - -1.120      -1.36          1      25   \n",
       "68      - -1.450      -2.12          1      25   \n",
       "69      - -0.896      -1.16          1      25   \n",
       "70      - -1.160      -1.80          1      25   \n",
       "71      - -0.968      -1.40          1      25   \n",
       "72      - -1.210      -1.48          1      25   \n",
       "\n",
       "                                             sequence  ...  \\\n",
       "2   CACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCC...  ...   \n",
       "3                            CGCTTAGCCTAGCCACACCCCCAC  ...   \n",
       "4                     CACCCCCTCCCCAATAAAGCTAAAACTCACC  ...   \n",
       "5           CTGAGCTAAACCTAGCCCCAAACCCACTCCACCTTACTACC  ...   \n",
       "6                             CCCACAGAACCCTCTAAATCCCC  ...   \n",
       "7                           CAACACCCACTACCTAAAAAATCCC  ...   \n",
       "10  CTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTC...  ...   \n",
       "11               CTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCC  ...   \n",
       "12                 CGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCC  ...   \n",
       "15                  CCCATTACAATCTCCAGCATTCCCCCTCAAACC  ...   \n",
       "16                        CTCCGTGCCACCTATCACACCCCATCC  ...   \n",
       "17  CCCTTCCCGTACTAATTAATCCCCTGGCCCAACCCGTCATCTACTC...  ...   \n",
       "18                          CTAGCCCCCATCTCAATCATATACC  ...   \n",
       "19  CTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGC...  ...   \n",
       "22  CATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCA...  ...   \n",
       "23         CACTCAGCCATTTTACCTCACCCCCACTGATGTTCGCCGACC  ...   \n",
       "24                     CTCTTACCTCCCTCTCTCCTACTCCTGCTC  ...   \n",
       "25  CAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTC...  ...   \n",
       "26                       CCTCAACACCACCTTCTTCGACCCCGCC  ...   \n",
       "27                         CTACCATAATCATCGCTATCCCCACC  ...   \n",
       "28         CTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCC  ...   \n",
       "29        CATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTAC  ...   \n",
       "30            CTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACC  ...   \n",
       "31  CCCGTATTTACCCTATAGCACCCCCTCTACCCCCTCTAGAGCCCAC...  ...   \n",
       "32                 CACAAACTACCACCTACCTCCCTCACCAAAGCCC  ...   \n",
       "35                          CCTCACTCATTTACACCAACCACCC  ...   \n",
       "36         CAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATC  ...   \n",
       "39  CCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCA...  ...   \n",
       "40           CGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCC  ...   \n",
       "41    CTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAAC  ...   \n",
       "42                CTCTCATAACCCTCAACACCCACTCCCTCTTAGCC  ...   \n",
       "43  CTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTA...  ...   \n",
       "44          CTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCC  ...   \n",
       "45             CTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCAC  ...   \n",
       "46  CGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACC...  ...   \n",
       "47             CAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGC  ...   \n",
       "48  CCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAA...  ...   \n",
       "49  CACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCA...  ...   \n",
       "50                   CCACAACCCAAACAACCCAGCTCTCCCTAAGC  ...   \n",
       "53       CTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCAC  ...   \n",
       "54                      CCATACCTCTCACTTCAACCTCCCTCACC  ...   \n",
       "55  CCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGA...  ...   \n",
       "56  CTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAA...  ...   \n",
       "57     CCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACC  ...   \n",
       "58                         CTCCACCTCCATCATCACCTCAACCC  ...   \n",
       "59  CTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAACCTATT...  ...   \n",
       "60  CGCCCATAATCATACAAAGCCCCCGCACCAATAGGATCCTCCCGAA...  ...   \n",
       "66  CCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTC...  ...   \n",
       "67                   CAAACAACCCCCTAGGAATCACCTCCCATTCC  ...   \n",
       "68  CCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCAC...  ...   \n",
       "69   CCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAAC  ...   \n",
       "70      CCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTAC  ...   \n",
       "71  CAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAA...  ...   \n",
       "72            CAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTC  ...   \n",
       "\n",
       "                          note start-100bp end+100bp  start-50bp  end+50bp  \\\n",
       "2   5 overlapping g4 in africa         318       692         368       642   \n",
       "3                          NaN         685       908         735       858   \n",
       "4                          NaN         854      1084         904      1034   \n",
       "5          more than in africa        1567      1807        1617      1757   \n",
       "6                          NaN        1957      2179        2007      2129   \n",
       "7                          NaN        2112      2336        2162      2286   \n",
       "10  5 overlapping g4 in africa        3371      3704        3421      3654   \n",
       "11                         NaN        3774      4009        3824      3959   \n",
       "12            more than africa        3851      4084        3901      4034   \n",
       "15                         NaN        4128      4360        4178      4310   \n",
       "16                         NaN        4272      4498        4322      4448   \n",
       "17  3 overlapping g4 in africa        4356      4605        4406      4555   \n",
       "18                         NaN        4775      4999        4825      4949   \n",
       "19  2 overlapping g4 in africa        5087      5355        5137      5305   \n",
       "22              2 g4 in africa        5324      5599        5374      5549   \n",
       "23            more than africa        5775      6016        5825      5966   \n",
       "24  2 overlapping g4 in africa        6113      6342        6163      6292   \n",
       "25            more than africa        6310      6564        6360      6514   \n",
       "26            less than africa        6445      6672        6495      6622   \n",
       "27                         NaN        6726      6951        6776      6901   \n",
       "28            more than africa        7286      7527        7336      7477   \n",
       "29                         NaN        7698      7940        7748      7890   \n",
       "30                         NaN        7836      8074        7886      8024   \n",
       "31            more than africa        8152      8405        8202      8355   \n",
       "32                         NaN        8355      8588        8405      8538   \n",
       "35                         NaN        8691      8915        8741      8865   \n",
       "36            more than africa        8817      9058        8867      9008   \n",
       "39              2 g4 in africa        9405      9707        9455      9657   \n",
       "40            more than africa       10075     10314       10125     10264   \n",
       "41            more than africa       10157     10403       10207     10353   \n",
       "42                         NaN       10503     10737       10553     10687   \n",
       "43              2 g4 in africa       10812     11090       10862     11040   \n",
       "44            more than africa       11022     11262       11072     11212   \n",
       "45            more than africa       11112     11349       11162     11299   \n",
       "46  4 overlapping g4 in africa       11392     11655       11442     11605   \n",
       "47            more than africa       11555     11792       11605     11742   \n",
       "48            more than africa       11963     12224       12013     12174   \n",
       "49  2 overlapping g4 in africa       12246     12550       12296     12500   \n",
       "50            more than africa       12442     12673       12492     12623   \n",
       "53  2 overlapping g4 in africa       13007     13250       13057     13200   \n",
       "54  2 overlapping g4 in africa       13330     13558       13380     13508   \n",
       "55            more than africa       13525     13805       13575     13755   \n",
       "56            more than africa       13635     13909       13685     13859   \n",
       "57            more than africa       13859     14104       13909     14054   \n",
       "58                         NaN       13949     14174       13999     14124   \n",
       "59  2 overlapping g4 in africa       14009     14269       14059     14219   \n",
       "60  3 overlapping g4 in africa       14125     14392       14175     14342   \n",
       "66   3 overlapping g4 in afica       14651     14932       14701     14882   \n",
       "67                         NaN       15260     15491       15310     15441   \n",
       "68                         NaN       15402     15656       15452     15606   \n",
       "69            more than africa       15536     15783       15586     15733   \n",
       "70            more than africa       16067     16311       16117     16261   \n",
       "71            more than africa       16139     16401       16189     16351   \n",
       "72            more than africa       16244     16482       16294     16432   \n",
       "\n",
       "    bp_untill_next_G4  bp_untill_previous_G4  G4_id  up_win  down_win  \n",
       "2                 193                     40      3     378       692  \n",
       "3                 146                    193      4     685       908  \n",
       "4                 683                    146      5     854      1084  \n",
       "5                 350                    683      6    1567      1807  \n",
       "6                 133                    350      7    1957      2179  \n",
       "7                 912                    133      8    2112      2336  \n",
       "10                270                    249     11    3371      3704  \n",
       "11                 42                    270     12    3774      3951  \n",
       "12                102                     42     13    3909      4084  \n",
       "15                112                     67     16    4161      4360  \n",
       "16                 58                    112     17    4272      4456  \n",
       "17                370                     58     18    4398      4605  \n",
       "18                288                    370     19    4775      4999  \n",
       "19                 33                    288     20    5087      5288  \n",
       "22                376                     32     23    5392      5599  \n",
       "23                297                    376     24    5775      6016  \n",
       "24                168                    297     25    6113      6342  \n",
       "25                 81                    168     26    6310      6545  \n",
       "26                254                     81     27    6464      6672  \n",
       "27                535                    254     28    6726      6951  \n",
       "28                371                    535     29    7286      7527  \n",
       "29                 96                    371     30    7698      7936  \n",
       "30                278                     96     31    7840      8074  \n",
       "31                150                    278     32    8152      8405  \n",
       "32                 56                    150     33    8355      8544  \n",
       "35                102                    148     36    8691      8915  \n",
       "36                285                    102     37    8817      9058  \n",
       "39                568                    212     40    9405      9707  \n",
       "40                 43                    568     41   10075     10257  \n",
       "41                300                     43     42   10214     10403  \n",
       "42                275                    300     43   10503     10737  \n",
       "43                132                    275     44   10812     11090  \n",
       "44                 50                    132     45   11022     11212  \n",
       "45                243                     50     46   11162     11349  \n",
       "46                100                    243     47   11392     11655  \n",
       "47                371                    100     48   11555     11792  \n",
       "48                222                    371     49   11963     12224  \n",
       "49                 92                    222     50   12246     12542  \n",
       "50                378                     92     51   12450     12673  \n",
       "53                280                     25     54   13082     13250  \n",
       "54                167                    280     55   13330     13558  \n",
       "55                 30                    167     56   13525     13735  \n",
       "56                150                     30     57   13705     13909  \n",
       "57                 45                    150     58   13859     14049  \n",
       "58                 35                     45     59   14004     14109  \n",
       "59                 56                     35     60   14074     14225  \n",
       "60                 36                     56     61   14169     14328  \n",
       "66                528                     98     67   14653     14932  \n",
       "67                111                    528     68   15260     15491  \n",
       "68                 80                    111     69   15402     15636  \n",
       "69                484                     80     70   15556     15783  \n",
       "70                 28                    484     71   16067     16239  \n",
       "71                 43                     28     72   16211     16344  \n",
       "72                145                     43     73   16301     16482  \n",
       "\n",
       "[54 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_25bp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b7c62411-85ae-47ce-be2d-a8ff82ff1582",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_25bp['start'] = df_25bp['start'].apply(lambda x: x-25)\n",
    "# df_25bp['end'] = df_25bp['end'].apply(lambda x: x+25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3aee0d0d-a77a-467f-be17-1f1434aa79a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_start = []\n",
    "\n",
    "for start, end, G4_id in zip(df_25bp['start'], \n",
    "                             df_25bp['end'], \n",
    "                             df_25bp['G4_id']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i in range(start-25, end+25):\n",
    "        row = []\n",
    "        \n",
    "        if i+WINDOW_SIZE <= end+25:\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            break\n",
    "\n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            if pos != 3107:\n",
    "                window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "            else:\n",
    "                window_seq += ''\n",
    "        \n",
    "        # print(i, window, WINDOW_SIZE, end)  \n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('check_start')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_check_start.append(row)\n",
    "# print(data_check_start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "05d43e76-5d2a-4d22-9cb8-d017a355dfc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_end = []\n",
    "\n",
    "for start, end, G4_id in zip(df_25bp['start'], \n",
    "                             df_25bp['end'], \n",
    "                             df_25bp['G4_id']):\n",
    "    \n",
    "    number_of_win = 0\n",
    "    \n",
    "    for i in range(end+25, start-25, -1):\n",
    "        row = []\n",
    "\n",
    "        \n",
    "        if i+WINDOW_SIZE > end+25:\n",
    "            continue  \n",
    "        elif (i+WINDOW_SIZE <= end+25):\n",
    "            window = [*range(i, i + WINDOW_SIZE)]       \n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        window_seq = ''\n",
    "        for pos in window:\n",
    "            if pos != 3107:\n",
    "                window_seq += str(ref_annot[ref_annot['Pos'] == pos].Nuc.values[0])\n",
    "            else:\n",
    "                window_seq += ''\n",
    "        \n",
    "        number_of_win += 1\n",
    "\n",
    "        row.append(G4_id)\n",
    "        row.append('check_end')\n",
    "        row.append(number_of_win)\n",
    "        row.append(window[0])\n",
    "        row.append(window[-1])\n",
    "        row.append(window_seq)\n",
    "        \n",
    "        data_check_end.append(row)\n",
    "# print(data_check_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "51dad145-e328-43d4-861b-5a0427c3b1fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_check_sides = data_check_start + data_check_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc4b7853-81ed-4d14-97e6-e2e9ab836451",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_check_sides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b22715ac-2844-46f0-a0be-8962237f98d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = pd.DataFrame(data=data_check_sides, \n",
    "                             columns = ['G4_id', 'side', 'number_of_win', 'window_start', 'window_end', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "00dd3f17-098e-48af-9313-7bfefbb2a03a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4901</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>1</td>\n",
       "      <td>607</td>\n",
       "      <td>616</td>\n",
       "      <td>CACTGAAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4902</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>2</td>\n",
       "      <td>606</td>\n",
       "      <td>615</td>\n",
       "      <td>ACACTGAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4903</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>3</td>\n",
       "      <td>605</td>\n",
       "      <td>614</td>\n",
       "      <td>TACACTGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4904</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>4</td>\n",
       "      <td>604</td>\n",
       "      <td>613</td>\n",
       "      <td>ATACACTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4905</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>5</td>\n",
       "      <td>603</td>\n",
       "      <td>612</td>\n",
       "      <td>AATACACTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5110</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>210</td>\n",
       "      <td>398</td>\n",
       "      <td>407</td>\n",
       "      <td>TTTTATCTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5111</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>211</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5112</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>212</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5113</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>213</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5114</th>\n",
       "      <td>3</td>\n",
       "      <td>check_end</td>\n",
       "      <td>214</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id       side  number_of_win  window_start  window_end    sequence\n",
       "4901      3  check_end              1           607         616  CACTGAAAAT\n",
       "4902      3  check_end              2           606         615  ACACTGAAAA\n",
       "4903      3  check_end              3           605         614  TACACTGAAA\n",
       "4904      3  check_end              4           604         613  ATACACTGAA\n",
       "4905      3  check_end              5           603         612  AATACACTGA\n",
       "...     ...        ...            ...           ...         ...         ...\n",
       "5110      3  check_end            210           398         407  TTTTATCTTT\n",
       "5111      3  check_end            211           397         406  ATTTTATCTT\n",
       "5112      3  check_end            212           396         405  AATTTTATCT\n",
       "5113      3  check_end            213           395         404  AAATTTTATC\n",
       "5114      3  check_end            214           394         403  CAAATTTTAT\n",
       "\n",
       "[214 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_check_sides[(df_win_check_sides['side'] == 'check_end') & (df_win_check_sides['G4_id'] == 3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "358e7e47-6499-4e36-9995-c248bfa19092",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_win_check_sides = df_win_check_sides['sequence'].apply(lambda x: Counter(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "45f31c06-2989-4e49-9cc6-f99a3a499ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_count_df_win_check_sides = pd.DataFrame.from_records(nuc_count_win_check_sides).sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "56b82dde-2e52-493f-8621-b2a785802bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = df_win_check_sides.join(nuc_count_df_win_check_sides).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "07ae5c08-8751-49d0-82e9-5988297cd0a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>1</td>\n",
       "      <td>393</td>\n",
       "      <td>402</td>\n",
       "      <td>TCAAATTTTA</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>2</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>3</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>4</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>5</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9743</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>74</td>\n",
       "      <td>16324</td>\n",
       "      <td>16333</td>\n",
       "      <td>TTACCGTACA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9744</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>75</td>\n",
       "      <td>16323</td>\n",
       "      <td>16332</td>\n",
       "      <td>TTTACCGTAC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9745</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>76</td>\n",
       "      <td>16322</td>\n",
       "      <td>16331</td>\n",
       "      <td>ATTTACCGTA</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9746</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>77</td>\n",
       "      <td>16321</td>\n",
       "      <td>16330</td>\n",
       "      <td>CATTTACCGT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9747</th>\n",
       "      <td>73</td>\n",
       "      <td>check_end</td>\n",
       "      <td>78</td>\n",
       "      <td>16320</td>\n",
       "      <td>16329</td>\n",
       "      <td>CCATTTACCG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9748 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      G4_id         side  number_of_win  window_start  window_end    sequence  \\\n",
       "0         3  check_start              1           393         402  TCAAATTTTA   \n",
       "1         3  check_start              2           394         403  CAAATTTTAT   \n",
       "2         3  check_start              3           395         404  AAATTTTATC   \n",
       "3         3  check_start              4           396         405  AATTTTATCT   \n",
       "4         3  check_start              5           397         406  ATTTTATCTT   \n",
       "...     ...          ...            ...           ...         ...         ...   \n",
       "9743     73    check_end             74         16324       16333  TTACCGTACA   \n",
       "9744     73    check_end             75         16323       16332  TTTACCGTAC   \n",
       "9745     73    check_end             76         16322       16331  ATTTACCGTA   \n",
       "9746     73    check_end             77         16321       16330  CATTTACCGT   \n",
       "9747     73    check_end             78         16320       16329  CCATTTACCG   \n",
       "\n",
       "        A    C    G    T  \n",
       "0     4.0  1.0  0.0  5.0  \n",
       "1     4.0  1.0  0.0  5.0  \n",
       "2     4.0  1.0  0.0  5.0  \n",
       "3     3.0  1.0  0.0  6.0  \n",
       "4     2.0  1.0  0.0  7.0  \n",
       "...   ...  ...  ...  ...  \n",
       "9743  3.0  3.0  1.0  3.0  \n",
       "9744  2.0  3.0  1.0  4.0  \n",
       "9745  3.0  2.0  1.0  4.0  \n",
       "9746  2.0  3.0  1.0  4.0  \n",
       "9747  2.0  4.0  1.0  3.0  \n",
       "\n",
       "[9748 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_check_sides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "60162525-7f65-423f-b8d1-de2ab56c24e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_check_sides.to_csv('../data/processed/df_win_check_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65bfe663-615f-49cc-a352-7e56b9a2db5a",
   "metadata": {},
   "source": [
    "Collect mutations from cancer dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ec18b2df-8934-43bf-ac78-663d5f46ee5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cancers = pd.read_csv('../data/raw/mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0b3f4996-771e-4be3-84ec-11ba5e205960",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_check_sides = pd.read_csv('../data/processed/df_win_check_sides.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1f9bb812-c75a-48e2-860e-b173bb19fca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_check_sides = []\n",
    "for window_start, window_end in zip(df_win_check_sides['window_start'], df_win_check_sides['window_end']):\n",
    "    row = [*range(window_start, window_end)]\n",
    "    Mut_win_check_sides.append(str(df_cancers[df_cancers['Pos'].isin(row)].Mut.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0f238d1e-4d9c-47a3-bfad-cad3a809ba60",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_check_sides = pd.DataFrame(Mut_win_check_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "22564adc-bbdd-4285-ab84-ac9094f5b063",
   "metadata": {},
   "outputs": [],
   "source": [
    "Mut_win_df_check_sides[0] = Mut_win_df_check_sides[0].str.findall(\"[A-T]>[A-T]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "15b6e619-bfe3-47ab-abdc-e285c92fa083",
   "metadata": {},
   "outputs": [],
   "source": [
    "obsMutwin_check_sides = pd.DataFrame.from_records(Mut_win_df_check_sides[0].apply(lambda x: Counter(x))) \\\n",
    "    .fillna(0) \\\n",
    "    .sort_index(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3220bf42-2b40-4f03-be42-9f42b350ad09",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_check_sides = df_win_check_sides.join(obsMutwin_check_sides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9f520752-d0df-489a-8a98-250e2a33eff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_win_cancer_check_sides['number_of_Mut'] = df_win_cancer_check_sides[['A>C','A>G','A>T',\n",
    "                                                              'C>A','C>G','C>T',\n",
    "                                                              'G>A','G>C','G>T',\n",
    "                                                              'T>A','T>C','T>G']].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2e333902-32ab-4cb6-b040-a1460db43805",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>G4_id</th>\n",
       "      <th>side</th>\n",
       "      <th>number_of_win</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "      <th>sequence</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>...</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>number_of_Mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>1</td>\n",
       "      <td>393</td>\n",
       "      <td>402</td>\n",
       "      <td>TCAAATTTTA</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>2</td>\n",
       "      <td>394</td>\n",
       "      <td>403</td>\n",
       "      <td>CAAATTTTAT</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>3</td>\n",
       "      <td>395</td>\n",
       "      <td>404</td>\n",
       "      <td>AAATTTTATC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>4</td>\n",
       "      <td>396</td>\n",
       "      <td>405</td>\n",
       "      <td>AATTTTATCT</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>check_start</td>\n",
       "      <td>5</td>\n",
       "      <td>397</td>\n",
       "      <td>406</td>\n",
       "      <td>ATTTTATCTT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   G4_id         side  number_of_win  window_start  window_end    sequence  \\\n",
       "0      3  check_start              1           393         402  TCAAATTTTA   \n",
       "1      3  check_start              2           394         403  CAAATTTTAT   \n",
       "2      3  check_start              3           395         404  AAATTTTATC   \n",
       "3      3  check_start              4           396         405  AATTTTATCT   \n",
       "4      3  check_start              5           397         406  ATTTTATCTT   \n",
       "\n",
       "     A    C    G    T  ...  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G  \\\n",
       "0  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "1  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "2  4.0  1.0  0.0  5.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3  3.0  1.0  0.0  6.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "4  2.0  1.0  0.0  7.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "\n",
       "   number_of_Mut  \n",
       "0            0.0  \n",
       "1            0.0  \n",
       "2            1.0  \n",
       "3            1.0  \n",
       "4            1.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_win_cancer_check_sides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7be9f257-ac80-4d4b-8bb3-38b245f1ab31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_win_cancer_check_sides.to_csv('../data/processed/df_win_cancer_check_sides.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b97fbb8e-0bed-4e14-805f-bd06e7af308a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
